{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# find tracerRNA\n",
    "\n",
    "example id:\n",
    "8774__OBKKDPJE_00146__BritoIL_2016__M2.46.ST__bin.47;length1073;#sequences1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First created Thu Jul 11 09:51:15 CEST 2019\n",
    "# Made by L-F-S\n",
    "# At the University Of Trento, Italy\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna, generic_protein\n",
    "\n",
    "\n",
    "sys.path.insert(0, '/home/lorenzo.signorini/cas_mining/utils/')\n",
    "import filename_discrepancies\n",
    "\n",
    "feature=\"Cas9\" #WARNING!!! CHANGE THIS!!\n",
    "#outpath=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/5caslocitable/out/\"+feature+\"/\"\n",
    "datadir=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Seq ID', 'Seq Description', 'Seq', 'Contig', 'Genome Name', 'Study',\n",
       "       'Sample Name', 'SGB ID', 'pilercr_CRISPR', 'minced_CRISPR',\n",
       "       'prokka_cas1', 'prokka_cas2', 'uSGB', 'Level of estimated taxonomy',\n",
       "       'Estimated taxonomy'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import data\n",
    "cas_dataset=pd.read_csv(\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/5caslocitable/known_\"+feature+\"_variants_table.csv\", index_col=0)\n",
    "cas_dataset.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0.1', 'Genome Name', 'Study', 'Sample Name', 'Assembler',\n",
       "       'Genome Size', 'N50', '# Contigs', 'Completeness', 'Contamination',\n",
       "       'Strain Heterogeneity', 'SGB ID', 'pilercr_CRISPR', 'minced_CRISPR',\n",
       "       'prokka_CRISPR', 'prokka_cas', 'uniref_cas'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 186,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tabellazza=pd.read_csv(datadir+\"/3tabellazza/crisprcas_hits_table.csv\", index_col=0)\n",
    "tabellazza.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SEQ ID:  OBKKDPJE_00146\n",
      "NODE_3_length_470079_cov_10.1699\n",
      "BritoIL_2016\n",
      "BritoIL_2016__M2.46.ST__bin.47\n",
      "1073\n",
      "[('.', '179452', '179677')]\n",
      "----------------------------------------------------------------------\n",
      "SEQ ID:  LPEAHEHG_01071\n",
      "NODE_190_length_65757_cov_25.3876\n",
      "ChengpingW_2017\n",
      "ChengpingW_2017__AS43raw__bin.51\n",
      "1074\n",
      "[('.', '9802', '10102')]\n",
      "----------------------------------------------------------------------\n",
      "SEQ ID:  CJFLGLGO_01367\n",
      "NODE_137_length_63930_cov_6.02239\n",
      "Obregon-TitoAJ_2015\n",
      "Obregon-TitoAJ_2015__SM40__bin.4\n",
      "1079\n",
      "[('.', '44486', '44718')]\n",
      "----------------------------------------------------------------------\n",
      "SEQ ID:  FLJLKKHA_00012\n",
      "NODE_49_length_176785_cov_12.2725\n",
      "BritoIL_2016\n",
      "BritoIL_2016__M1.29.ST__bin.30\n",
      "979\n",
      "[('.', '13625', '14320')]\n",
      "----------------------------------------------------------------------\n",
      "SEQ ID:  LJKGNKPJ_00564\n",
      "NODE_211_length_45519_cov_7.71687\n",
      "KarlssonFH_2013\n",
      "KarlssonFH_2013__S420__bin.8\n",
      "1002\n",
      "[('.', '12991', '13885')]\n",
      "----------------------------------------------------------------------\n",
      "SEQ ID:  CPMHNJAN_01529\n",
      "NODE_8014_length_5265_cov_2.43301\n",
      "LiJ_2014\n",
      "LiJ_2014__O2.UC47-1__bin.10\n",
      "1059\n",
      "[('.', '6', '637')]\n",
      "----------------------------------------------------------------------\n",
      "SEQ ID:  AKIPNPCL_00899\n",
      "NODE_1451_length_29619_cov_4.50203\n",
      "CosteaPI_2017\n",
      "CosteaPI_2017__SID713A009-11-90-0__bin.81\n",
      "1005\n",
      "[('.', '21722', '24134')]\n",
      "----------------------------------------------------------------------\n",
      "SEQ ID:  AJHGDLDC_00764\n",
      "NODE_3931_length_8539_cov_7.89097\n",
      "HMP_2012\n",
      "HMP_2012__SRS015065__bin.2\n",
      "1074\n",
      "[('.', '1618', '2313')]\n",
      "----------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "for seqid in [\"OBKKDPJE_00146\",\"LPEAHEHG_01071\",\"CJFLGLGO_01367\",\"FLJLKKHA_00012\",\"LJKGNKPJ_00564\",\"CPMHNJAN_01529\", \"AKIPNPCL_00899\", \"AJHGDLDC_00764\"]:\n",
    "    print(\"SEQ ID: \", seqid)\n",
    "    print(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Contig\"].iloc[0])\n",
    "    print(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Study\"].iloc[0])\n",
    "    print(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Genome Name\"].iloc[0])\n",
    "    print(len(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Seq\"].iloc[0]))\n",
    "    print(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"minced_CRISPR\"].iloc[0])\n",
    "    print(\"----------------------------------------------------------------------\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.Alphabet import RNAAlphabet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tOBKKDPJE_00146\n",
      "SGB:\t8774\n",
      "Genome Name:\t BritoIL_2016__M2.46.ST__bin.47\n",
      "Contig:\tNODE_3_length_470079_cov_10.1699 \n",
      "-----------------------------------------------------\n",
      "\n",
      "CRISPR spacer sequences:\n",
      " ['GTCAAAACTAATACGTTATAATCGGGGCGCGCCAGATG', 'GTCAAAACTTATGGCAAAATCTACAAAAATCGCAAACG', 'CTCAAAACTGGTTGCCATTTGGTTTGAAATTCCTCTGC'] \n",
      "CRISPR repeat sequences:\n",
      " ['AATCTAAGAAAGAACACTTTGCTGTAAA', 'AATCTAAGAAAGAACACTTTGCTGTAAA', 'AATCTAAGAAAGAACACTTTGCTGTAAA', 'AATCTAAGGAAGGTTAGTTTGCTAAATA']\n",
      "Retrieving locus\n",
      "CRISPR si trova 179452 179677\n",
      "Cas9 si trova [174787, 178008]\n",
      "Cas2 si trova [178941, 179336]\n",
      "Cas1 si trova [178008, 178898]\n",
      "1073.0 1073\n",
      "Cas locus on plus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? True\n",
      "MYTKEIIWGFDIGKGSLGEAVRVGNEFKHVQSLEIAPDFAETKTTAAARRAWRTREAHKARELWLEKCLSEAGIEVLLRRKVGVVDGAWKLVSKGDVRLEREFPPAGEDLCYNSIALRCKLLLGEKLESWQVFKALNSAIQNRGYDANIPWGDEDAADGGDGSSRKSSDSDASGDGESYAEKLAHYEREFDGILAELDNREKYEFPCFFKAAKMGLWSPENPDFVGVRIDNTAEKAKGYVMPREAVVREFRALVNAAALQYPNLKGKADFIMFGRACEPYASYNPELRKRFGLKRGAATDWTALGQKTPRFDNRVIDKCKLIPRLNVCKIKPLDECKTDDDFLSYTITLSLKLLNLRFFRGAGVDSLSFDEFLRAFELAKNSKFKLGKAALKKFFKTIGAEVMDESQSEIEQPRNGGRAAFSRPAMRILRDLIFSGKSPAEFYATAIENIANTDKNKGLVAGDLDFLKLMGDAPWSGIFIPDVGLSDFADASAAAREAQINKLIGSQNDPIVRHRLSFFYERIKSLEAKFGVPDKIVLEFVRDDFLGKKAKSDLNKAIKTRFDEKKKLAEDLDSKREEFGFKYASKKMLLKFELYKKQGGECVYTGNALPLSELESLEIEHIVPRSRGGPDAMYNYVLTDEKTNKEKGDRTPFEWRSTMGAEWESYCERVRKLTKELGAKRCKLLLSPNAAELVEKYTALAETAWISKLAQRICCLHFGFQFGGLAGAKKVWTVSGNTTAMIRGRYGLNRLLHDEESAEDLSLFDRICMEKKFEEKNRKNKKHHALDAMCLCFAPTSNAKKASAYKNILPEEITRNPESDEARKFFGKYLASVIPNKVAVKKSALEQTIYALRKGRIVKKCNVRDLAYTGINPKYNYSTLVKSIKSIVGEVPKRIVSDFVLSNSDLSESEWEDWCKNNAAIPSKNGAPTKLLRVLCDAGNPEEYKDLSKDGCGAYRKGDSHKGQFIWKTKQGRYVVAPVYAHASKAKVLAALKNNPDCAEVCGFFRQNCLVEIPNDVIDKSGKMLLKSGIYKSNTIMSDGRAKFTNSDGISTNPISIQYLMPAGMKRIKLDTV\n",
      "start:  174787\n",
      "end:  179677\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n",
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/8774__OBKKDPJE_00146__BritoIL_2016__M2.46.ST__bin.47__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11\n",
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['179452', '179518', '179584', '179650'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "and finally, just for sgb 15299, the tracrRNA CACAGGCTTTCGCCGCGCCTGACCGCAAGGAA\n",
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tLPEAHEHG_01071\n",
      "SGB:\t15299\n",
      "Genome Name:\t ChengpingW_2017__AS43raw__bin.51\n",
      "Contig:\tNODE_190_length_65757_cov_25.3876 \n",
      "-----------------------------------------------------\n",
      "\n",
      "CRISPR spacer sequences:\n",
      " ['CTTCCGTGCCATTTCCTCTATGACAGAATC', 'CTGCCCGGCGACGTGCGTGTGAGCGATACG', 'CCTGCCGCCCAATGCCCAGGACAACGAACT', 'AACGACTATTCAACGCACGGTAATAGCGCCT'] \n",
      "CRISPR repeat sequences:\n",
      " ['GGGATGATTTACTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAACAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT']\n",
      "Retrieving locus\n",
      "CRISPR si trova 9802 10102\n",
      "Cas9 si trova [11349, 14573]\n",
      "Cas2 si trova [10151, 10459]\n",
      "Cas1 si trova [10464, 11345]\n",
      "1074.0 1074\n",
      "Cas locus on minus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? True\n",
      "MKHPYGIGLDIGIASVGWAVVALNENAEPYGLIRCGSRIFDAAEQPNTGESLAAPRREARSTRRRLRRRSLRKADLYELMAQNGLPGRAEIEEAVQAGHLPDIYALRVQALDGPVTAMDFARILLHLMQRRGFRSNRKADDAQKDGKLLQAIDANTRRMEENHYRTVGEMMYRDPVFAEHKRNKSENYLSTVRRDQIVEEAVQLFAAQRQYGAAWASPKMEAEYLTILTRQRSFAEGPGGDSPYRGMQKVGPCALEGKSERRAPKATFAFEYFTLLQKINHIRIAENGVNRILTAEERQELLSACYRTDKLDFARIRKLLALPGEARFNMVRYRENQTAEECEKKEKITVLPCYHKMRKVLNTLRKDHIRSISREQLDAAATALTNPENEEKLRAALEEAQFEPLEVDALMALPSFSGFGHISVKACRKLIPYLEQGLNYNDACKEAGYDFQGNYRGEKSQFLPASTEEMEDITSPVVRRAVAQTIKVVNAIIREQGESPVNIHLELAREMSKNYKQRNELADAMEKNQAENARLMEKLHDLFRGRTITGQTLVKYRLWKEQKEVCAYSLQMMKLDSVITDSSYAEVDHIVPYSRSFDDRRTNKVLVLTSENRKKGDRLPLEYLQGKRREDFIVYTKANVKNYRKRQNLLKEGLSKEESREFIQRNLQDTQYSASFMLNYIRNHLAFADCSAAGKQRVVAVNGAVTAFLRKRWGLSKVRADGDLHHAVDATVIACTTPSMVKRVTEFCKQEETNHVRNEYFPEPWPRFRDELMQRLSACPQENLMQINPVYYQNVDIASIRPVFVSRMPRHKMTGAAHKETIKGRLDDAHTVQRRSITDLKLDKEGEIAGYFNPSSDTLLYNALKEQLTAFGGDGKKAFAEPFFKPRADGTPGAQVRKVKVYDKATSMVAVHEGKGVADNDTMVRIDVYFMPGDGYYWVPIYVADTVKPELPNKAVIAHKNYAEWKEMDEKDFLFSLCPNDLIRVESKRTMKFKVQNKDSTLEKEMQTNGDMVYFEGGNISGGSITVATHDGAYMVESLGLKTLQKVQKYQVDALGNYTLVKKEKRQYFPAQRR\n",
      "start:  9802\n",
      "end:  14573\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n",
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/15299__LPEAHEHG_01071__ChengpingW_2017__AS43raw__bin.51__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11\n",
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['9802', '9868', '9934', '10000', '10067'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "rpt1|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t100.00\t36\t25\t0\t0\tCTAATAGCTCTTGGTATGGTATAAT\tCTAATAGCTCTTGGTATGGTATAAT\t9945\t9969\t9e-09\tplus\n",
      "\n",
      "rpt1|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t100.00\t36\t25\t0\t0\tCTAATAGCTCTTGGTATGGTATAAT\tCTAATAGCTCTTGGTATGGTATAAT\t10011\t10035\t9e-09\tplus\n",
      "\n",
      "rpt1|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t100.00\t36\t25\t0\t0\tCTAATAGCTCTTGGTATGGTATAAT\tCTAATAGCTCTTGGTATGGTATAAT\t10078\t10102\t9e-09\tplus\n",
      "\n",
      "rpt1|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t96.00\t36\t25\t1\t0\tCTAATAGCTCTTGGTATGGTATAAT\tCTAACAGCTCTTGGTATGGTATAAT\t9879\t9903\t4e-07\tplus\n",
      "\n",
      "rpt1|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t95.65\t36\t23\t1\t0\tTAATAGCTCTTGGTATGGTATAA\tTAATAGTTCTTGGTATGGTATAA\t14729\t14751\t5e-06\tplus\n",
      "\n",
      "rpt2|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t96.00\t36\t25\t1\t0\tCTAACAGCTCTTGGTATGGTATAAT\tCTAATAGCTCTTGGTATGGTATAAT\t9813\t9837\t4e-07\tplus\n",
      "\n",
      "rpt2|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t91.30\t36\t23\t2\t0\tTAACAGCTCTTGGTATGGTATAA\tTAATAGTTCTTGGTATGGTATAA\t14729\t14751\t2e-04\tplus\n",
      "\n",
      "rpt3|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t100.00\t36\t25\t0\t0\tCTAATAGCTCTTGGTATGGTATAAT\tCTAATAGCTCTTGGTATGGTATAAT\t9813\t9837\t9e-09\tplus\n",
      "\n",
      "rpt3|NODE_190_length_65757_cov_25.3876|ChengpingW_2017__AS43raw__bin.51|LPEAHEHG_01071\tNODE_190_length_65757_cov_25.3876\t95.65\t36\t23\t1\t0\tTAATAGCTCTTGGTATGGTATAA\tTAATAGTTCTTGGTATGGTATAA\t14729\t14751\t5e-06\tplus\n",
      "\n",
      "and finally, just for sgb 15299, the tracrRNA TAGCAACCATAATAGTTCTTGGTATGGTATAA\n",
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tCJFLGLGO_01367\n",
      "SGB:\t4329\n",
      "Genome Name:\t Obregon-TitoAJ_2015__SM40__bin.4\n",
      "Contig:\tNODE_137_length_63930_cov_6.02239 \n",
      "-----------------------------------------------------\n",
      "\n",
      "CRISPR spacer sequences:\n",
      " ['AAGGTAATGGTCTTACCGAAGCTAATCTT', 'CCGGCACCAACCGACTCGTCACTCATGACC', 'GAATTAAGATTCAAAGGGTCAGTAGGAAGA'] \n",
      "CRISPR repeat sequences:\n",
      " ['ATTATAGTTTCTTAATTTTTCTTGGTATGTTATAAT', 'ATTGTAGTTCCCTAATTTTTCTTGGTATATTATAAT', 'ATTGTAGTTCCCTAATTTTTCTTGGTATGTTATAAT', 'ATTGTAGTTCCCTAATTTTTCTTGGTATGTTATAAT']\n",
      "Retrieving locus\n",
      "CRISPR si trova 44486 44718\n",
      "Cas9 si trova [46031, 49270]\n",
      "Cas2 si trova [44816, 45136]\n",
      "Cas1 si trova [45129, 46028]\n",
      "1079.0 1079\n",
      "Cas locus on minus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? True\n",
      "MKYGIGLDCGIASVGYCITELDSNDEPKRIVRLGSRIFDKAENPKDGSSLAKPRREARGLRRRIRRHQHRLQRIRYILVSDGIVTQDELDTMFQGQLSDIYEIRARALDEPISNVEFARVLINLAQRRGFKSNRKVDENTKDKETGKLLGAIEKNRENIAKNGYRTVGEMLYKDERYSKYKRNKGEDYLNTVSRDMIADEIKMIFSAQRSFRMPFATEEIEARYTDIVMSQRPFDLGPGEGNEKSPSPYAGNQIEKMVGRCTFFPDEYRAVKASYSFQLFSLLQGINNITLVDDGGNAFPLSQQERNELKEYCFKTKSVTYASIRKKLKISPDFKFKNITYTEKGVEESEKKTKFQHLNIYHQMKSTLGDAMNSLSHDELDEIGRIFTFYKNDTKIIEALEKTDIDKDLYPALLNLPSFSKTGHISVKACKLLIPFLEEGKTYNEACECAGLDFKAHSNAEKKILLPPKSDELDDIVNPVVRRAVSQTVKVVNAIIREMGTSPTYLNIELAQELSKSKKERDEIEKNYLLNRAKNEKIKKEIVDNFGFEPKGQDIVKLKLYHEQDGICPYSLEPIKYDRLFEVGYVDIDHIIPYSACFDDSYNNKVLTFSSENRQKRNRVPLEYLAENKKSDYRVWVNSNIRNYRKKQNLLKEHFTEKDKEDFKQRNLNDTKYLSRVLYNYINDNLIFEDFANGRKRHVISVNGAATAHMRKCWGIDKIRENGDLHHAVDAAVISCVTQGMVNKISKYSFDRETMFDVDYRTGEVKERFPLPYPNFRKELEARSEIEDEKRLKSVLLSFPNYSYDDAESAKPAFVSRMPRRKVTGPAHKDTIRSGKVDGCKISKVELTSLKLKDGAIENYYNPESDTLLYNALLQRLIEFDGDAKQAFAEELHKPKADGTPGPIVKKVKIIERSSSSVRARNENKLGIADNGSMVRIDVFYVENDGYYFVPIYVADTIKPTLPNLACVPGGKPWKEMDDKDFVFSLYQNDLIRITAKKDMKFSVVNKDSTLPANKYENELLLYYGGADISTASINGITHDNSYKFKSCGIKSLVNIEKFTVDPLGNVNKVNKEKRMYFN\n",
      "start:  44486\n",
      "end:  49270\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/4329__CJFLGLGO_01367__Obregon-TitoAJ_2015__SM40__bin.4__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11\n",
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['44486', '44551', '44617', '44683'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "rpt1|NODE_137_length_63930_cov_6.02239|Obregon-TitoAJ_2015__SM40__bin.4|CJFLGLGO_01367\tNODE_137_length_63930_cov_6.02239\t95.45\t36\t22\t1\t0\tAATTTTTCTTGGTATGTTATAA\tAATTGTTCTTGGTATGTTATAA\t49359\t49380\t2e-05\tplus\n",
      "\n",
      "rpt3|NODE_137_length_63930_cov_6.02239|Obregon-TitoAJ_2015__SM40__bin.4|CJFLGLGO_01367\tNODE_137_length_63930_cov_6.02239\t88.24\t36\t34\t4\t0\tTTGTAGTTCCCTAATTTTTCTTGGTATGTTATAA\tTTGTAGTAACCCAATTGTTCTTGGTATGTTATAA\t49347\t49380\t4e-07\tplus\n",
      "\n",
      "and finally, just for sgb 15299, the tracrRNA ACGCACTCTTCAAGTTCGTCAAGTATTACCGG\n",
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tFLJLKKHA_00012\n",
      "SGB:\t9311\n",
      "Genome Name:\t BritoIL_2016__M1.29.ST__bin.30\n",
      "Contig:\tNODE_49_length_176785_cov_12.2725 \n",
      "-----------------------------------------------------\n",
      "\n",
      "CRISPR spacer sequences:\n",
      " ['GCGGGTTGGTCCTTCGGCGCTTCGCTCTCT', 'CGGCCTCGGCAACAGATACACCGTCAACCT', 'ACCTTGCCCTTTGTTTGTTTTGGAAATCGT', 'TCGTTCTTCAACCCCTGCAACCCATACTGA', 'CTGCCACCGCGCCGAAAACACCCGCCGCAA', 'CGGCGTAAATGCGTTTTGACGCGTATCCGT', 'CTATATCATTAAAACAAGCGACGGGATACA', 'TCCGCTTCCGCCGTTTTGGACTGATATTTA', 'TAATCCGCTTTCCCGTTTGATAGGTGATTA', 'AGGCTTCTACGGCTTTCCGCTGTTTACTAA', 'CAACATAAAATTCATTGTATTCCCCTATCT', 'AGTACAAGCCGTCATCCGACGGATGCGGGT', 'GAAGAAAACGTCCGCTTGCGTAAAGCGTTG', 'CAAATTCAGCAAGCCCTGCGACAACGCGAT', 'GTCAGTTCTGTTTCATTGTATCCGAGTTTA', 'TCGGTGAGGTTACCGTTTATTCAGCAAATT', 'ACAGGAGCTTGAGAAATGACAACGTATTTT', 'ATTCCCGCCGATATGGTTGGCAGGGACAGC', 'CGCGCGTTTTGTCTGTTCCTTATTGCGCCG', 'CAATTATTACACCGATTGGACACAACGAAC', 'GGCGTTTCCGTTCCGTGCCTATGCGTTGAC', 'CAGTTCGGCGGTATCGGCTTATTCTGATGA', 'CTTGTGGCGAATATGGAGATAAAACTAAG', 'AAGAGCGAAATTCTGGTATTTCAAACGTCG', 'ACGGTAAGGTTCGTTATTTTGCGTGTGGCGA', 'AAGAGCATTGCACTTTCGTCACGTTAACGT', 'ACGATCGCGAAGCTCATCGATATGAGCTTT', 'ACGGTAAAGTGCGTTATTTTGCTTGTGGCG', 'CGGTCTTAACAAGAATGATGCCGATGGTTC', 'AAGCGGCAGAGTATTGTGATGCGGATTTGG', 'ATATAGCCACTGGCGACGAATACCCTGCAT', 'TAAGGCGGTGCTTTATGGTCGGCGACAAAT', 'AGCCCTTGCCCTACAGGACATATTCCCAAC', 'AAACTGGATTTAGGTAGCGGAATAGAGAGA', 'ATAGCGTCATTATTGATGATTATCCGCCTT', 'ATACCTGTTGCAAGGTCTTGCTTGAGAAGC', 'ACAATCTGTTTAGCTTTCGCGTGCTTTTCT', 'CGGGCGTATCGTGTAGCGAAAGCCACACTC', 'TGAATAATTATTTCGGTAAGAAAGCTGCTAAA', 'CAACTATCGGTTTACCGTTAGAGGCTTCTT', 'CATTTCATTGCATCTTCGGCAGCTTCCGTC', 'TTGTTTGTGTTGTTTGATAGGAAAGCGGAA', 'TTTGTTTGTCGTTGCGATTCATATGAATCG', 'CCGTACTTCGCGCGGAAGGATTCAACCGCC'] \n",
      "CRISPR repeat sequences:\n",
      " ['GTTATCGCTTCCTTTCCAGCCTAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC', 'ATTATAGCAGATGTTTATTTTGTGTCAAACCGCAAC']\n",
      "Retrieving locus\n",
      "CRISPR si trova 13625 14320\n",
      "Cas9 si trova [98503, 101700]\n",
      "Cas2 si trova [102599, 102928]\n",
      "Cas1 si trova [101693, 102598]\n",
      "1065.0 979\n",
      "Cas locus on minus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? False\n",
      "RQIRFVPHSGFDITALHR*KNWFWEI*YAKLAFVKQKKRFFHENGFQSVKRAFRHIRDAG*QIAVPPRFDAGRTGKRLYENQGFAGKRRRYGDLVRHAQTGG*HRKATRRNMDGCGRRFRLARARFRFGHGQRGHGNAAVGGTGLHSPV*NALYGRRQPLFPPDEPHHASFGANGGAV*NGAGRDFAHDRHRRESPQADYLSPARRVGASQIRRFAGGTEH*RRHDRHRTRRLEKPRSHRTHAARIRRGFGYCRRFGRRAHPHRSRQNLFDGVRFDRSRRHIVRRFSDCRRFDYPRFGAFPARHRNQPAQKRHSDGFKGHGGGHYA*KQPNRRGRTRRGHSRQIVRTARR*FARFACAVDD*RIPDSGNRRRFCRRKNDLARGVGAESQRKQPIRRDFKRTESERSRLFRPWQRHRNQGDGKKTLRRGTD*GKTGSQNCHVLFGNGHGGGKPRGGGRHIVDCDQFSELYRTDEQIGSKHS*LSRLTDRRRRAKARWRKTLPNISTTLFWTRADCTAPSDTPFCRRTKIRRMKPPRPKPPKRSTRKASTPF*KTPLCATKRQATRRPKSPSFPPCARLCWTCSAISPRTRFSRTAPPPKARCWTDATSERSFAPTRT*NSL*PPAPKSAPTAASRNCRTRTNPPTTTRF*RTSKNATNAIPNARPPP*NPRKTPAFWTRRT*TPRKCSNPRWTLSRLTTNKPICPNVSPRPSIPPAPN*FWGRCRG*NRWNKRSITPTRKTRSGGLWQSCSARPARSLRMTTS*ICSCRTILPCGT*FRPAPATAAWIPISATPSRTTLNAFSNNIRQ*RQSFSTGKKRTTAL*KRSKQHRSSTR*RLFPCRPPRPPTRICRLTTS*RRGACCCKNSICKQPRAAPYLSSRSPLGLWRSICRMPIAEAESMDCVVASLRAMTILLCFIRTDKG*DFRVTTSSLLVFSIFYLFSI**KNIQNHKEQIETPTPALGQKSAYNFSVAIHVCDFYRKGFLTPCSRFREQKKLIKLPLKRKAFLFRRKPLQNSLFSCIQSAKPQGTGGFVVFMCRPSDTTEWQVKS*RKFFNGNKKYSRRN\n",
      "start:  13625\n",
      "end:  102928\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n",
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/9311__FLJLKKHA_00012__BritoIL_2016__M1.29.ST__bin.30__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11\n",
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['13625', '13691', '13757', '13823', '13889', '13955', '14021', '14087', '14153', '14219', '14285', '102979', '103045', '103111', '103177', '103243', '103309', '103375', '103441', '103507', '103573', '103639', '103705', '103771', '103836', '103902', '103969', '104035', '104101', '104167', '104233', '104299', '104365', '104431', '104497', '104563', '104629', '104695', '104761', '104827', '104895', '104961', '105027', '105093', '105159', '105225'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "rpt2|NODE_49_length_176785_cov_12.2725|BritoIL_2016__M1.29.ST__bin.30|FLJLKKHA_00012\tNODE_49_length_176785_cov_12.2725\t96.30\t36\t27\t1\t0\tTTATCGCTTCCTTTCCAATCAAAACGC\tTTATCGCTTCCTTTCCAATTAAAACGC\t18691\t18665\t8e-08\tminus\n",
      "\n",
      "rpt3|NODE_49_length_176785_cov_12.2725|BritoIL_2016__M1.29.ST__bin.30|FLJLKKHA_00012\tNODE_49_length_176785_cov_12.2725\t100.00\t36\t17\t0\t0\tTTATCGCTTCCTTTCCA\tTTATCGCTTCCTTTCCA\t18691\t18675\t6e-04\tminus\n",
      "\n",
      "and finally, just for sgb 15299, the tracrRNA CGAACAGCGGGGCTTTTGAAAATTCCGGCAAA\n",
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tLJKGNKPJ_00564\n",
      "SGB:\t14454\n",
      "Genome Name:\t KarlssonFH_2013__S420__bin.8\n",
      "Contig:\tNODE_211_length_45519_cov_7.71687 \n",
      "-----------------------------------------------------\n",
      "\n",
      "CRISPR spacer sequences:\n",
      " ['GAAGTCCTGGAGTACTTCCAAGGCAAAGAG', 'GAGGCGCAGGCCGAGATGGTCAACCCGCCC', 'CCGGGCGGGTCAAGTATGTCTCGCTGCTTG', 'TTGCCGAAGGCACTCCAATCGGTCTTCTTC', 'CGAATGGCGAAGGGGCAGATGCTTTGCGCA', 'GAATCTACTGGCAAGGTGTACACTGTCACG', 'GCTCAATCACATTACCATCGGCATCAAGCA', 'AGTACTTGCGCGAATCGAAGCACTTGTACA', 'TTCTTCTAGTGTCTCAGCTTAACAGTAACT', 'TGGCCTTGCTGGTCATATACATACAATGTA', 'AGTCTGTGTCGGATGGAATGACGAGAGGTA', 'TCAGTAAGTTGGATACCTGCAGATACCATGT', 'ATTGATAGACTCGATGTCTTCCAGATCATA'] \n",
      "CRISPR repeat sequences:\n",
      " ['GTCTTGAGCAGGTGCCCTTCCCTAAAGTGATACGCT', 'GCCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT', 'GTCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT']\n",
      "Retrieving locus\n",
      "CRISPR si trova 12991 13885\n",
      "Cas9 si trova [9720, 12728]\n",
      "Cas2 si trova [15225, 15542]\n",
      "Cas1 si trova [14325, 15215]\n",
      "1002.0 1002\n",
      "Cas locus on plus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? True\n",
      "MEITINREIGKLGLPRHLVLGMDPGIASCGFALIDTANHEILDLGVRLFDSPTHPKTGQSLAVIRRGFRSTRRNIDRTQARLKHCLQILKAYGLIPQDATKEYFHTTKGDKQPLKLRVDGLDRLLNDREWALVLYSLCKRRGYIPHGEGNQDKSSEGGKVLSALAANKEAIAETSCRTVGEWLAQQPQSRNRGGNYDKCVTHAQLIEETHILFDAQRSFGSKYASPEFETAYIEVCDWERSRKDFDRRTYDLVGHCSYFPTEKRAARCTLTSELVSAYGALGNITIIHENGTSRALSATERDECIAILFSCEPIRGNKDCAVKFGALRKALDLSSGDYFKGVPAADEKTREVYKPKGWRVLRNTLNATNPILLQRLRDDRNLADAVMEAVAYSSALPVLQEQLQGLPLSEAEIEALCRLPYSSKALNGYGNRSKKALDMLLDCLEEPEVLNLTQAENDCGLLGLRIAGAQLERSDRLMPYETWIERTGRTNNNPVVIRSMSQMRKVVNAICRKWGVPNEIHVELDRELRLPQRAKDEIAKANKKNEKNRERIAGQIAELRGCAADEVTGKQIEKYRLWEEQEYFDLYTGAKIEVDRLISDDTYTQIDHILPFSRTGENSRNNKVLVLAKSNQDKREQTPYEWMSHDGAPSWDAFERRVQENQKLSRRKKNFLLEKDLDTKEGEFLARSFTDTAYMSREVCAYLADCLLFPDDGAKAHVVPTTGRATAWLRRRWGLNFGSDGEKDRSDDRHHATDACVIAACSPSLVIKTARINQETHWSITRGMNETQRRDAIMKALESVMPWETFANEVRAAHDFVVPTRFVPRKGKGELFEQTVYRYAGVNAQGKDIARKASSDKDIVMGNAVVSADEKSVIKVSEMLCLRLWHDPEAKKGQGAWYADPVYKADIPSLKDGTYVPKIAKAATGRKAWKAIPQNVLNSAPLEIYLGDLVQIGNQKGRFRGINIATVSWSFVDTLSKLPVKMPTIGRLNNETIPTVIRELLI\n",
      "start:  9720\n",
      "end:  15542\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/14454__LJKGNKPJ_00564__KarlssonFH_2013__S420__bin.8__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11\n",
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['12991', '13057', '13123', '13189', '13255', '13321', '13387', '13453', '13519', '13585', '13651', '13717', '13784', '13850'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13125\t13158\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13191\t13224\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13257\t13290\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13323\t13356\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13389\t13422\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13455\t13488\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13521\t13554\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13587\t13620\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13653\t13686\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13719\t13752\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13786\t13819\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\t13852\t13885\t3e-12\tplus\n",
      "\n",
      "rpt1|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t92.00\t36\t25\t1\t1\tGAGCACGCACCCTTCCCCAAGGTGA\tGAGC-CGCGCCCTTCCCCAAGGTGA\t14247\t14270\t5e-05\tplus\n",
      "\n",
      "rpt2|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t97.06\t36\t34\t1\t0\tCTTGAGCACGCGCCCTTCCCCAAGGTGATACGCT\tCTTGAGCACGCACCCTTCCCCAAGGTGATACGCT\t13059\t13092\t3e-12\tplus\n",
      "\n",
      "rpt2|NODE_211_length_45519_cov_7.71687|KarlssonFH_2013__S420__bin.8|LJKGNKPJ_00564\tNODE_211_length_45519_cov_7.71687\t93.33\t36\t30\t1\t1\tGTCTTGAGCACGCGCCCTTCCCCAAGGTGA\tGTCTGGAGC-CGCGCCCTTCCCCAAGGTGA\t14242\t14270\t8e-08\tplus\n",
      "\n",
      "and finally, just for sgb 15299, the tracrRNA TTATGCCAAAACCGTTCTTTCGGGCGACACGG\n",
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tCPMHNJAN_01529\n",
      "SGB:\t8767\n",
      "Genome Name:\t LiJ_2014__O2.UC47-1__bin.10\n",
      "Contig:\tNODE_8014_length_5265_cov_2.43301 \n",
      "-----------------------------------------------------\n",
      "\n",
      "CRISPR spacer sequences:\n",
      " ['TTTAAGCACTTATCTTGGGAAATGCAAAACG', 'GGGCAAAATCGAGGAGTCTTTGCAAGAAAA', 'GCGCGTATTATTGGGGCGGATACGACGCGA', 'TGCCTTCCTGTTTTAAGTGATGTTGCTGCT', 'AGTTGACATCGTATATCCGCTGACATTTTT', 'AAGTCGAAGAGTGTCATCTTCGCTCTTTAT', 'GAAAAATTTAGGGATTACAAAGGTTGTGCG', 'AAAGATAAGCATTTCGCAGAATGTCTTTGAT', 'CAAGTGGACACCTATAATTCCATTCTTCTT'] \n",
      "CRISPR repeat sequences:\n",
      " ['GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT', 'GTTTTGACTTTACAACAAATTGTTCTTTCTAAAAAT']\n",
      "Retrieving locus\n",
      "CRISPR si trova 6 637\n",
      "Cas9 si trova [2023, 5202]\n",
      "Cas2 si trova [731, 1099]\n",
      "Cas1 si trova [1130, 2026]\n",
      "1059.0 1059\n",
      "Cas locus on minus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? True\n",
      "MEKEKNNHVWGFDIGKGSLGEAVRIGGDFKHIASVLLDPEFGEIKTAALARRQMRTRKAHKAREKWLEECLEGTGVEILKRREVGIVDGQWQLISKGDERLEREFPPSGEDVCYNSIALRCKLLLGEKLEGWQVFKALNSAIQKRGYDENIPWGEAEEKSSKKDDDDYAQKLSQYEKEKSELFESFADGEKYDFPCFFKAYKMGLWSPENPTRVEVRIDCRAQKAKSYVIARKYVEREFECLVEAAAKFFPKLKGRAKFILYGVSETPYASYYGNMRKKFGLKRGAESDWTALGQKVPRFDNRIIDKCRFIPRMNVCKIRPLNEARNEKDLLYYEVTLALKLLNLRFFRNSNIEQLTFEEFKKAFEIAAGAKYKMTKTAMKKFLKSISATALGDDYAEIEPPKESGRASFSRPAMEILKELIFSGMPPREFYGRKISEISNTDKNKGVVSSDLDFIKLMGDCPWGGIFIPDGQTYQFANAAFPDPDAQINKLIGSQNNPIVRHRLTFFFERIKFLESKYGVPDKVVLEFVRDDFMGEKEKKEMRIAMKKRAEEKMRIAKDLDEHGFKGGNLLLKMELYRKQNGICLYTGDALPLSELATLEIEHIVPRSRGGPDAQYNYVITREKTNKLKGDKTPYEWLSSDAKEWASYTDRVNACAKGLGKKRCRLLLEKDAEELVEKYTALAETAHIAKLAQKIVCMHFGFQFGGLSGTKRVFTVPGALTSRIRKFYGLDRILHNSASESENLSFEEFLKFSEDLEKKNRKNKKHHALDAMCLCFAPTGVDSRRAKLGEILPEKLRSEKAAREFFKSYLEKIMPVDVAPKKPRLEDGIYSKRIIGGKACMVKRNNLVDLAYKSGLKPVFDIPTLIKLVDKKEKGIINPQIRKMIGEFAATNPDESAWRKWCEEVRLPSKSGLGARIMRVLVYYGEADEYKDLSKDGYGAYRKGDGHKGQVIWESVDGKYYVEPVYVHASKAGVMAALNANPKKKRICGMFNSHCTVDVGDVYNDKGEFILPAGRYMVNTIRSDNQCVLTNADGEKRNPININYLMRAGMRRVDLSEL\n",
      "start:  6\n",
      "end:  5202\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n",
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/8767__CPMHNJAN_01529__LiJ_2014__O2.UC47-1__bin.10__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11\n",
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['6', '73', '139', '205', '271', '337', '403', '469', '536', '602'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "and finally, just for sgb 15299, the tracrRNA \n",
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tAKIPNPCL_00899\n",
      "SGB:\t9340\n",
      "Genome Name:\t CosteaPI_2017__SID713A009-11-90-0__bin.81\n",
      "Contig:\tNODE_1451_length_29619_cov_4.50203 \n",
      "-----------------------------------------------------\n",
      "\n",
      "CRISPR spacer sequences:\n",
      " ['GTCTATTCGGCGGCAATTACCGGCAATACG', 'GAACCGGCTTGTCGGCAGACAGGGCAAGCG', 'CGATAGCTTCGTATAAGGAGACAAACATGA', 'TTCTTCCTGCTTTGGTTAAGGTATAGTCAA', 'GCCCCGTGACGGCAAGCGCTATACTTTGAG', 'CGATAGCTTCGTATAAGGAGACAAACATGA', 'TTCTTCCTGCTTTGGTTAAGGTATAGTCAA', 'GCCCCGTGACGGCAAGCGCTATACTTTGAG', 'CGATAGCTTCGTATAAGGAGACAAACATGA', 'GACGTCGGAAACCGCTCTTGCCGCCTTTTT', 'GTACGTGAAGCGAAGCATTGGGCGAAGTCA', 'GACTTCTGGGCAAGAACTATATAATCTGCC', 'CAAATCAACAATCGAAAGGCAGAAAAAAGG', 'CCAACCGGCGTGGTCAGCATGGCAATCCCG', 'CGATCGCAAGGCTTTTGCCCGCATTGTGGT', 'GGCTGATATTGTCATGACCGGTCTCCATAA', 'GACCGTTGGCCGGCGCATCCGGGTCGGATT', 'CGGACGTTCAAGAGGAACGGCGCCGGTGGT', 'GAATTATGTCCTCTTCCTTGGTTCTACCGG', 'CTTCTGAGGATTTGATAGGCTATGCTTACC', 'CCGTCATCAACGCCATAAACGCCGCTGAAG', 'CGTTGACAATATCGGCATAAGTTGAACGGG', 'ATACACATATTTCACAAGACAAGTATATAC', 'ACCGTCAATGCTCTTAACGTGCGGGCATAC', 'AAAAGGAGCAACACAATATGAATAAAAACA', 'CAACAACAACATTTATAACTTCATGGACAA', 'ATAAGAGCGCGAGCTATATCTCACGGTACG', 'ACCTACCAGCAAGGTGTTGCAAGACAGTGG', 'CAAACCGATTGCCGAAGCGAAAAGCTTTGC', 'TAATTGAGGGAAACGATGTAAGTTATAGTT', 'TCATTTTGAAAACAAACGATATTTTTTTGAG', 'AAAGTTAGCATCCGAAAGGCTATTTGGCAC', 'GCTGAAATCAGTCAAGAGTATATATGAAAA', 'TGATGAACTGGATTCAATATTCGTGCCTGG', 'CTATCTACAATAGCGAAATCTACGCTACCG', 'AAAGCACAGTATATTACTGTGCTTTTTTTT'] \n",
      "CRISPR repeat sequences:\n",
      " ['ATTGTAACGTTCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT', 'GTTATAGCTTCCTTTCCAAATCAGATATGCTATAAT']\n",
      "Retrieving locus\n",
      "CRISPR si trova 21722 24134\n",
      "Cas9 si trova [25441, 28458]\n",
      "Cas2 si trova [24192, 24497]\n",
      "Cas1 si trova [24534, 25436]\n",
      "1005.0 1005\n",
      "Cas locus on minus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? True\n",
      "MKKKIFGFDLGIASIGWAVIDHSDENFDPETGEIIEGKVVGCGVRCFPVAENPKDGSSLAAPRREKRLLRRITRRKARRMLGIKRLFVAKGLVASTTELEALYAEQTGGDVWNLRAEALRRPLSKEELLRVLTHLAKHRGFKSYRKAAEEADKESGRILTAIAENRKETVGFQTLAQMIVERAKHSDDHKMRNYTTQKGPNKGVAVYVNSIPREEIEKETKLIFEYQKQFGLFTEDLYRDFCKIAFRYREAGSVGHMVGRCRFESEQPRAPKEAPSAELFVALSKINNLKVTVDGERRFLNGEERKALLELLKNTKEVKYSTIKNKLFKSREVFFDDVNYAQKTKKGKSGEEKAVNPEDAKFYAMKGWHKLKAAFSPEQWKEVGSNLPLLDLGMTAVVCEKNDAGIERFLSEKGIPEDYREVFKKLTGSEFINLSLKALYKLNPYLAEGLKYNQACEKAGYDFREDGIKLAEEKGLLLPPIANDKLTTVPVVNRAVAQFRKVYNAMVRTYGAPDQINLEIGRDLKKSRDERNQIMRRQKENETERKEAEDGLEKEGLAANGKNMLKYRLYRQQNGKCIYSGKAIDLRRLDENGYCDVDHIIPYSRSLDDGQNNKVLCLAEENRKKGSQTPYEYLEPLGRWEEFETVVNTTPSINRYKRNNLLNKDYKEKENDLEFRERNANDNSYIARYVKRYLEDAVDFSASSCAIKNRIQVRTGSLTDYLRYQWGLIKDRNESDRHHAQDAVVVACATQGMVQKLSKLSAIFENKDDFRRKKAEKLGHEEAEVWYKYIKQQIREPWSGFRTEVLASLEKVFVSRPPRKNATGEIHQETIRTVNPKRKKYNEKEILSGIKIRGGLANNGLMLRTDVFVKKNKKGKDEFYLVPVYLSDMGKELPNKAMVPGKKENEWLELDETYQFKFSFYMDDLIKIKKKENEIFGYFRGTNRATASVSVATHDRSHTFEGIGVKTLDGIEKYQVDPLGRIAKVKKETRLPLTMIKKNRHKKGE\n",
      "start:  21722\n",
      "end:  28458\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/9340__AKIPNPCL_00899__CosteaPI_2017__SID713A009-11-90-0__bin.81__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11\n",
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['21722', '21788', '21854', '21920', '21986', '22052', '22118', '22184', '22250', '22316', '22382', '22448', '22514', '22580', '22646', '22712', '22778', '22844', '22910', '22976', '23042', '23108', '23174', '23240', '23306', '23372', '23438', '23504', '23570', '23636', '23702', '23769', '23835', '23901', '23967', '24033', '24099'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t21799\t21823\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t21865\t21889\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t21931\t21955\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t21997\t22021\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22063\t22087\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22129\t22153\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22195\t22219\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22261\t22285\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22327\t22351\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22393\t22417\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22459\t22483\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22525\t22549\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22591\t22615\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22657\t22681\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22723\t22747\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22789\t22813\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22855\t22879\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22921\t22945\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t22987\t23011\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23053\t23077\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23119\t23143\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23185\t23209\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23251\t23275\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23317\t23341\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23383\t23407\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23449\t23473\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23515\t23539\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23581\t23605\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23647\t23671\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23713\t23737\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23780\t23804\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23846\t23870\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23912\t23936\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t23978\t24002\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t24044\t24068\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t24110\t24134\t4e-09\tplus\n",
      "\n",
      "rpt1|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t17\t0\t0\tCTTTCCAAATCAGATAT\tCTTTCCAAATCAGATAT\t29574\t29558\t1e-04\tminus\n",
      "\n",
      "rpt2|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t27\t0\t0\tTTATAGCTTCCTTTCCAAATCAGATAT\tTTATAGCTTCCTTTCCAAATCAGATAT\t29584\t29558\t3e-10\tminus\n",
      "\n",
      "rpt2|NODE_1451_length_29619_cov_4.50203|CosteaPI_2017__SID713A009-11-90-0__bin.81|AKIPNPCL_00899\tNODE_1451_length_29619_cov_4.50203\t100.00\t36\t25\t0\t0\tCTTTCCAAATCAGATATGCTATAAT\tCTTTCCAAATCAGATATGCTATAAT\t21733\t21757\t4e-09\tplus\n",
      "\n",
      "and finally, just for sgb 15299, the tracrRNA CCACAGCCCGCACGAGAGTCAAAATTATGGTT\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tAJHGDLDC_00764\n",
      "SGB:\t15286\n",
      "Genome Name:\t HMP_2012__SRS015065__bin.2\n",
      "Contig:\tNODE_3931_length_8539_cov_7.89097 \n",
      "-----------------------------------------------------\n",
      "\n",
      "CRISPR spacer sequences:\n",
      " ['AGCGTCATGTTCTCGGTAATCGGTGCATCT', 'GCGGCAAATCTGGCCGCCACACTGAAATTC', 'GTGCCGATAAACTGCCCCTTGGAGATTGTG', 'GGCTGCACGGATAGATAGGCGTTGCCGTTG', 'CCGTGTGTGTAGTAGATAACATTGTTCCAC', 'ATCTCCCATTGCATATAATAGGGGTTTGCC', 'CGTACATATCCACGCTCTGGCCGTCCTCCA', 'GCAAGCGGTTACTGGGCCAGGTGTCGGGGA', 'TACATCGACATGAACAGCATACCGATGCTG', 'ACCTCATACATTGCCCTTATGAATTCCACT'] \n",
      "CRISPR repeat sequences:\n",
      " ['ATCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT', 'GTCATAGTTCCCTAATAGCTCTTGGTATGGTATAAT']\n",
      "Retrieving locus\n",
      "CRISPR si trova 1618 2313\n",
      "Cas9 si trova [3560, 6784]\n",
      "Cas2 si trova [294, 584]\n",
      "Cas1 si trova [594, 1625]\n",
      "1074.0 1074\n",
      "Cas locus on minus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? True\n",
      "MKHPYGIGLDIGIASVGWAVVALNENAEPYGLIRCGSRIFDKAEQPKTGDSLAAPRREARSARRRLRRRSLRKADLYELMEKNGLPGKAEIEQAVQAGHLPDVYALRVQALDGPVTALDFARILLHLMQRRGFRSNRKADDAQKDGKLLQAIDANTRRMEANRYRTVGEMMYRDPVFAEHKRNKAENYLSTVKRDQIIDEVRLVFAAQRQYGATWASPEMEAEYLCILTRQRSFAEGPGKGSPYSGSNRVGTCTLEGKSEQRAAKAAFSFEYFTLLQKINHIRIAENGTSRTLTPAERQILLSVCCQTDKMDFARVRKALALPEEARFNMVRYRGEQTAEACEKKEKITALPCYHKMRKALNTLRKDHIRNISREQLDAAGAALTNPENEDKLREALKQAEFEPLEIEALLTLPSFAGYGHISVKACRKLIPYLEQGMNYNDACQAAGYDFQGRQNGEKAQFLPASTEEMDDITSPVVRRAVAQTIKVVNAIIREQGESPVSIHLELAREMNKNFQQRSELDKAMRDNSAENERLMKELNELFPGRTVTGQDLVKYRLWKEQNEICAYSLEKLDLTRVITEAGYAEVDHIVPYSISFDDRRTNKVLVLASENRQKGNRLPLQYLQGKRRDDFIVYTKANVKNFRKRQNLLKERLSEEDGKGYIQRNLQDTQYIAAFMLNYIRNHLAFADCSGAGKRRVVAVNGAVTAFLRKRWGLSKVRTDGDLHHAADAAVIACTTQGMIKRVSDFCKRAETTAVRNEHFPEPWPRFRDELTQRLSACPQEDLMKINPVYYATVDISSIQPVFVSRMPRHKVTGAAHEETIRSQVNEKYTAVRKSITELSLDKDGKIKDYFKPSSDTLLYEALKKRLTEFGGNAKKAFAEPFYKPRADGTPGAQVRKVKVVAKMNNTIPVRSGGGVAKGGDMIRIDVYYVPGDGYYWVPIYVADTVKETLPNKAVVRDKTMEEWKEMKEDDFLFSLYSNDLILVERDKPICFSLMHEDSTLPKKYETKKELVYYKGGDISNGGIRIETHEGAYFLKSLTFGIVQKVQKYQVDVLGNYTPVKKEKRQTFPAQRR\n",
      "start:  294\n",
      "end:  6784\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n",
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/15286__AJHGDLDC_00764__HMP_2012__SRS015065__bin.2__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11\n",
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['1618', '1684', '1750', '1816', '1882', '1948', '2014', '2080', '2146', '2212', '2278'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "rpt1|NODE_3931_length_8539_cov_7.89097|HMP_2012__SRS015065__bin.2|AJHGDLDC_00764\tNODE_3931_length_8539_cov_7.89097\t86.11\t36\t36\t4\t1\tATCATAG-TTCCCTAATAGCTCTTGGTATGGTATAA\tATCATAGCAACCATAATAGTTCTTGGTATGGTATAA\t6928\t6963\t7e-07\tplus\n",
      "\n",
      "rpt2|NODE_3931_length_8539_cov_7.89097|HMP_2012__SRS015065__bin.2|AJHGDLDC_00764\tNODE_3931_length_8539_cov_7.89097\t95.65\t36\t23\t1\t0\tTAATAGCTCTTGGTATGGTATAA\tTAATAGTTCTTGGTATGGTATAA\t6941\t6963\t7e-07\tplus\n",
      "\n",
      "and finally, just for sgb 15299, the tracrRNA \n"
     ]
    }
   ],
   "source": [
    "for seqid in [\"OBKKDPJE_00146\",\"LPEAHEHG_01071\",\"CJFLGLGO_01367\",\"FLJLKKHA_00012\",\"LJKGNKPJ_00564\",\"CPMHNJAN_01529\", \"AKIPNPCL_00899\", \"AJHGDLDC_00764\"]:\n",
    "    spacers=[]\n",
    "    repeats=[]\n",
    "    repeat_start_pos=[]\n",
    "\n",
    " \n",
    "    cas9_aa=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Seq\"].iloc[0]\n",
    "    dataset=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Study\"].iloc[0]\n",
    "    genomename=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Genome Name\"].iloc[0]\n",
    "    SGB=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"SGB ID\"].iloc[0]\n",
    "    contig=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Contig\"].iloc[0]\n",
    "    print(\"\\n-----------------------------------------------------\\n\"+feature+\" id:\\t\"+seqid+\"\\nSGB:\\t\"+str(SGB)+\"\\nGenome Name:\\t\",\\\n",
    "          genomename+\"\\nContig:\\t\"+cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Contig\"].iloc[0],\"\\n-----------------------------------------------------\\n\")\n",
    "\n",
    "    ######################################################################TODO THIS MUST BE INSIDE filename_discrepancies\n",
    "    # take into account filename discrepancies\n",
    "    if dataset.startswith(\"ZeeviD\"):  #TODO guarda se funziona quando finisci di girare minced_runner\n",
    "             s3_dataset=\"ZeeviD_2015\"\n",
    "             #genomename=filename_discrepancies.change_to_megahit(genomename)\n",
    "             s3_genomename=genomename\n",
    "             try:\n",
    "                 genomename=genomename.replace(\"ZeeviD_2015\", \"ZeeviD_2015_B\")\n",
    "                 prokkafile_ofsample=open(annodir+\"/justminced/ZeeviD_2015_B/\"+genomename+\".crisprcas.gff.minced\")\n",
    "                 prokkafile_ofsample.close()\n",
    "                 dataset=\"ZeeviD_2015_B\"\n",
    "             except:\n",
    "                 genomename=genomename.replace(\"ZeeviD_2015_B\", \"ZeeviD_2015_A\")\n",
    "                 dataset=\"ZeeviD_2015_A\"\n",
    "    else:\n",
    "        s3_dataset=dataset\n",
    "        s3_genomename=genomename\n",
    "        dataset=filename_discrepancies.s3(s3_dataset,r=True) # get working dataset name from s3 dataset name\n",
    "        genomename=genomename.replace(s3_dataset, dataset)\n",
    "\n",
    "    if filename_discrepancies.dataset_has_megahit(dataset,genomename):\n",
    "        genomename=filename_discrepancies.change_to_megahit(genomename)\n",
    "    \n",
    "    samplename = genomename.split(\"__\")[1]\n",
    "    ########################################################################\n",
    "    \n",
    "    # get CRISPR spacer and repeat sequence\n",
    " \n",
    "    mincedCRISPRfilename=datadir+\"1crisprsearch/out/\"+s3_dataset+\"/\"+s3_genomename+\".fa.minced.out\"\n",
    "    f=open(mincedCRISPRfilename, \"r\")\n",
    "    for line in f.readlines():\n",
    "        if line[0]== \"1\" or line[0]== \"2\" or line[0]== \"3\" or line[0]== \"4\" or line[0]== \"5\" or \\\n",
    "        line[0]== \"6\" or line[0]== \"7\" or line[0]== \"8\" or line[0]== \"9\" or line[0]== \"0\":\n",
    "            \n",
    "            repeats.append(line.split(\"\\t\")[2])\n",
    "            \n",
    "            repeat_start_pos.append(line.split(\"\\t\")[0])\n",
    "            \n",
    "            if not line.split(\"\\t\")[3]==\"\\n\":\n",
    "                spacers.append(line.split(\"\\t\")[3])\n",
    "    f.close()\n",
    "    \n",
    "    print(\"CRISPR spacer sequences:\\n\",spacers,\"\\nCRISPR repeat sequences:\\n\",repeats)\n",
    "    # get whole locus sequence todo this could become a function of its own, dentro utils, gli dai genomename della cas e/o seqid della cas (anzi famo solo seqid, che magari ci sn doppioni in genomi) e ti dà tt il locus\n",
    "\n",
    "    # print start and stop of CRISPR\n",
    "    unostranosplit=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"minced_CRISPR\"].iloc[0].split(\"\\', \\'\")\n",
    "    crispr_start, crispr_stop = int(unostranosplit[1]), int(unostranosplit[2].strip(\"\\')]\\\"\"))\n",
    "\n",
    "    # > get start and stop position of all Cass. questo devi metterlo nella tabella di cas non ci sono santi.\n",
    "\n",
    "    prokka_anno=tabellazza[tabellazza[\"Genome Name\"]==s3_genomename][\"prokka_cas\"].iloc[0]\n",
    "    # so già che all casses inside here are on the same contig which is 'contig', xke le ho estratte per essere così\n",
    "\n",
    "    prokkaslist=prokka_anno.split(\">\")\n",
    "    prokkaslist.pop(0)\n",
    "    # get contigname of current effectorcas   \n",
    "    cas_position={}\n",
    "    for casanno in prokkaslist:\n",
    "        unostranosplit=casanno.split(\"Name=\")\n",
    "        casnumber=unostranosplit[1][3] #because we know that we are loooking at 'good' loci, this is just going to be 9, 1 or 2\n",
    "        unaltrostranosplit=casanno.split(\"\\t\")\n",
    "        cas_start=unaltrostranosplit[3]\n",
    "        cas_stop=unaltrostranosplit[4]\n",
    "        cas_position[\"Cas\"+casnumber]=[int(cas_start),int(cas_stop)]\n",
    "\n",
    "    genomefilename= \"/shares/CIBIO-Storage/CM/scratch/tmp_projects/epasolli_darkmatter/allcontigs/ALLreconstructedgenomes/\"+str(SGB)+\"/\"+genomename+\".fa\"\n",
    "    print(\"Retrieving locus\")\n",
    "    \n",
    "    \n",
    "    # access contig:\n",
    "    \n",
    "    for record in SeqIO.parse(genomefilename,\"fasta\"):  #c'è modo di non fare questo 'ciclo'?\n",
    "        if record.id.startswith(contig):\n",
    "            #test se effettivamente la seq di cas è dove voglio io\n",
    "            \n",
    "            SeqIO.write(record, \"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb\", \"fasta\")\n",
    "            print(\"CRISPR si trova\", crispr_start, crispr_stop)\n",
    "            print(\"Cas9 si trova\", cas_position[\"Cas9\"])\n",
    "            print(\"Cas2 si trova\", cas_position[\"Cas2\"])\n",
    "            print(\"Cas1 si trova\", cas_position[\"Cas1\"])\n",
    "\n",
    "            #estrai la posizione di cas9 dal contig, e traducila\n",
    "            cas9start=[int(x) for x in cas_position[\"Cas9\"]][0]\n",
    "            cas9stop=[int(x) for x in cas_position[\"Cas9\"]][1]\n",
    "            cas9_nnseq=record.seq[cas9start-1:cas9stop-3] #gff should have a 1-based positional annotation (ma sto andando a occhio finchè non sono uguali, è già un oggetto Bio.Seq\n",
    "            \n",
    "            tracrRNA_of_15299=record.seq[14719:14751]\n",
    "            \n",
    "            my_translated_cas9=cas9_nnseq.transcribe().translate()\n",
    "\n",
    "            #print(my_translated_cas9)\n",
    "            #print(\"++++++++++annotatedcas9+++++++++++++\\n\",cas9_aa)\n",
    "\n",
    "            print(len(cas9_nnseq)/3,len(cas9_aa))\n",
    "\n",
    "\n",
    "            if my_translated_cas9==cas9_aa:\n",
    "                print(\"Cas locus on plus strand\")\n",
    "                plus=True\n",
    "            if not my_translated_cas9==cas9_aa:\n",
    "                #proviamo col reverse complement\n",
    "                cas9_nnseq=record.seq[cas9start+2:cas9stop]\n",
    "                my_translated_cas9=cas9_nnseq.reverse_complement().transcribe().translate()\n",
    "                print(\"Cas locus on minus strand\")\n",
    "            print(\"Is Cas9 translated from the genome equal to the orignal annotation?\", my_translated_cas9==cas9_aa)\n",
    "            print(my_translated_cas9)\n",
    "    \n",
    "            #extact a locus, 200 bp a monte e a valle dei cosi che ho:\n",
    "            starts, stops=[startstop[0]for startstop in cas_position.values()], [startstop[1]for startstop in cas_position.values() ]\n",
    "            minimo=min([crispr_start]+starts)\n",
    "            print(\"start: \", minimo)\n",
    "            massimo=max([crispr_stop]+stops)\n",
    "            print(\"end: \", massimo)\n",
    "\n",
    "            #opzione 1 blast\n",
    "          #1. salva np.uniq(repeats) in un fasta file: build temporary query file for blastn search\n",
    "\n",
    "            print(\"let's now blast the repeat against  the genome\")\n",
    "            blast_folder=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/\"\n",
    "            os.chdir(blast_folder)\n",
    "            tempfile=open(\"temp_repeat_seq\", \"w\")\n",
    "            tempfile.close()\n",
    "            tempfile=open(\"temp_repeat_seq\", \"a\")\n",
    "            for n, repeat in enumerate([Seq(sequence) for sequence in np.unique(repeats)]):\n",
    "                tempfile.write(\">rpt\"+str(n+1)+\"|\"+contig+\"|\"+genomename+\"|\"+seqid+\"\\n\"+str(repeat)+\"\\n\") \n",
    "            tempfile.close()\n",
    "            \n",
    "            #2. Blast query file against db\n",
    "            print(\"Running blastn of query file agianst all contigs\")\n",
    "            \n",
    "            #3. make db file\n",
    "            dbfile=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb\"\n",
    "            os.system(\"makeblastdb -in \"+dbfile+\" -parse_seqids  -dbtype nucl\")\n",
    "\n",
    "            \n",
    "            \n",
    "            blastoutfile=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/\"\\\n",
    "            +str(SGB)+\"__\"+seqid+\"__\"+genomename+\"__\"+feature+\".trcr.blastout\"\n",
    "            blastn_command = \"blastn -out \"+blastoutfile+\" -outfmt \\\"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\\\" -query temp_repeat_seq -db \"+dbfile+\" -evalue 0.001 -word_size 11\"\n",
    "            print(blastn_command)\n",
    "            os.system(blastn_command)\n",
    "           # os.system(\"rm /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracerRNA/temp*\")\n",
    "        \n",
    "           #4. parse blast output: only print output if it is NOT one of the old repeats.\n",
    "            print(\"\\n-----------------------------------------------------\\nParsing BLAST output, looking for hit which are NOT the same repeats\\ni.e. not starting with:\", repeat_start_pos,\":\\n-----------------------------------------------------\\n\")\n",
    "            f=open(blastoutfile,\"r\")\n",
    "            for line in f.readlines():\n",
    "                start_pos=line.split(\"\\t\")[9]\n",
    "                if start_pos not in repeat_start_pos:\n",
    "                    if str(int(start_pos)+1) not in repeat_start_pos:\n",
    "                        if str(int(start_pos)-1) not in repeat_start_pos:\n",
    "                            print(line) \n",
    "               \n",
    "            print(\"and finally, just for sgb 15299, the tracrRNA\", tracrRNA_of_15299)\n",
    "    # certo che 1) avrei potuto mettere nu indicazione di poosizione dentro il cas9, ma vbb 2) trovare il modo di capire le posizioni di cirspr dentro l'arrai a cosa si riferiscono. e qst lo posso vede ora.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "we\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " [print(\"we\") for number in [1,2,3] if number in [4,5,6,1]]\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## _S. aureus_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAAAATCTCGCCAACAAGTTGACGAGATAAACACGGCATTTTGCCTTGTTTTAGTAGATTCTGTTTCCAGAGTACTAAAAC\n"
     ]
    }
   ],
   "source": [
    "Sau_sgRNA=Seq(\"GTTTTAGTACTCTGGAAACAGAATCTACTAAAACAAGGCAAAATGCCGTGTTTATCTCGTCAACTTGTTGGCGAGATTTTT\")\n",
    "print(Sau_sgRNA.reverse_complement())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence 'NC_007795.1' (2821361 bp)\n",
      "\n",
      "\n",
      "\n",
      "CRISPR 1   Range: 860412 - 860694\n",
      "\n",
      "POSITION\tREPEAT\t\t\t\tSPACER\n",
      "\n",
      "--------\t------------------------------\t---------------------------------\n",
      "\n",
      "860412\t\tAAAGTCAGCTTACAATAGTGTGCAAGTTGG\tCGGGGCCCCAACACAGAGGCTGGCGG\t[ 30, 26 ]\n",
      "\n",
      "860468\t\tAAAGTCAGCTTACAATAGTGTGCAAGTTGG\tGGTGGAGCCCCGACACAGAGAAATTAGC\t[ 30, 28 ]\n",
      "\n",
      "860526\t\tTCCTCAATTTCTACAGACAATGCAAGTTGG\tCGGGGCCCCAACACAGAAGCTGGCGA\t[ 30, 26 ]\n",
      "\n",
      "860582\t\tAAAGTCAGCTTACAATAGTGTGCAAGTTGG\tGGTGGGACGACGAAATAAATTTTGCGAAAATATCATTTCTGTCCCACTCCCAA\t[ 30, 53 ]\n",
      "\n",
      "860665\t\tAAAGACAGCAGTAGGATAATTTCAATTTGG\t\n",
      "\n",
      "--------\t------------------------------\t---------------------------------\n",
      "\n",
      "Repeats: 5\tAverage Length: 30\t\tAverage Length: 33\n",
      "\n",
      "\n",
      "\n",
      "CRISPR 2   Range: 1776440 - 1776694\n",
      "\n",
      "POSITION\tREPEAT\t\t\t\tSPACER\n",
      "\n",
      "--------\t-----------------------\t-----------------------------------\n",
      "\n",
      "1776440\t\tAGTTGGCGGGGCCCCAACATAGA\tAGCTGGCCAATAGTTAGCTTTCAATAATGTGCAAGT\t[ 23, 36 ]\n",
      "\n",
      "1776499\t\tTGGGGTAAGGGCCCCAACACAGA\tAGCTGGCCAATAGTCAGCTTTCAATAATGTGCAAGT\t[ 23, 36 ]\n",
      "\n",
      "1776558\t\tTGGGGTAAGGGCCCCAACACAGA\tGAATTTCGAAAAGAAATTCTACAGACAATGCA\t[ 23, 32 ]\n",
      "\n",
      "1776613\t\tAGTTGGCGGGGCCCCAACACAGA\tAGCTGGCCAATAGTCAGCTTTCAATAATGTGCAAGT\t[ 23, 36 ]\n",
      "\n",
      "1776672\t\tTGGGGTAAGGGCCCCAACACAGA\t\n",
      "\n",
      "--------\t-----------------------\t-----------------------------------\n",
      "\n",
      "Repeats: 5\tAverage Length: 23\t\tAverage Length: 35\n",
      "\n",
      "\n",
      "\n",
      "CRISPR 3   Range: 2335711 - 2335904\n",
      "\n",
      "POSITION\tREPEAT\t\t\t\tSPACER\n",
      "\n",
      "--------\t-----------------------\t----------------------------------\n",
      "\n",
      "2335711\t\tTGTTGGGGCCCCGCCAACTTGCA\tCATTATTGTATGCTGACTTTTCGTCACCTTCTG\t[ 23, 33 ]\n",
      "\n",
      "2335767\t\tTGTTGGGGCCCCGTTCCCCAACT\tTGCATTGTCTGTAGAATTTCTTTTTGAAATTCTCTA\t[ 23, 36 ]\n",
      "\n",
      "2335826\t\tTGTTGGGGCCCCGCCAACTTGCA\tCATTATTGTAAGCTGACTTTCTGTCAGCTTCTG\t[ 23, 33 ]\n",
      "\n",
      "2335882\t\tTGTTGGGGCCCCGCCAACTTGTA\t\n",
      "\n",
      "--------\t-----------------------\t----------------------------------\n",
      "\n",
      "Repeats: 4\tAverage Length: 23\t\tAverage Length: 34\n",
      "\n",
      "\n",
      "\n",
      "Time to find repeats: 1209 ms\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "# unique repeats: 9 \n",
      "\n",
      "Running blastn of query file agianst all contigs\n",
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/control/Sau_ctrl.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/control/temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/control/Sau_ctrl/Sau_ctrl.fna -evalue 0.001 -word_size 11\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomename=\"Sau_ctrl\"\n",
    "spacers=[]\n",
    "repeats=[]\n",
    "\n",
    "\n",
    "# get CRISPR spacer and repeat sequence\n",
    "\n",
    "mincedCRISPRfilename=datadir+\"1crisprsearch/control/\"+genomename+\".minced.out.txt\"\n",
    "f=open(mincedCRISPRfilename, \"r\")\n",
    "for line in f.readlines():\n",
    "    print(line)\n",
    "    if line[0]== \"1\" or line[0]== \"2\" or line[0]== \"3\" or line[0]== \"4\" or line[0]== \"5\" or \\\n",
    "    line[0]== \"6\" or line[0]== \"7\" or line[0]== \"8\" or line[0]== \"9\" or line[0]== \"0\":\n",
    "        repeats.append(line.split(\"\\t\")[2])\n",
    "        if not line.split(\"\\t\")[3]==\"\\n\":\n",
    "            spacers.append(line.split(\"\\t\")[3])\n",
    "f.close()\n",
    "\n",
    "#print(\"CRISPR spacer sequences:\\n\",spacers,\"\\nCRISPR repeat sequences:\\n\",repeats,\"\\n# repeats:\", len(repeats))\n",
    "\n",
    "#\n",
    "genomefilename= \"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/control/\"+genomename+\"/\"+genomename+\".fna\"\n",
    "print(\"# unique repeats:\", len(np.unique(repeats)),\"\\n\")\n",
    "\n",
    "#opzione 1 blast\n",
    "#1. salva np.uniq(repeats) in un fasta file: build temporary query file for blastn search\n",
    "\n",
    "blast_folder=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/control/\"\n",
    "os.chdir(blast_folder)\n",
    "\n",
    "tempfile=open(\"temp_repeat_seq\", \"w\")\n",
    "tempfile.close()\n",
    "tempfile=open(\"temp_repeat_seq\", \"a\")\n",
    "for n, repeat in enumerate([Seq(sequence).reverse_complement() for sequence in np.unique(repeats)]):\n",
    "    tempfile.write(\">cr_rpt\"+str(n+1)+\"|\"+genomename+\"\\n\"+str(repeat)+\"\\n\") \n",
    "tempfile.close()\n",
    "\n",
    "#2. Blast query file against db\n",
    "print(\"Running blastn of query file agianst all contigs\")\n",
    "\n",
    "#3. make db file\n",
    "dbfile=genomefilename\n",
    "os.system(\"makeblastdb -in \"+dbfile+\" -parse_seqids  -dbtype nucl\")\n",
    "\n",
    "#4 blast\n",
    "blastoutfile=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/control/\"+genomename+\".trcr.blastout\"\n",
    "query=blast_folder+\"temp_repeat_seq\"#Sau_paper_tracrRNA\"\n",
    "blastn_command = \"blastn -out \"+blastoutfile+\" -outfmt \\\"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\\\"\\\n",
    " -query \"+query+\" -db \"\\\n",
    "+dbfile+\" -evalue 0.001 -word_size 11\"\n",
    "print(blastn_command)\n",
    "os.system(blastn_command)\n",
    "\n",
    "#os.system(\"rm /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracerRNA/temp*\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTTTAGTACTCTGTAATTTTAGGTATGAGGTAGAC\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "False\n",
      "rev compl:\n",
      " GTCTACCTCATACCTAAAATTACAGAGTACTAAAAC\n",
      "36\n"
     ]
    }
   ],
   "source": [
    "\n",
    "Sau_paper_repeat= Seq(\"GUUUUAGUACUCUGUAAUUUUAGGUAUGAGGUAGAC\", RNAAlphabet()).back_transcribe()\n",
    "print(Sau_paper_repeat)\n",
    "for repeat in repeats:\n",
    "    print(str(Sau_paper_repeat)==repeat)\n",
    "\n",
    "revc=Sau_paper_repeat.reverse_complement()\n",
    "print(\"rev compl:\\n\", revc)\n",
    "print(len(Sau_paper_repeat))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## _S. pyogenes_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence 'NC_002737.2' (1852433 bp)\n",
      "\n",
      "\n",
      "\n",
      "CRISPR 1   Range: 860819 - 861250\n",
      "\n",
      "POSITION\tREPEAT\t\t\t\tSPACER\n",
      "\n",
      "--------\t------------------------------------\t------------------------------\n",
      "\n",
      "860819\t\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tTGCGCTGGTTGATTTCTTCTTGCGCTTTTT\t[ 36, 30 ]\n",
      "\n",
      "860885\t\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tTTATATGAACATAACTCAATTTGTAAAAAA\t[ 36, 30 ]\n",
      "\n",
      "860951\t\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tAGGAATATCCGCAATAATTAATTGCGCTCT\t[ 36, 30 ]\n",
      "\n",
      "861017\t\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tAGTGCCGAGGAAAAATTAGGTGCGCTTGGC\t[ 36, 30 ]\n",
      "\n",
      "861083\t\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tTAAATTTGTTTAGCAGGTAAACCGTGCTTT\t[ 36, 30 ]\n",
      "\n",
      "861149\t\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tTTCAGCACACTGAGACTTGTTGAGTTCCAT\t[ 36, 30 ]\n",
      "\n",
      "861215\t\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCATTC\t\n",
      "\n",
      "--------\t------------------------------------\t------------------------------\n",
      "\n",
      "Repeats: 7\tAverage Length: 36\t\tAverage Length: 30\n",
      "\n",
      "\n",
      "\n",
      "# unique repeats: 2 \n",
      "\n",
      "Running blastn of query file agianst all contigs\n",
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/control/Spy_ctrl.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/control/temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/control/Spy_ctrl/Spy_ctrl.fna -evalue 0.001 -word_size 11\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sta per diventarew specifico per Spy, xke gli levo il secondo giro di CRISPRspacer.\n",
    "genomename=\"Spy_ctrl\"\n",
    "spacers=[]\n",
    "repeats=[]\n",
    "\n",
    "\n",
    "# get CRISPR spacer and repeat sequence\n",
    "\n",
    "mincedCRISPRfilename=datadir+\"1crisprsearch/control/\"+genomename+\".minced.out.txt\"\n",
    "f=open(mincedCRISPRfilename, \"r\")\n",
    "for line in f.readlines(): \n",
    "\n",
    "    #                               <-------------------- THIS LINE OF CODE IS SPECIFIC FOR S. pyogenes! occhio\n",
    "    if line.startswith(\"CRISPR 2\"):\n",
    "        break\n",
    "    #\n",
    "    \n",
    "    print(line)\n",
    "\n",
    "    if line[0]== \"1\" or line[0]== \"2\" or line[0]== \"3\" or line[0]== \"4\" or line[0]== \"5\" or \\\n",
    "    line[0]== \"6\" or line[0]== \"7\" or line[0]== \"8\" or line[0]== \"9\" or line[0]== \"0\":\n",
    "        repeats.append(line.split(\"\\t\")[2])\n",
    "        if not line.split(\"\\t\")[3]==\"\\n\":\n",
    "            spacers.append(line.split(\"\\t\")[3])\n",
    "f.close()\n",
    "\n",
    "#print(\"CRISPR spacer sequences:\\n\",spacers,\"\\nCRISPR repeat sequences:\\n\",repeats,\"\\n# repeats:\", len(repeats))\n",
    "\n",
    "#\n",
    "genomefilename= \"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/control/\"+genomename+\"/\"+genomename+\".fna\"\n",
    "print(\"# unique repeats:\", len(np.unique(repeats)),\"\\n\")\n",
    "\n",
    "#opzione 1 blast\n",
    "#1. salva np.uniq(repeats) in un fasta file: build temporary query file for blastn search\n",
    "\n",
    "blast_folder=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/control/\"\n",
    "os.chdir(blast_folder)\n",
    "\n",
    "tempfile=open(\"temp_repeat_seq\", \"w\")\n",
    "tempfile.close()\n",
    "tempfile=open(\"temp_repeat_seq\", \"a\")\n",
    "for n, repeat in enumerate([Seq(sequence) for sequence in np.unique(repeats)]):\n",
    "    tempfile.write(\">cr_rpt\"+str(n+1)+\"|\"+genomename+\"\\n\"+str(repeat)+\"\\n\") \n",
    "tempfile.close()\n",
    "\n",
    "#2. Blast query file against db\n",
    "print(\"Running blastn of query file agianst all contigs\")\n",
    "\n",
    "#3. make db file\n",
    "dbfile=genomefilename\n",
    "os.system(\"makeblastdb -in \"+dbfile+\" -parse_seqids  -dbtype nucl\")\n",
    "\n",
    "#4 blast\n",
    "blastoutfile=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/control/\"+genomename+\".trcr.blastout\"\n",
    "query=blast_folder+\"temp_repeat_seq\"#Sau_paper_tracrRNA\"\n",
    "blastn_command = \"blastn -out \"+blastoutfile+\" -outfmt \\\"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\\\"\\\n",
    " -query \"+query+\" -db \"\\\n",
    "+dbfile+\" -evalue 0.001 -word_size 11\"\n",
    "print(blastn_command)\n",
    "os.system(blastn_command)\n",
    "\n",
    "#os.system(\"rm /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracerRNA/temp*\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## blastoutput parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cr_rpt1|Spy_ctrl\tNC_002737.2\t100.00\t36\t36\t0\t0\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\t860819\t860854\t2e-13\tplus\n",
      "\n",
      "cr_rpt1|Spy_ctrl\tNC_002737.2\t100.00\t36\t36\t0\t0\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\t860885\t860920\t2e-13\tplus\n",
      "\n",
      "cr_rpt1|Spy_ctrl\tNC_002737.2\t100.00\t36\t36\t0\t0\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\t860951\t860986\t2e-13\tplus\n",
      "\n",
      "cr_rpt1|Spy_ctrl\tNC_002737.2\t100.00\t36\t36\t0\t0\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\t861017\t861052\t2e-13\tplus\n",
      "\n",
      "cr_rpt1|Spy_ctrl\tNC_002737.2\t100.00\t36\t36\t0\t0\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\t861083\t861118\t2e-13\tplus\n",
      "\n",
      "cr_rpt1|Spy_ctrl\tNC_002737.2\t100.00\t36\t36\t0\t0\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\tGTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\t861149\t861184\t2e-13\tplus\n",
      "\n",
      "cr_rpt1|Spy_ctrl\tNC_002737.2\t96.97\t36\t33\t0\t1\tGTTTTAGAGCTATGCTGTTTTGAATGGTC-CCA\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCA\t861215\t861247\t1e-09\tplus\n",
      "\n",
      "cr_rpt1|Spy_ctrl\tNC_002737.2\t96.00\t36\t25\t1\t0\tGCTATGCTGTTTTGAATGGTCCCAA\tGCTATGCTGTTTTGAATGGTTCCAA\t854433\t854457\t1e-05\tplus\n",
      "\n",
      "cr_rpt2|Spy_ctrl\tNC_002737.2\t100.00\t36\t36\t0\t0\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCATTC\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCATTC\t861215\t861250\t2e-13\tplus\n",
      "\n",
      "cr_rpt2|Spy_ctrl\tNC_002737.2\t96.97\t36\t33\t0\t1\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCA\tGTTTTAGAGCTATGCTGTTTTGAATGGTC-CCA\t860819\t860850\t1e-09\tplus\n",
      "\n",
      "cr_rpt2|Spy_ctrl\tNC_002737.2\t96.97\t36\t33\t0\t1\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCA\tGTTTTAGAGCTATGCTGTTTTGAATGGTC-CCA\t860885\t860916\t1e-09\tplus\n",
      "\n",
      "cr_rpt2|Spy_ctrl\tNC_002737.2\t96.97\t36\t33\t0\t1\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCA\tGTTTTAGAGCTATGCTGTTTTGAATGGTC-CCA\t860951\t860982\t1e-09\tplus\n",
      "\n",
      "cr_rpt2|Spy_ctrl\tNC_002737.2\t96.97\t36\t33\t0\t1\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCA\tGTTTTAGAGCTATGCTGTTTTGAATGGTC-CCA\t861017\t861048\t1e-09\tplus\n",
      "\n",
      "cr_rpt2|Spy_ctrl\tNC_002737.2\t96.97\t36\t33\t0\t1\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCA\tGTTTTAGAGCTATGCTGTTTTGAATGGTC-CCA\t861083\t861114\t1e-09\tplus\n",
      "\n",
      "cr_rpt2|Spy_ctrl\tNC_002737.2\t96.97\t36\t33\t0\t1\tGTTTTAGAGCTATGCTGTTTTGAATGGTCTCCA\tGTTTTAGAGCTATGCTGTTTTGAATGGTC-CCA\t861149\t861180\t1e-09\tplus\n",
      "\n",
      "cr_rpt2|Spy_ctrl\tNC_002737.2\t96.00\t36\t25\t0\t1\tGCTATGCTGTTTTGAATGGTCTCCA\tGCTATGCTGTTTTGAATGGT-TCCA\t854433\t854456\t3e-05\tplus\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f=open(blastoutfile)\n",
    "for line in f.readlines():\n",
    "    print(line)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "rpt1 has a hit at position 854433  854457, which is 60knt upstream of CRISPR, and just 300nt upstream of spCas9 (854751,858857), according to NCBI's gff RefSeq annotation. And Cas1 is at 858857,859726\n",
    "this is also hit by rpt2 (with one gap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGGAACCATTCAAAACAGCATAGC\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this is the sequence matching the repeat:\n",
    "spy_matched_rep=Seq(\"GCTATGCTGTTTTGAATGGTTCCAA\")\n",
    "\n",
    "# since repeat is on the + strand, and this is all on the + strand, tracrRNA should be its reverse complement:\n",
    "spy_antir=spy_matched_rep.reverse_complement()\n",
    "#this should go from 854457-854433 = 25 nt  (24+1)\n",
    "print(spy_antir)\n",
    "len(spy_antir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ARNold, do,main finder, finds something at 854300, and 854365, which is però 433"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  85446     Erpin - AAAACAAAAAACAAGGTTTACAAACCTTGTTTTTCAAGTATA NA\n",
      "\n",
      "  85448      Both + TACTTGAAAAACAAGGTTTGTAAACCTTGTTTTTTGTTTTGA NA\n",
      "\n",
      " 854300      Both - AACCAAAAAAACAAGCGCTTTCAAAACGCTTGTTTTATCATTTT -6.70\n",
      "\n",
      " 854300      Both + AAAATGATAAAACAAGCGTTTTGAAAGCGCTTGTTTTTTTGGTTTGC -10.00\n",
      "\n",
      " 854365      Both - TTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTGATAC -9.00\n",
      "\n"
     ]
    }
   ],
   "source": [
    "f=open(blast_folder+\"Spy_ctrl.terminator\")\n",
    "for line in f.readlines():\n",
    "    if line.strip().startswith(\"854\"):\n",
    "        print(line)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38\n"
     ]
    }
   ],
   "source": [
    "# this is the ARNold terminator, going from 854365 to 854402 (+1 perché conti il primo tipo), \n",
    "# but actually from 854402 - 854364 = 38 nt (37+1)\n",
    "terminator=Seq(\"TTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTGATAC\")\n",
    "print(len(terminator))\n",
    "\n",
    "#in the genome, I should find its reverse complement:\n",
    "terminator_rev_comp=terminator.reverse_complement()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1852434\n"
     ]
    }
   ],
   "source": [
    "# find this inside the genome, and link it to the thing found by ARNold:\n",
    "f=open(\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/control/Spy_ctrl/Spy_ctrl.fna\")\n",
    "for line in f.readlines():\n",
    "    if not line.strip().startswith(\">\"):\n",
    "        Spy_genome=line\n",
    "f.close()\n",
    "\n",
    "print(len(Spy_genome))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCTATGCTGTTTTGAATGGTTCCAA\n",
      "GTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\n"
     ]
    }
   ],
   "source": [
    "print(spy_antir.reverse_complement())\n",
    "rpt1=Seq(\"GTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC\")\n",
    "\n",
    "print(rpt1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGGAACCATTCAAAACAGCATAGC 25 True\n",
      "TTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTGATAC 38 True\n"
     ]
    }
   ],
   "source": [
    "#check antirepeat\n",
    "extracted_antirepeat=Seq(Spy_genome[854432:854457]).reverse_complement()  # 854432 invece che 854433 . why? Indexing diverso evidentemente col blastout, vbb\n",
    "print(extracted_antirepeat, len(extracted_antirepeat), extracted_antirepeat==spy_antir)\n",
    "\n",
    "# check terminator\n",
    "extracted_terminator=Seq(Spy_genome[854364:854402]).reverse_complement()\n",
    "print(extracted_terminator, len(extracted_terminator), extracted_terminator==terminator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGGAACCATTCAAAACAGCATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTGATAC 93\n"
     ]
    }
   ],
   "source": [
    "#per cui, final tracrRNA for spyogenes:\n",
    "spy_tracrRNA=Seq(Spy_genome[854364:854457]).reverse_complement()\n",
    "print(spy_tracrRNA, len(spy_tracrRNA))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAGTTAAAATAAGGCTAGTCCGTTATCAAC\n",
      "True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "30"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#between the ant repeat and the terminator, lies the mario sequence, which should be 93-(25+38)=30nt long\n",
    "spy_mario=Seq(Spy_genome[854364+38:854457-25]).reverse_complement()\n",
    "print(spy_mario)\n",
    "print(spy_mario==Seq(\"AAGTTAAAATAAGGCTAGTCCGTTATCAAC\"))\n",
    "len(spy_mario)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "#allignment with official sgRNA\n",
    "spCas9_sgRNA=Seq(\"GTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTT\")\n",
    "\n",
    "#which is made by:\n",
    "spCas9_tracrRNA=spCas9_sgRNA[16:]\n",
    "spCas9_tetraloop=spCas9_sgRNA[12:16]\n",
    "spCas9_repeat=spCas9_sgRNA[:12]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGGAACCATTCAAAACAGCATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGCTTTTTTTGATAC\n",
      "                     |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||||  |\n",
      "---------------------TAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGTGGCACCGAGTCGGTGC-TTTTTT--T--\n",
      "  Score=67\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# algin our tracrRNA with the official tracrRNA (the official sgRNA after the official tetraloop)\n",
    "alignments = pairwise2.align.localxx(spy_tracrRNA, spCas9_tracrRNA)\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "si ALLINEA (!= appaia, ricorda) molto bene, ma solo con gli ultimi 4 dell'anti-repeat, + cose nel mezzo (mario), + terminator.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCTATGCTGTTTTGAATG-GTTCCA-A\n",
      "        |||||  | | |   |  |\n",
      "--------GTTTT--A-GAG---C-TA\n",
      "  Score=10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Align our anti_repeat  (= revcompl of antirepeat). with sgRNA's repeat\n",
    "alignments = pairwise2.align.localxx(spy_antir.reverse_complement(), spCas9_repeat)\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "the antirepeat is different, probably because it is not matured....\n",
    "si potrebbe trovare il GTTTT signal e partire tipo da qll...\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTTTAGAGCTA\n",
      "||||||||||||\n",
      "GTTTTAGAGCTA\n",
      "  Score=12\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#align sgRNA's repeat with the first bit of genome minced extracted repeat... \n",
    "alignments = pairwise2.align.localxx(rpt1[:12], spCas9_repeat)\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OPPURE, semplicemente, come vediamo sopra, prendere i primi 12 cosi della repeat del crisprarray, e sono esattamente loro.\n",
    "Ma ora mi viene il dubbio, questi quanto bene si appaiano con l'inizo del tracrRNA ufficiale?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTTTAGA---GCTA\n",
      " ||||| |   ||||\n",
      "ATTTTA-ACTTGCTA\n",
      "  Score=10\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#align sgRNA repeat, with sgRNA antirepeat's rev compl\n",
    "alignments = pairwise2.align.localxx(spCas9_repeat, spCas9_tracrRNA[:14].reverse_complement()) # è TCGA+ un pezzo di mario!\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LA RISPOSTAE': NON BENISSIMO!\n",
    "benebene. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTGGAAC-CA--TTCAAAAC-A-G-CATAG--C\n",
      " |   |  ||  || ||||  | | | |||  |\n",
      "-T---A-GCAAGTT-AAAA-TAAGGC-TAGTCC\n",
      "  Score=17\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Align our anti_repeat. with sgRNA's antirepeat (and rest of tracrRNA)\n",
    "alignments = pairwise2.align.localxx(spy_antir, spCas9_tracrRNA[:25])\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "hmmm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TAGCAAGTTAAAAT\n",
      "||||||||||||||\n",
      "TAGCAAGTTAAAAT\n",
      "  Score=14\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#proviamo a allineare, che invece è giusto, ce l'abbiamo davanti agli occhi, TAGC + un pezzo di mario (i primi 10 nt di mario), con TAGC + mario nostro. beh viene uguale lo so già\n",
    "alignments = pairwise2.align.localxx(spy_antir[-4:]+spy_mario[:10], spCas9_tracrRNA[:14])\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35 TTGGAACCATTCAAAACAGCATAGCAAGTTAAAAT ATTTTAACTTGCTATGCTGTTTTGAATGGTTCCAA\n"
     ]
    }
   ],
   "source": [
    "#quella sequenza lì è la nostra vera antirepeat!!!!\n",
    "#ma chec cos'è? si può generalizzare??? l'inizio di mario, matcha contor qualche repeat?\n",
    "\n",
    "real_antirepeat=spy_antir[-31:]+spy_mario[:10]\n",
    "print(len(real_antirepeat),real_antirepeat, real_antirepeat.reverse_complement())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATTTTA-ACTTGCTATGCTGTTTTGAATGGTTCC-A--A-\n",
      " ||||| |   ||||||||||||||||||| ||| |  |\n",
      "GTTTTAGA---GCTATGCTGTTTTGAATGG-TCCCAAAAC\n",
      "  Score=30\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# ma qst xke in realtà, l'antirepeat è più lunga: parte di marioi è in realtà antirepeat\n",
    "alignments = pairwise2.align.localxx(real_antirepeat.reverse_complement(), rpt1)\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(rpt1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTTTAGAGCTATGCTGTTTTGAATGGTC-CCAAAAC\n",
      "        ||||||||||||||||||||  ||  ||\n",
      "--------GCTATGCTGTTTTGAATGGT-TCC--AA-\n",
      "  Score=24\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# align teh revcompl of our anti repeat, with our repeat (già presente nell'outupt di blast):\n",
    "alignments = pairwise2.align.localxx(rpt1,spy_antir.reverse_complement())\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTTTA-GA\n",
      "|||||  |.\n",
      "GTTTT-GGG\n",
      "  Score=6\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# quindi mi stai dicendo che la parte finale del repeat si appaia con la parte iniziale?, e una parte si overlappa?\n",
    "# no, da dove l'hai capito?\n",
    "alignments = pairwise2.align.localxx(rpt1[:8],rpt1[-8:].reverse_complement())\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# zanzaaaaaan. si lo fa. questo ci fa piacere xke è in linea col fatto che crispr voglia dire 'palyndromic'\n",
    "#tipo i primi 8 nucleotidi si appaiicchiano circa, e poi ce ne sn 4 che sono gli ACTG tipo."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTTTAGAGCTATG-CTGTT--TTG-AATG----GTC----C-C-AAAAC\n",
      "||||| | |    | |      ||  ||      | |    | | |||||\n",
      "GTTTT-G-G----GAC----CATT-CAA--AACAG-CATAGCTCTAAAAC\n",
      "  Score=22\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.globalxx(rpt1,rpt1.reverse_complement())\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('GTTTTAGAGCTATGCTGTTTTGAATGGTCCCAAAAC')"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpt1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
