{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Streptococcus gallolyticus\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First created Thu Jul 11 09:51:15 CEST 2019\n",
    "# Made by L-F-S\n",
    "# At the University Of Trento, Italy\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna, generic_protein, RNAAlphabet\n",
    "from Bio import pairwise2\n",
    "\n",
    "\n",
    "sys.path.insert(0, '/home/lorenzo.signorini/cas_mining/utils/')\n",
    "import filename_discrepancies\n",
    "\n",
    "feature=\"Cas9\" #WARNING!!! CHANGE THIS!!\n",
    "#outpath=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/5caslocitable/out/\"+feature+\"/\"\n",
    "datadir=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Seq ID', 'Seq Description', 'Seq', 'Contig', 'Genome Name', 'Study',\n",
       "       'Sample Name', 'SGB ID', 'Pos', 'pilercr_CRISPR', 'minced_CRISPR',\n",
       "       'prokka_cas1', 'prokka_cas2', 'uSGB', 'Level of estimated taxonomy',\n",
       "       'Estimated taxonomy', 'prokka_cas9'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Import data\n",
    "cas_dataset=pd.read_csv(\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/5caslocitable/known_\"+feature+\"prokkacas9_variants_table.csv\", index_col=0)\n",
    "cas_dataset.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## inspect casdataset, especially the last column.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Seq ID</th>\n",
       "      <th>Seq Description</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Study</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SGB ID</th>\n",
       "      <th>Pos</th>\n",
       "      <th>pilercr_CRISPR</th>\n",
       "      <th>minced_CRISPR</th>\n",
       "      <th>prokka_cas1</th>\n",
       "      <th>prokka_cas2</th>\n",
       "      <th>uSGB</th>\n",
       "      <th>Level of estimated taxonomy</th>\n",
       "      <th>Estimated taxonomy</th>\n",
       "      <th>prokka_cas9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2214</td>\n",
       "      <td>OEDMBJBD_00132</td>\n",
       "      <td>OEDMBJBD_00132 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MIKDVESDESVLLRFLSRSDMSNRNKVELKPYDKETFSDGEYLMSK...</td>\n",
       "      <td>NODE_71_length_152938_cov_11.0153</td>\n",
       "      <td>BritoIL_2016__M1.1.ST__bin.108</td>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>M1.1.ST</td>\n",
       "      <td>8018</td>\n",
       "      <td>122069 122356</td>\n",
       "      <td>[('ID=CRISPR1', '119549', '9')]</td>\n",
       "      <td>[('.', '118396', '119549')]</td>\n",
       "      <td>[('ID=OEDMBJBD_00131', '120985', '121896')]</td>\n",
       "      <td>[('ID=OEDMBJBD_00130', '120660', '120983')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=OEDMBJBD_00132', '122069', '122356')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2215</td>\n",
       "      <td>OEDMBJBD_00133</td>\n",
       "      <td>OEDMBJBD_00133 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...</td>\n",
       "      <td>NODE_71_length_152938_cov_11.0153</td>\n",
       "      <td>BritoIL_2016__M1.1.ST__bin.108</td>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>M1.1.ST</td>\n",
       "      <td>8018</td>\n",
       "      <td>122457 125459</td>\n",
       "      <td>[('ID=CRISPR1', '119549', '9')]</td>\n",
       "      <td>[('.', '118396', '119549')]</td>\n",
       "      <td>[('ID=OEDMBJBD_00131', '120985', '121896')]</td>\n",
       "      <td>[('ID=OEDMBJBD_00130', '120660', '120983')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=OEDMBJBD_00133', '122457', '125459')</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Seq ID                                    Seq Description  \\\n",
       "2214  OEDMBJBD_00132  OEDMBJBD_00132 CRISPR-associated endonuclease ...   \n",
       "2215  OEDMBJBD_00133  OEDMBJBD_00133 CRISPR-associated endonuclease ...   \n",
       "\n",
       "                                                    Seq  \\\n",
       "2214  MIKDVESDESVLLRFLSRSDMSNRNKVELKPYDKETFSDGEYLMSK...   \n",
       "2215  MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...   \n",
       "\n",
       "                                 Contig                     Genome Name  \\\n",
       "2214  NODE_71_length_152938_cov_11.0153  BritoIL_2016__M1.1.ST__bin.108   \n",
       "2215  NODE_71_length_152938_cov_11.0153  BritoIL_2016__M1.1.ST__bin.108   \n",
       "\n",
       "             Study Sample Name  SGB ID            Pos  \\\n",
       "2214  BritoIL_2016     M1.1.ST    8018  122069 122356   \n",
       "2215  BritoIL_2016     M1.1.ST    8018  122457 125459   \n",
       "\n",
       "                       pilercr_CRISPR                minced_CRISPR  \\\n",
       "2214  [('ID=CRISPR1', '119549', '9')]  [('.', '118396', '119549')]   \n",
       "2215  [('ID=CRISPR1', '119549', '9')]  [('.', '118396', '119549')]   \n",
       "\n",
       "                                      prokka_cas1  \\\n",
       "2214  [('ID=OEDMBJBD_00131', '120985', '121896')]   \n",
       "2215  [('ID=OEDMBJBD_00131', '120985', '121896')]   \n",
       "\n",
       "                                      prokka_cas2 uSGB  \\\n",
       "2214  [('ID=OEDMBJBD_00130', '120660', '120983')]   No   \n",
       "2215  [('ID=OEDMBJBD_00130', '120660', '120983')]   No   \n",
       "\n",
       "     Level of estimated taxonomy  \\\n",
       "2214                     Species   \n",
       "2215                     Species   \n",
       "\n",
       "                                     Estimated taxonomy  \\\n",
       "2214  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "2215  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "\n",
       "                                    prokka_cas9  \n",
       "2214  ('ID=OEDMBJBD_00132', '122069', '122356')  \n",
       "2215  ('ID=OEDMBJBD_00133', '122457', '125459')  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cas_dataset[cas_dataset[\"Genome Name\"]==\"BritoIL_2016__M1.1.ST__bin.108\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This guy has 2 Cas9s on the same contig. Let's see how many 8018 SGB genomes do have 2 Cas9s on the same contig"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Unnamed: 0.1', 'Genome Name', 'Study', 'Sample Name', 'Assembler',\n",
       "       'Genome Size', 'N50', '# Contigs', 'Completeness', 'Contamination',\n",
       "       'Strain Heterogeneity', 'SGB ID', 'pilercr_CRISPR', 'minced_CRISPR',\n",
       "       'prokka_CRISPR', 'prokka_cas', 'uniref_cas'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tabellazza=pd.read_csv(datadir+\"/3tabellazza/crisprcas_hits_table.csv\", index_col=0)\n",
    "tabellazza.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Seq ID</th>\n",
       "      <th>Seq Description</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Study</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SGB ID</th>\n",
       "      <th>Pos</th>\n",
       "      <th>pilercr_CRISPR</th>\n",
       "      <th>minced_CRISPR</th>\n",
       "      <th>prokka_cas1</th>\n",
       "      <th>prokka_cas2</th>\n",
       "      <th>uSGB</th>\n",
       "      <th>Level of estimated taxonomy</th>\n",
       "      <th>Estimated taxonomy</th>\n",
       "      <th>prokka_cas9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>274</td>\n",
       "      <td>KMPIJEMK_00757</td>\n",
       "      <td>KMPIJEMK_00757 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MILIKDIASGEQEIYRFLSRTMPNVNHYVELKPYDKEKFDNVQELV...</td>\n",
       "      <td>NODE_18_length_199723_cov_56.3678</td>\n",
       "      <td>BackhedF_2015__SID203_B__bin.2</td>\n",
       "      <td>BackhedF_2015</td>\n",
       "      <td>SID203_B</td>\n",
       "      <td>8018</td>\n",
       "      <td>20514 20813</td>\n",
       "      <td>[('ID=CRISPR1', '17994', '16')]</td>\n",
       "      <td>[('.', '16904', '17994')]</td>\n",
       "      <td>[('ID=KMPIJEMK_00756', '19430', '20341')]</td>\n",
       "      <td>[('ID=KMPIJEMK_00755', '19105', '19428')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=KMPIJEMK_00757', '20514', '20813')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>275</td>\n",
       "      <td>KMPIJEMK_00758</td>\n",
       "      <td>KMPIJEMK_00758 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...</td>\n",
       "      <td>NODE_18_length_199723_cov_56.3678</td>\n",
       "      <td>BackhedF_2015__SID203_B__bin.2</td>\n",
       "      <td>BackhedF_2015</td>\n",
       "      <td>SID203_B</td>\n",
       "      <td>8018</td>\n",
       "      <td>20810 21355</td>\n",
       "      <td>[('ID=CRISPR1', '17994', '16')]</td>\n",
       "      <td>[('.', '16904', '17994')]</td>\n",
       "      <td>[('ID=KMPIJEMK_00756', '19430', '20341')]</td>\n",
       "      <td>[('ID=KMPIJEMK_00755', '19105', '19428')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=KMPIJEMK_00758', '20810', '21355')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2214</td>\n",
       "      <td>OEDMBJBD_00132</td>\n",
       "      <td>OEDMBJBD_00132 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MIKDVESDESVLLRFLSRSDMSNRNKVELKPYDKETFSDGEYLMSK...</td>\n",
       "      <td>NODE_71_length_152938_cov_11.0153</td>\n",
       "      <td>BritoIL_2016__M1.1.ST__bin.108</td>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>M1.1.ST</td>\n",
       "      <td>8018</td>\n",
       "      <td>122069 122356</td>\n",
       "      <td>[('ID=CRISPR1', '119549', '9')]</td>\n",
       "      <td>[('.', '118396', '119549')]</td>\n",
       "      <td>[('ID=OEDMBJBD_00131', '120985', '121896')]</td>\n",
       "      <td>[('ID=OEDMBJBD_00130', '120660', '120983')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=OEDMBJBD_00132', '122069', '122356')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2215</td>\n",
       "      <td>OEDMBJBD_00133</td>\n",
       "      <td>OEDMBJBD_00133 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...</td>\n",
       "      <td>NODE_71_length_152938_cov_11.0153</td>\n",
       "      <td>BritoIL_2016__M1.1.ST__bin.108</td>\n",
       "      <td>BritoIL_2016</td>\n",
       "      <td>M1.1.ST</td>\n",
       "      <td>8018</td>\n",
       "      <td>122457 125459</td>\n",
       "      <td>[('ID=CRISPR1', '119549', '9')]</td>\n",
       "      <td>[('.', '118396', '119549')]</td>\n",
       "      <td>[('ID=OEDMBJBD_00131', '120985', '121896')]</td>\n",
       "      <td>[('ID=OEDMBJBD_00130', '120660', '120983')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=OEDMBJBD_00133', '122457', '125459')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3831</td>\n",
       "      <td>IOOHGADO_00343</td>\n",
       "      <td>IOOHGADO_00343 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...</td>\n",
       "      <td>NODE_94_length_65207_cov_49.5489</td>\n",
       "      <td>CM_madagascar__A13_03_1FE__bin.27</td>\n",
       "      <td>CM_madagascar</td>\n",
       "      <td>A13_03_1FE</td>\n",
       "      <td>8018</td>\n",
       "      <td>58862 62989</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[('ID=IOOHGADO_00344', '62989', '63855')]</td>\n",
       "      <td>[('ID=IOOHGADO_00345', '63852', '64196')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=IOOHGADO_00343', '58862', '62989')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3854</td>\n",
       "      <td>IKGOIMFN_00491</td>\n",
       "      <td>IKGOIMFN_00491 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...</td>\n",
       "      <td>NODE_82_length_57976_cov_50.4232</td>\n",
       "      <td>CM_madagascar__A20_03_1FE__bin.9</td>\n",
       "      <td>CM_madagascar</td>\n",
       "      <td>A20_03_1FE</td>\n",
       "      <td>8018</td>\n",
       "      <td>6554 10681</td>\n",
       "      <td>[('ID=CRISPR1', '14571', '11')]</td>\n",
       "      <td>[('.', '13349', '14571'), ('.', '21160', '2244...</td>\n",
       "      <td>[('ID=IKGOIMFN_00492', '10681', '11547'), ('ID...</td>\n",
       "      <td>[('ID=IKGOIMFN_00493', '11559', '11888'), ('ID...</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=IKGOIMFN_00491', '6554', '10681')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3855</td>\n",
       "      <td>IKGOIMFN_00499</td>\n",
       "      <td>IKGOIMFN_00499 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAK...</td>\n",
       "      <td>NODE_82_length_57976_cov_50.4232</td>\n",
       "      <td>CM_madagascar__A20_03_1FE__bin.9</td>\n",
       "      <td>CM_madagascar</td>\n",
       "      <td>A20_03_1FE</td>\n",
       "      <td>8018</td>\n",
       "      <td>15251 18640</td>\n",
       "      <td>[('ID=CRISPR1', '14571', '11')]</td>\n",
       "      <td>[('.', '13349', '14571'), ('.', '21160', '2244...</td>\n",
       "      <td>[('ID=IKGOIMFN_00492', '10681', '11547'), ('ID...</td>\n",
       "      <td>[('ID=IKGOIMFN_00493', '11559', '11888'), ('ID...</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=IKGOIMFN_00499', '15251', '18640')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3949</td>\n",
       "      <td>CMOIDPAO_00382</td>\n",
       "      <td>CMOIDPAO_00382 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...</td>\n",
       "      <td>NODE_151_length_51672_cov_99.9895</td>\n",
       "      <td>CM_madagascar__A53_01_1FE__bin.2</td>\n",
       "      <td>CM_madagascar</td>\n",
       "      <td>A53_01_1FE</td>\n",
       "      <td>8018</td>\n",
       "      <td>37456 41583</td>\n",
       "      <td>[('ID=CRISPR1', '29765', '22')]</td>\n",
       "      <td>[('.', '28343', '29765')]</td>\n",
       "      <td>[('ID=CMOIDPAO_00381', '36590', '37456')]</td>\n",
       "      <td>[('ID=CMOIDPAO_00380', '36249', '36578')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=CMOIDPAO_00382', '37456', '41583')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4002</td>\n",
       "      <td>AIJKJLCA_00348</td>\n",
       "      <td>AIJKJLCA_00348 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...</td>\n",
       "      <td>NODE_96_length_51336_cov_1088.99</td>\n",
       "      <td>CM_madagascar__A68_01_1FE__bin.24</td>\n",
       "      <td>CM_madagascar</td>\n",
       "      <td>A68_01_1FE</td>\n",
       "      <td>8018</td>\n",
       "      <td>37067 41194</td>\n",
       "      <td>[('ID=CRISPR1', '29368', '16')]</td>\n",
       "      <td>[('.', '28343', '29368')]</td>\n",
       "      <td>[('ID=AIJKJLCA_00347', '36201', '37067')]</td>\n",
       "      <td>[('ID=AIJKJLCA_00346', '35860', '36189')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=AIJKJLCA_00348', '37067', '41194')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4006</td>\n",
       "      <td>EJAINPIA_00527</td>\n",
       "      <td>EJAINPIA_00527 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...</td>\n",
       "      <td>NODE_66_length_67225_cov_92.2439</td>\n",
       "      <td>CM_madagascar__A69_02_1FE__bin.22</td>\n",
       "      <td>CM_madagascar</td>\n",
       "      <td>A69_02_1FE</td>\n",
       "      <td>8018</td>\n",
       "      <td>59139 63266</td>\n",
       "      <td>[('ID=CRISPR1', '66103', '3')]</td>\n",
       "      <td>[('.', '65934', '66564')]</td>\n",
       "      <td>[('ID=EJAINPIA_00528', '63266', '64132')]</td>\n",
       "      <td>[('ID=EJAINPIA_00529', '64129', '64473')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=EJAINPIA_00527', '59139', '63266')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4464</td>\n",
       "      <td>KKNCMIMF_00309</td>\n",
       "      <td>KKNCMIMF_00309 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNTE...</td>\n",
       "      <td>NODE_14_length_177436_cov_11.6738</td>\n",
       "      <td>ChengpingW_2017__AS10raw__bin.39</td>\n",
       "      <td>ChengpingW_2017</td>\n",
       "      <td>AS10raw</td>\n",
       "      <td>8018</td>\n",
       "      <td>64299 67691</td>\n",
       "      <td>[('ID=CRISPR1', '70444', '4')]</td>\n",
       "      <td>[('.', '70211', '71830')]</td>\n",
       "      <td>[('ID=KKNCMIMF_00310', '67864', '68775')]</td>\n",
       "      <td>[('ID=KKNCMIMF_00311', '68777', '69100')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=KKNCMIMF_00309', '64299', '67691')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4614</td>\n",
       "      <td>BAOICCCB_01645</td>\n",
       "      <td>BAOICCCB_01645 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_3808_length_4248_cov_2.6041</td>\n",
       "      <td>ChengpingW_2017__AS37raw__bin.3</td>\n",
       "      <td>ChengpingW_2017</td>\n",
       "      <td>AS37raw</td>\n",
       "      <td>8018</td>\n",
       "      <td>114 4235</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=BAOICCCB_01645', '114', '4235')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6665</td>\n",
       "      <td>LFBAJCND_01267</td>\n",
       "      <td>LFBAJCND_01267 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_35_length_40069_cov_14.9237</td>\n",
       "      <td>DavidLA_2015__LD-Run1-10__bin.2</td>\n",
       "      <td>DavidLA_2015</td>\n",
       "      <td>LD-Run1-10</td>\n",
       "      <td>8018</td>\n",
       "      <td>33228 37367</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[('ID=LFBAJCND_01268', '37367', '38233')]</td>\n",
       "      <td>[('ID=LFBAJCND_01269', '38245', '38574')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=LFBAJCND_01267', '33228', '37367')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7622</td>\n",
       "      <td>OMDKNLNP_01483</td>\n",
       "      <td>OMDKNLNP_01483 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_92_length_29327_cov_8.9586</td>\n",
       "      <td>FerrettiP_2018__CA_C10023IS2397FE_t3M15__bin.1</td>\n",
       "      <td>FerrettiP_2018</td>\n",
       "      <td>CA_C10023IS2397FE_t3M15</td>\n",
       "      <td>8018</td>\n",
       "      <td>10441 14574</td>\n",
       "      <td>[('ID=CRISPR1', '18329', '13')]</td>\n",
       "      <td>[('.', '17237', '18329'), ('.', '18971', '2032...</td>\n",
       "      <td>[('ID=OMDKNLNP_01484', '14574', '15440')]</td>\n",
       "      <td>[('ID=OMDKNLNP_01485', '15452', '15781')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=OMDKNLNP_01483', '10441', '14574')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7637</td>\n",
       "      <td>CDJLDPBF_00866</td>\n",
       "      <td>CDJLDPBF_00866 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_46_length_45597_cov_7.57947</td>\n",
       "      <td>FerrettiP_2018__CA_C10026IS2444FE_t1M15__bin.4</td>\n",
       "      <td>FerrettiP_2018</td>\n",
       "      <td>CA_C10026IS2444FE_t1M15</td>\n",
       "      <td>8018</td>\n",
       "      <td>10441 14574</td>\n",
       "      <td>[('ID=CRISPR1', '17999', '8')]</td>\n",
       "      <td>[('.', '17237', '17999'), ('.', '18641', '1986...</td>\n",
       "      <td>[('ID=CDJLDPBF_00867', '14574', '15440')]</td>\n",
       "      <td>[('ID=CDJLDPBF_00868', '15452', '15781')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=CDJLDPBF_00866', '10441', '14574')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7638</td>\n",
       "      <td>OCFCKFDJ_01244</td>\n",
       "      <td>OCFCKFDJ_01244 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_44_length_51823_cov_17.3485</td>\n",
       "      <td>FerrettiP_2018__CA_C10026IS2449FE_t2M15__bin.2</td>\n",
       "      <td>FerrettiP_2018</td>\n",
       "      <td>CA_C10026IS2449FE_t2M15</td>\n",
       "      <td>8018</td>\n",
       "      <td>10433 14566</td>\n",
       "      <td>[('ID=CRISPR1', '17991', '8')]</td>\n",
       "      <td>[('.', '17229', '17991'), ('.', '18633', '1985...</td>\n",
       "      <td>[('ID=OCFCKFDJ_01245', '14566', '15432')]</td>\n",
       "      <td>[('ID=OCFCKFDJ_01246', '15444', '15773')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=OCFCKFDJ_01244', '10433', '14566')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7704</td>\n",
       "      <td>CMOGHOMO_00555</td>\n",
       "      <td>CMOGHOMO_00555 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_32_length_111102_cov_34.8219</td>\n",
       "      <td>FerrettiP_2018__CA_C10042IS2731FE_t2M16__bin.6</td>\n",
       "      <td>FerrettiP_2018</td>\n",
       "      <td>CA_C10042IS2731FE_t2M16</td>\n",
       "      <td>8018</td>\n",
       "      <td>10450 14571</td>\n",
       "      <td>[('ID=CRISPR1', '18857', '15')]</td>\n",
       "      <td>[('.', '17237', '18857'), ('.', '19499', '2164...</td>\n",
       "      <td>[('ID=CMOGHOMO_00556', '14571', '15437')]</td>\n",
       "      <td>[('ID=CMOGHOMO_00557', '15449', '15778')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=CMOGHOMO_00555', '10450', '14571')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10452</td>\n",
       "      <td>BAEJBBCH_00525</td>\n",
       "      <td>BAEJBBCH_00525 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...</td>\n",
       "      <td>NODE_5_length_339048_cov_101.011</td>\n",
       "      <td>HeQ_2017__RSZAXPI003108-148__bin.2</td>\n",
       "      <td>HeQ_2017</td>\n",
       "      <td>RSZAXPI003108-148</td>\n",
       "      <td>8018</td>\n",
       "      <td>158056 161448</td>\n",
       "      <td>[('ID=CRISPR1', '153165', '9')]</td>\n",
       "      <td>[('.', '152537', '155536')]</td>\n",
       "      <td>[('ID=BAEJBBCH_00524', '156972', '157883')]</td>\n",
       "      <td>[('ID=BAEJBBCH_00523', '156647', '156970')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=BAEJBBCH_00525', '158056', '161448')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10529</td>\n",
       "      <td>PHEOENKP_01065</td>\n",
       "      <td>PHEOENKP_01065 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MILIKDTASGEQEIYRFLSRTMPNVNHYAELKPYDKEKFDGGQELM...</td>\n",
       "      <td>NODE_8163_length_2246_cov_1.94432</td>\n",
       "      <td>HeQ_2017__SZAXPI029486-84__bin.3</td>\n",
       "      <td>HeQ_2017</td>\n",
       "      <td>SZAXPI029486-84</td>\n",
       "      <td>8018</td>\n",
       "      <td>1879 2178</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[('ID=PHEOENKP_01064', '795', '1706')]</td>\n",
       "      <td>[('ID=PHEOENKP_01063', '470', '778')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=PHEOENKP_01065', '1879', '2178')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10585</td>\n",
       "      <td>HLPGCKEJ_00912</td>\n",
       "      <td>HLPGCKEJ_00912 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_267_length_56193_cov_4.66183</td>\n",
       "      <td>HeQ_2017__SZAXPI029502-105__bin.12</td>\n",
       "      <td>HeQ_2017</td>\n",
       "      <td>SZAXPI029502-105</td>\n",
       "      <td>8018</td>\n",
       "      <td>43363 47496</td>\n",
       "      <td>[('ID=CRISPR1', '38966', '31')]</td>\n",
       "      <td>[('.', '36950', '38966'), ('.', '39608', '4070...</td>\n",
       "      <td>[('ID=HLPGCKEJ_00911', '42497', '43363')]</td>\n",
       "      <td>[('ID=HLPGCKEJ_00910', '42156', '42485')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=HLPGCKEJ_00912', '43363', '47496')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>10630</td>\n",
       "      <td>GFDNGKMC_00295</td>\n",
       "      <td>GFDNGKMC_00295 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_182_length_91995_cov_4.66868</td>\n",
       "      <td>HeQ_2017__SZAXPI029553-32__bin.14</td>\n",
       "      <td>HeQ_2017</td>\n",
       "      <td>SZAXPI029553-32</td>\n",
       "      <td>8018</td>\n",
       "      <td>10446 14567</td>\n",
       "      <td>[('ID=CRISPR1', '19183', '30')]</td>\n",
       "      <td>[('.', '17233', '19183'), ('.', '19825', '2137...</td>\n",
       "      <td>[('ID=GFDNGKMC_00296', '14567', '15433')]</td>\n",
       "      <td>[('ID=GFDNGKMC_00297', '15445', '15774')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=GFDNGKMC_00295', '10446', '14567')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>19847</td>\n",
       "      <td>NKOAIHNC_01426</td>\n",
       "      <td>NKOAIHNC_01426 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_300_length_28736_cov_3.78449</td>\n",
       "      <td>QinJ_2012__DOF010__bin.10</td>\n",
       "      <td>QinJ_2012</td>\n",
       "      <td>DOF010</td>\n",
       "      <td>8018</td>\n",
       "      <td>10450 14571</td>\n",
       "      <td>[('ID=CRISPR1', '18859', '25')]</td>\n",
       "      <td>[('.', '17237', '18859'), ('.', '19501', '1993...</td>\n",
       "      <td>[('ID=NKOAIHNC_01427', '14571', '15437')]</td>\n",
       "      <td>[('ID=NKOAIHNC_01428', '15449', '15778')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=NKOAIHNC_01426', '10450', '14571')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21404</td>\n",
       "      <td>OMJCMEOF_01997</td>\n",
       "      <td>OMJCMEOF_01997 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNTE...</td>\n",
       "      <td>NODE_353_length_20445_cov_5.46126</td>\n",
       "      <td>QinN_2014__LD-62__bin.17</td>\n",
       "      <td>QinN_2014</td>\n",
       "      <td>LD-62</td>\n",
       "      <td>8018</td>\n",
       "      <td>2620 6012</td>\n",
       "      <td>[('ID=CRISPR1', '9821', '20')]</td>\n",
       "      <td>[('.', '8532', '10941')]</td>\n",
       "      <td>[('ID=OMJCMEOF_01998', '6185', '7096')]</td>\n",
       "      <td>[('ID=OMJCMEOF_01999', '7098', '7421')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=OMJCMEOF_01997', '2620', '6012')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21440</td>\n",
       "      <td>ANGLCDIO_00392</td>\n",
       "      <td>ANGLCDIO_00392 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_135_length_106229_cov_4.73162</td>\n",
       "      <td>QinN_2014__LD-69__bin.10</td>\n",
       "      <td>QinN_2014</td>\n",
       "      <td>LD-69</td>\n",
       "      <td>8018</td>\n",
       "      <td>101622 105743</td>\n",
       "      <td>[('ID=CRISPR1', '96496', '8')]</td>\n",
       "      <td>[('.', '95800', '96496'), ('.', '97138', '9895...</td>\n",
       "      <td>[('ID=ANGLCDIO_00391', '100756', '101622')]</td>\n",
       "      <td>[('ID=ANGLCDIO_00390', '100415', '100744')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=ANGLCDIO_00392', '101622', '105743')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24219</td>\n",
       "      <td>MLDCFJFN_01355</td>\n",
       "      <td>MLDCFJFN_01355 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_32_length_67127_cov_10.9528</td>\n",
       "      <td>VogtmannE_2016__MMRS51728985ST-27-0-0__bin.2</td>\n",
       "      <td>VogtmannE_2016</td>\n",
       "      <td>MMRS51728985ST-27-0-0</td>\n",
       "      <td>8018</td>\n",
       "      <td>62899 67032</td>\n",
       "      <td>[('ID=CRISPR1', '58371', '15')]</td>\n",
       "      <td>[('.', '56818', '58371'), ('.', '59013', '6023...</td>\n",
       "      <td>[('ID=MLDCFJFN_01354', '62033', '62899')]</td>\n",
       "      <td>[('ID=MLDCFJFN_01353', '61692', '62021')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=MLDCFJFN_01355', '62899', '67032')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>26561</td>\n",
       "      <td>MPHFNKCK_01164</td>\n",
       "      <td>MPHFNKCK_01164 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...</td>\n",
       "      <td>NODE_613_length_43171_cov_3.94376</td>\n",
       "      <td>XieH_2016__YSZC12003_36697__bin.20</td>\n",
       "      <td>XieH_2016</td>\n",
       "      <td>YSZC12003_36697</td>\n",
       "      <td>8018</td>\n",
       "      <td>5051 8443</td>\n",
       "      <td>[('ID=CRISPR1', '11724', '12')]</td>\n",
       "      <td>[('.', '10963', '12845')]</td>\n",
       "      <td>[('ID=MPHFNKCK_01165', '8616', '9527')]</td>\n",
       "      <td>[('ID=MPHFNKCK_01166', '9529', '9852')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=MPHFNKCK_01164', '5051', '8443')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27219</td>\n",
       "      <td>EEPDNCNB_00596</td>\n",
       "      <td>EEPDNCNB_00596 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNTE...</td>\n",
       "      <td>NODE_1_length_811617_cov_14.3029</td>\n",
       "      <td>XieH_2016__YSZC12003_37276__bin.66</td>\n",
       "      <td>XieH_2016</td>\n",
       "      <td>YSZC12003_37276</td>\n",
       "      <td>8018</td>\n",
       "      <td>600351 603743</td>\n",
       "      <td>[('ID=CRISPR1', '597831', '30')]</td>\n",
       "      <td>[('.', '594961', '597831')]</td>\n",
       "      <td>[('ID=EEPDNCNB_00595', '599267', '600178')]</td>\n",
       "      <td>[('ID=EEPDNCNB_00594', '598942', '599265')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=EEPDNCNB_00596', '600351', '603743')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>27993</td>\n",
       "      <td>PHIIAFGJ_00210</td>\n",
       "      <td>PHIIAFGJ_00210 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_247_length_84988_cov_4.56296</td>\n",
       "      <td>YuJ_2015__SZAXPI003434-7__bin.5</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>SZAXPI003434-7</td>\n",
       "      <td>8018</td>\n",
       "      <td>4328 8449</td>\n",
       "      <td>[('ID=CRISPR1', '11942', '13')]</td>\n",
       "      <td>[('.', '11115', '12602'), ('.', '13244', '1374...</td>\n",
       "      <td>[('ID=PHIIAFGJ_00211', '8449', '9315')]</td>\n",
       "      <td>[('ID=PHIIAFGJ_00212', '9327', '9656')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=PHIIAFGJ_00210', '4328', '8449')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>28084</td>\n",
       "      <td>DAJGLKOL_01122</td>\n",
       "      <td>DAJGLKOL_01122 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MPNQFTEEQISEIIKVRKSQSTAFNKGWHSFSAKLMNELIPELYAT...</td>\n",
       "      <td>NODE_167_length_76680_cov_8.70181</td>\n",
       "      <td>YuJ_2015__SZAXPI015219-129__bin.21</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>SZAXPI015219-129</td>\n",
       "      <td>8018</td>\n",
       "      <td>74346 76553</td>\n",
       "      <td>[('ID=CRISPR1', '71826', '12')]</td>\n",
       "      <td>[('.', '71002', '71826')]</td>\n",
       "      <td>[('ID=DAJGLKOL_01121', '73262', '74173')]</td>\n",
       "      <td>[('ID=DAJGLKOL_01120', '72937', '73260')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=DAJGLKOL_01122', '74346', '76553')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>28312</td>\n",
       "      <td>GEFHLDIP_00071</td>\n",
       "      <td>GEFHLDIP_00071 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNTE...</td>\n",
       "      <td>NODE_80_length_121637_cov_4.97806</td>\n",
       "      <td>YuJ_2015__SZAXPI015257-56__bin.40</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>SZAXPI015257-56</td>\n",
       "      <td>8018</td>\n",
       "      <td>74259 77651</td>\n",
       "      <td>[('ID=CRISPR1', '80535', '6')]</td>\n",
       "      <td>[('.', '80171', '81987')]</td>\n",
       "      <td>[('ID=GEFHLDIP_00072', '77824', '78735')]</td>\n",
       "      <td>[('ID=GEFHLDIP_00073', '78737', '79060')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=GEFHLDIP_00071', '74259', '77651')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>28442</td>\n",
       "      <td>IMAHBFIH_00347</td>\n",
       "      <td>IMAHBFIH_00347 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_24_length_176822_cov_5.8476</td>\n",
       "      <td>YuJ_2015__SZAXPI017451-16__bin.7</td>\n",
       "      <td>YuJ_2015</td>\n",
       "      <td>SZAXPI017451-16</td>\n",
       "      <td>8018</td>\n",
       "      <td>162252 166373</td>\n",
       "      <td>[('ID=CRISPR1', '158248', '7')]</td>\n",
       "      <td>[('.', '157684', '158248'), ('.', '158890', '1...</td>\n",
       "      <td>[('ID=IMAHBFIH_00346', '161386', '162252')]</td>\n",
       "      <td>[('ID=IMAHBFIH_00345', '161045', '161374')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=IMAHBFIH_00347', '162252', '166373')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>30584</td>\n",
       "      <td>KIPICGBL_01758</td>\n",
       "      <td>KIPICGBL_01758 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MKKSYSIGLDIGTNSVGFSAIDDNYKVITKKMRVMGNTSCQYANKS...</td>\n",
       "      <td>NODE_3376_length_8018_cov_10.0000_ID_68660</td>\n",
       "      <td>ZeeviD_2015__PNP_Main_40__bin.41</td>\n",
       "      <td>ZeeviD_2015</td>\n",
       "      <td>PNP_Main_40</td>\n",
       "      <td>8018</td>\n",
       "      <td>36 4100</td>\n",
       "      <td>[('ID=CRISPR1', '.', '7')]</td>\n",
       "      <td>[('.', '6770', '7203'), ('.', '7846', '8015')]</td>\n",
       "      <td>[('ID=KIPICGBL_01759', '4100', '4966')]</td>\n",
       "      <td>[('ID=KIPICGBL_01760', '4963', '5307')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=KIPICGBL_01758', '36', '4100')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>32475</td>\n",
       "      <td>GMAILNMJ_00134</td>\n",
       "      <td>GMAILNMJ_00134 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MDKAFLDDKVNEEVIEDIIKTLTLFEDKDMIHERLQKYSDIFTADQ...</td>\n",
       "      <td>NODE_3204_length_6699_cov_5.0000_ID_15084</td>\n",
       "      <td>ZeeviD_2015__PNP_Main_99__bin.11</td>\n",
       "      <td>ZeeviD_2015</td>\n",
       "      <td>PNP_Main_99</td>\n",
       "      <td>8018</td>\n",
       "      <td>4350 6671</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[('ID=GMAILNMJ_00133', '3484', '4350')]</td>\n",
       "      <td>[('ID=GMAILNMJ_00132', '3143', '3487')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=GMAILNMJ_00134', '4350', '6671')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33316</td>\n",
       "      <td>DEJDICEE_00395</td>\n",
       "      <td>DEJDICEE_00395 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...</td>\n",
       "      <td>NODE_50_length_91725_cov_10.2022</td>\n",
       "      <td>ZellerG_2014__CCIS59903910ST-4-0__bin.16</td>\n",
       "      <td>ZellerG_2014</td>\n",
       "      <td>CCIS59903910ST-4-0</td>\n",
       "      <td>8018</td>\n",
       "      <td>10450 14571</td>\n",
       "      <td>[('ID=CRISPR1', '19319', '28')]</td>\n",
       "      <td>[('.', '17237', '19319'), ('.', '19961', '2171...</td>\n",
       "      <td>[('ID=DEJDICEE_00396', '14571', '15437')]</td>\n",
       "      <td>[('ID=DEJDICEE_00397', '15449', '15778')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=DEJDICEE_00395', '10450', '14571')</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>33593</td>\n",
       "      <td>PMBNEDJK_01979</td>\n",
       "      <td>PMBNEDJK_01979 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MIKRYAEHHEDLEKLKEFIKANKSELYHDIFKDETKNGYAGYIENG...</td>\n",
       "      <td>NODE_1431_length_7325_cov_6.40248</td>\n",
       "      <td>ZellerG_2014__CCIS93040568ST-20-0__bin.13</td>\n",
       "      <td>ZellerG_2014</td>\n",
       "      <td>CCIS93040568ST-20-0</td>\n",
       "      <td>8018</td>\n",
       "      <td>4127 7297</td>\n",
       "      <td>[('ID=CRISPR2', '1464', '16')]</td>\n",
       "      <td>[('.', '438', '1464')]</td>\n",
       "      <td>[('ID=PMBNEDJK_01978', '3261', '4127')]</td>\n",
       "      <td>[('ID=PMBNEDJK_01977', '2920', '3249')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "      <td>('ID=PMBNEDJK_01979', '4127', '7297')</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Seq ID                                    Seq Description  \\\n",
       "274    KMPIJEMK_00757  KMPIJEMK_00757 CRISPR-associated endonuclease ...   \n",
       "275    KMPIJEMK_00758  KMPIJEMK_00758 CRISPR-associated endonuclease ...   \n",
       "2214   OEDMBJBD_00132  OEDMBJBD_00132 CRISPR-associated endonuclease ...   \n",
       "2215   OEDMBJBD_00133  OEDMBJBD_00133 CRISPR-associated endonuclease ...   \n",
       "3831   IOOHGADO_00343  IOOHGADO_00343 CRISPR-associated endonuclease ...   \n",
       "3854   IKGOIMFN_00491  IKGOIMFN_00491 CRISPR-associated endonuclease ...   \n",
       "3855   IKGOIMFN_00499  IKGOIMFN_00499 CRISPR-associated endonuclease ...   \n",
       "3949   CMOIDPAO_00382  CMOIDPAO_00382 CRISPR-associated endonuclease ...   \n",
       "4002   AIJKJLCA_00348  AIJKJLCA_00348 CRISPR-associated endonuclease ...   \n",
       "4006   EJAINPIA_00527  EJAINPIA_00527 CRISPR-associated endonuclease ...   \n",
       "4464   KKNCMIMF_00309  KKNCMIMF_00309 CRISPR-associated endonuclease ...   \n",
       "4614   BAOICCCB_01645  BAOICCCB_01645 CRISPR-associated endonuclease ...   \n",
       "6665   LFBAJCND_01267  LFBAJCND_01267 CRISPR-associated endonuclease ...   \n",
       "7622   OMDKNLNP_01483  OMDKNLNP_01483 CRISPR-associated endonuclease ...   \n",
       "7637   CDJLDPBF_00866  CDJLDPBF_00866 CRISPR-associated endonuclease ...   \n",
       "7638   OCFCKFDJ_01244  OCFCKFDJ_01244 CRISPR-associated endonuclease ...   \n",
       "7704   CMOGHOMO_00555  CMOGHOMO_00555 CRISPR-associated endonuclease ...   \n",
       "10452  BAEJBBCH_00525  BAEJBBCH_00525 CRISPR-associated endonuclease ...   \n",
       "10529  PHEOENKP_01065  PHEOENKP_01065 CRISPR-associated endonuclease ...   \n",
       "10585  HLPGCKEJ_00912  HLPGCKEJ_00912 CRISPR-associated endonuclease ...   \n",
       "10630  GFDNGKMC_00295  GFDNGKMC_00295 CRISPR-associated endonuclease ...   \n",
       "19847  NKOAIHNC_01426  NKOAIHNC_01426 CRISPR-associated endonuclease ...   \n",
       "21404  OMJCMEOF_01997  OMJCMEOF_01997 CRISPR-associated endonuclease ...   \n",
       "21440  ANGLCDIO_00392  ANGLCDIO_00392 CRISPR-associated endonuclease ...   \n",
       "24219  MLDCFJFN_01355  MLDCFJFN_01355 CRISPR-associated endonuclease ...   \n",
       "26561  MPHFNKCK_01164  MPHFNKCK_01164 CRISPR-associated endonuclease ...   \n",
       "27219  EEPDNCNB_00596  EEPDNCNB_00596 CRISPR-associated endonuclease ...   \n",
       "27993  PHIIAFGJ_00210  PHIIAFGJ_00210 CRISPR-associated endonuclease ...   \n",
       "28084  DAJGLKOL_01122  DAJGLKOL_01122 CRISPR-associated endonuclease ...   \n",
       "28312  GEFHLDIP_00071  GEFHLDIP_00071 CRISPR-associated endonuclease ...   \n",
       "28442  IMAHBFIH_00347  IMAHBFIH_00347 CRISPR-associated endonuclease ...   \n",
       "30584  KIPICGBL_01758  KIPICGBL_01758 CRISPR-associated endonuclease ...   \n",
       "32475  GMAILNMJ_00134  GMAILNMJ_00134 CRISPR-associated endonuclease ...   \n",
       "33316  DEJDICEE_00395  DEJDICEE_00395 CRISPR-associated endonuclease ...   \n",
       "33593  PMBNEDJK_01979  PMBNEDJK_01979 CRISPR-associated endonuclease ...   \n",
       "\n",
       "                                                     Seq  \\\n",
       "274    MILIKDIASGEQEIYRFLSRTMPNVNHYVELKPYDKEKFDNVQELV...   \n",
       "275    MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...   \n",
       "2214   MIKDVESDESVLLRFLSRSDMSNRNKVELKPYDKETFSDGEYLMSK...   \n",
       "2215   MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...   \n",
       "3831   MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...   \n",
       "3854   MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...   \n",
       "3855   MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAK...   \n",
       "3949   MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...   \n",
       "4002   MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...   \n",
       "4006   MEKTYSIGLDIGTNSVGWSVITDDYKVPAKKMKVLGNTDKKYIKKN...   \n",
       "4464   MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNTE...   \n",
       "4614   MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "6665   MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "7622   MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "7637   MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "7638   MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "7704   MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "10452  MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...   \n",
       "10529  MILIKDTASGEQEIYRFLSRTMPNVNHYAELKPYDKEKFDGGQELM...   \n",
       "10585  MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "10630  MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "19847  MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "21404  MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNTE...   \n",
       "21440  MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "24219  MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "26561  MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...   \n",
       "27219  MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNTE...   \n",
       "27993  MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "28084  MPNQFTEEQISEIIKVRKSQSTAFNKGWHSFSAKLMNELIPELYAT...   \n",
       "28312  MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNTE...   \n",
       "28442  MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "30584  MKKSYSIGLDIGTNSVGFSAIDDNYKVITKKMRVMGNTSCQYANKS...   \n",
       "32475  MDKAFLDDKVNEEVIEDIIKTLTLFEDKDMIHERLQKYSDIFTADQ...   \n",
       "33316  MTKKNYSIGLDIGTNSVGWAVITDDYKVPAKKMKVLGNTDKKYIKK...   \n",
       "33593  MIKRYAEHHEDLEKLKEFIKANKSELYHDIFKDETKNGYAGYIENG...   \n",
       "\n",
       "                                           Contig  \\\n",
       "274             NODE_18_length_199723_cov_56.3678   \n",
       "275             NODE_18_length_199723_cov_56.3678   \n",
       "2214            NODE_71_length_152938_cov_11.0153   \n",
       "2215            NODE_71_length_152938_cov_11.0153   \n",
       "3831             NODE_94_length_65207_cov_49.5489   \n",
       "3854             NODE_82_length_57976_cov_50.4232   \n",
       "3855             NODE_82_length_57976_cov_50.4232   \n",
       "3949            NODE_151_length_51672_cov_99.9895   \n",
       "4002             NODE_96_length_51336_cov_1088.99   \n",
       "4006             NODE_66_length_67225_cov_92.2439   \n",
       "4464            NODE_14_length_177436_cov_11.6738   \n",
       "4614             NODE_3808_length_4248_cov_2.6041   \n",
       "6665             NODE_35_length_40069_cov_14.9237   \n",
       "7622              NODE_92_length_29327_cov_8.9586   \n",
       "7637             NODE_46_length_45597_cov_7.57947   \n",
       "7638             NODE_44_length_51823_cov_17.3485   \n",
       "7704            NODE_32_length_111102_cov_34.8219   \n",
       "10452            NODE_5_length_339048_cov_101.011   \n",
       "10529           NODE_8163_length_2246_cov_1.94432   \n",
       "10585           NODE_267_length_56193_cov_4.66183   \n",
       "10630           NODE_182_length_91995_cov_4.66868   \n",
       "19847           NODE_300_length_28736_cov_3.78449   \n",
       "21404           NODE_353_length_20445_cov_5.46126   \n",
       "21440          NODE_135_length_106229_cov_4.73162   \n",
       "24219            NODE_32_length_67127_cov_10.9528   \n",
       "26561           NODE_613_length_43171_cov_3.94376   \n",
       "27219            NODE_1_length_811617_cov_14.3029   \n",
       "27993           NODE_247_length_84988_cov_4.56296   \n",
       "28084           NODE_167_length_76680_cov_8.70181   \n",
       "28312           NODE_80_length_121637_cov_4.97806   \n",
       "28442            NODE_24_length_176822_cov_5.8476   \n",
       "30584  NODE_3376_length_8018_cov_10.0000_ID_68660   \n",
       "32475   NODE_3204_length_6699_cov_5.0000_ID_15084   \n",
       "33316            NODE_50_length_91725_cov_10.2022   \n",
       "33593           NODE_1431_length_7325_cov_6.40248   \n",
       "\n",
       "                                          Genome Name            Study  \\\n",
       "274                    BackhedF_2015__SID203_B__bin.2    BackhedF_2015   \n",
       "275                    BackhedF_2015__SID203_B__bin.2    BackhedF_2015   \n",
       "2214                   BritoIL_2016__M1.1.ST__bin.108     BritoIL_2016   \n",
       "2215                   BritoIL_2016__M1.1.ST__bin.108     BritoIL_2016   \n",
       "3831                CM_madagascar__A13_03_1FE__bin.27    CM_madagascar   \n",
       "3854                 CM_madagascar__A20_03_1FE__bin.9    CM_madagascar   \n",
       "3855                 CM_madagascar__A20_03_1FE__bin.9    CM_madagascar   \n",
       "3949                 CM_madagascar__A53_01_1FE__bin.2    CM_madagascar   \n",
       "4002                CM_madagascar__A68_01_1FE__bin.24    CM_madagascar   \n",
       "4006                CM_madagascar__A69_02_1FE__bin.22    CM_madagascar   \n",
       "4464                 ChengpingW_2017__AS10raw__bin.39  ChengpingW_2017   \n",
       "4614                  ChengpingW_2017__AS37raw__bin.3  ChengpingW_2017   \n",
       "6665                  DavidLA_2015__LD-Run1-10__bin.2     DavidLA_2015   \n",
       "7622   FerrettiP_2018__CA_C10023IS2397FE_t3M15__bin.1   FerrettiP_2018   \n",
       "7637   FerrettiP_2018__CA_C10026IS2444FE_t1M15__bin.4   FerrettiP_2018   \n",
       "7638   FerrettiP_2018__CA_C10026IS2449FE_t2M15__bin.2   FerrettiP_2018   \n",
       "7704   FerrettiP_2018__CA_C10042IS2731FE_t2M16__bin.6   FerrettiP_2018   \n",
       "10452              HeQ_2017__RSZAXPI003108-148__bin.2         HeQ_2017   \n",
       "10529                HeQ_2017__SZAXPI029486-84__bin.3         HeQ_2017   \n",
       "10585              HeQ_2017__SZAXPI029502-105__bin.12         HeQ_2017   \n",
       "10630               HeQ_2017__SZAXPI029553-32__bin.14         HeQ_2017   \n",
       "19847                       QinJ_2012__DOF010__bin.10        QinJ_2012   \n",
       "21404                        QinN_2014__LD-62__bin.17        QinN_2014   \n",
       "21440                        QinN_2014__LD-69__bin.10        QinN_2014   \n",
       "24219    VogtmannE_2016__MMRS51728985ST-27-0-0__bin.2   VogtmannE_2016   \n",
       "26561              XieH_2016__YSZC12003_36697__bin.20        XieH_2016   \n",
       "27219              XieH_2016__YSZC12003_37276__bin.66        XieH_2016   \n",
       "27993                 YuJ_2015__SZAXPI003434-7__bin.5         YuJ_2015   \n",
       "28084              YuJ_2015__SZAXPI015219-129__bin.21         YuJ_2015   \n",
       "28312               YuJ_2015__SZAXPI015257-56__bin.40         YuJ_2015   \n",
       "28442                YuJ_2015__SZAXPI017451-16__bin.7         YuJ_2015   \n",
       "30584                ZeeviD_2015__PNP_Main_40__bin.41      ZeeviD_2015   \n",
       "32475                ZeeviD_2015__PNP_Main_99__bin.11      ZeeviD_2015   \n",
       "33316        ZellerG_2014__CCIS59903910ST-4-0__bin.16     ZellerG_2014   \n",
       "33593       ZellerG_2014__CCIS93040568ST-20-0__bin.13     ZellerG_2014   \n",
       "\n",
       "                   Sample Name  SGB ID            Pos  \\\n",
       "274                   SID203_B    8018    20514 20813   \n",
       "275                   SID203_B    8018    20810 21355   \n",
       "2214                   M1.1.ST    8018  122069 122356   \n",
       "2215                   M1.1.ST    8018  122457 125459   \n",
       "3831                A13_03_1FE    8018    58862 62989   \n",
       "3854                A20_03_1FE    8018     6554 10681   \n",
       "3855                A20_03_1FE    8018    15251 18640   \n",
       "3949                A53_01_1FE    8018    37456 41583   \n",
       "4002                A68_01_1FE    8018    37067 41194   \n",
       "4006                A69_02_1FE    8018    59139 63266   \n",
       "4464                   AS10raw    8018    64299 67691   \n",
       "4614                   AS37raw    8018       114 4235   \n",
       "6665                LD-Run1-10    8018    33228 37367   \n",
       "7622   CA_C10023IS2397FE_t3M15    8018    10441 14574   \n",
       "7637   CA_C10026IS2444FE_t1M15    8018    10441 14574   \n",
       "7638   CA_C10026IS2449FE_t2M15    8018    10433 14566   \n",
       "7704   CA_C10042IS2731FE_t2M16    8018    10450 14571   \n",
       "10452        RSZAXPI003108-148    8018  158056 161448   \n",
       "10529          SZAXPI029486-84    8018      1879 2178   \n",
       "10585         SZAXPI029502-105    8018    43363 47496   \n",
       "10630          SZAXPI029553-32    8018    10446 14567   \n",
       "19847                   DOF010    8018    10450 14571   \n",
       "21404                    LD-62    8018      2620 6012   \n",
       "21440                    LD-69    8018  101622 105743   \n",
       "24219    MMRS51728985ST-27-0-0    8018    62899 67032   \n",
       "26561          YSZC12003_36697    8018      5051 8443   \n",
       "27219          YSZC12003_37276    8018  600351 603743   \n",
       "27993           SZAXPI003434-7    8018      4328 8449   \n",
       "28084         SZAXPI015219-129    8018    74346 76553   \n",
       "28312          SZAXPI015257-56    8018    74259 77651   \n",
       "28442          SZAXPI017451-16    8018  162252 166373   \n",
       "30584              PNP_Main_40    8018        36 4100   \n",
       "32475              PNP_Main_99    8018      4350 6671   \n",
       "33316       CCIS59903910ST-4-0    8018    10450 14571   \n",
       "33593      CCIS93040568ST-20-0    8018      4127 7297   \n",
       "\n",
       "                         pilercr_CRISPR  \\\n",
       "274     [('ID=CRISPR1', '17994', '16')]   \n",
       "275     [('ID=CRISPR1', '17994', '16')]   \n",
       "2214    [('ID=CRISPR1', '119549', '9')]   \n",
       "2215    [('ID=CRISPR1', '119549', '9')]   \n",
       "3831                                NaN   \n",
       "3854    [('ID=CRISPR1', '14571', '11')]   \n",
       "3855    [('ID=CRISPR1', '14571', '11')]   \n",
       "3949    [('ID=CRISPR1', '29765', '22')]   \n",
       "4002    [('ID=CRISPR1', '29368', '16')]   \n",
       "4006     [('ID=CRISPR1', '66103', '3')]   \n",
       "4464     [('ID=CRISPR1', '70444', '4')]   \n",
       "4614                                NaN   \n",
       "6665                                NaN   \n",
       "7622    [('ID=CRISPR1', '18329', '13')]   \n",
       "7637     [('ID=CRISPR1', '17999', '8')]   \n",
       "7638     [('ID=CRISPR1', '17991', '8')]   \n",
       "7704    [('ID=CRISPR1', '18857', '15')]   \n",
       "10452   [('ID=CRISPR1', '153165', '9')]   \n",
       "10529                               NaN   \n",
       "10585   [('ID=CRISPR1', '38966', '31')]   \n",
       "10630   [('ID=CRISPR1', '19183', '30')]   \n",
       "19847   [('ID=CRISPR1', '18859', '25')]   \n",
       "21404    [('ID=CRISPR1', '9821', '20')]   \n",
       "21440    [('ID=CRISPR1', '96496', '8')]   \n",
       "24219   [('ID=CRISPR1', '58371', '15')]   \n",
       "26561   [('ID=CRISPR1', '11724', '12')]   \n",
       "27219  [('ID=CRISPR1', '597831', '30')]   \n",
       "27993   [('ID=CRISPR1', '11942', '13')]   \n",
       "28084   [('ID=CRISPR1', '71826', '12')]   \n",
       "28312    [('ID=CRISPR1', '80535', '6')]   \n",
       "28442   [('ID=CRISPR1', '158248', '7')]   \n",
       "30584        [('ID=CRISPR1', '.', '7')]   \n",
       "32475                               NaN   \n",
       "33316   [('ID=CRISPR1', '19319', '28')]   \n",
       "33593    [('ID=CRISPR2', '1464', '16')]   \n",
       "\n",
       "                                           minced_CRISPR  \\\n",
       "274                            [('.', '16904', '17994')]   \n",
       "275                            [('.', '16904', '17994')]   \n",
       "2214                         [('.', '118396', '119549')]   \n",
       "2215                         [('.', '118396', '119549')]   \n",
       "3831                                                 NaN   \n",
       "3854   [('.', '13349', '14571'), ('.', '21160', '2244...   \n",
       "3855   [('.', '13349', '14571'), ('.', '21160', '2244...   \n",
       "3949                           [('.', '28343', '29765')]   \n",
       "4002                           [('.', '28343', '29368')]   \n",
       "4006                           [('.', '65934', '66564')]   \n",
       "4464                           [('.', '70211', '71830')]   \n",
       "4614                                                 NaN   \n",
       "6665                                                 NaN   \n",
       "7622   [('.', '17237', '18329'), ('.', '18971', '2032...   \n",
       "7637   [('.', '17237', '17999'), ('.', '18641', '1986...   \n",
       "7638   [('.', '17229', '17991'), ('.', '18633', '1985...   \n",
       "7704   [('.', '17237', '18857'), ('.', '19499', '2164...   \n",
       "10452                        [('.', '152537', '155536')]   \n",
       "10529                                                NaN   \n",
       "10585  [('.', '36950', '38966'), ('.', '39608', '4070...   \n",
       "10630  [('.', '17233', '19183'), ('.', '19825', '2137...   \n",
       "19847  [('.', '17237', '18859'), ('.', '19501', '1993...   \n",
       "21404                           [('.', '8532', '10941')]   \n",
       "21440  [('.', '95800', '96496'), ('.', '97138', '9895...   \n",
       "24219  [('.', '56818', '58371'), ('.', '59013', '6023...   \n",
       "26561                          [('.', '10963', '12845')]   \n",
       "27219                        [('.', '594961', '597831')]   \n",
       "27993  [('.', '11115', '12602'), ('.', '13244', '1374...   \n",
       "28084                          [('.', '71002', '71826')]   \n",
       "28312                          [('.', '80171', '81987')]   \n",
       "28442  [('.', '157684', '158248'), ('.', '158890', '1...   \n",
       "30584     [('.', '6770', '7203'), ('.', '7846', '8015')]   \n",
       "32475                                                NaN   \n",
       "33316  [('.', '17237', '19319'), ('.', '19961', '2171...   \n",
       "33593                             [('.', '438', '1464')]   \n",
       "\n",
       "                                             prokka_cas1  \\\n",
       "274            [('ID=KMPIJEMK_00756', '19430', '20341')]   \n",
       "275            [('ID=KMPIJEMK_00756', '19430', '20341')]   \n",
       "2214         [('ID=OEDMBJBD_00131', '120985', '121896')]   \n",
       "2215         [('ID=OEDMBJBD_00131', '120985', '121896')]   \n",
       "3831           [('ID=IOOHGADO_00344', '62989', '63855')]   \n",
       "3854   [('ID=IKGOIMFN_00492', '10681', '11547'), ('ID...   \n",
       "3855   [('ID=IKGOIMFN_00492', '10681', '11547'), ('ID...   \n",
       "3949           [('ID=CMOIDPAO_00381', '36590', '37456')]   \n",
       "4002           [('ID=AIJKJLCA_00347', '36201', '37067')]   \n",
       "4006           [('ID=EJAINPIA_00528', '63266', '64132')]   \n",
       "4464           [('ID=KKNCMIMF_00310', '67864', '68775')]   \n",
       "4614                                                 NaN   \n",
       "6665           [('ID=LFBAJCND_01268', '37367', '38233')]   \n",
       "7622           [('ID=OMDKNLNP_01484', '14574', '15440')]   \n",
       "7637           [('ID=CDJLDPBF_00867', '14574', '15440')]   \n",
       "7638           [('ID=OCFCKFDJ_01245', '14566', '15432')]   \n",
       "7704           [('ID=CMOGHOMO_00556', '14571', '15437')]   \n",
       "10452        [('ID=BAEJBBCH_00524', '156972', '157883')]   \n",
       "10529             [('ID=PHEOENKP_01064', '795', '1706')]   \n",
       "10585          [('ID=HLPGCKEJ_00911', '42497', '43363')]   \n",
       "10630          [('ID=GFDNGKMC_00296', '14567', '15433')]   \n",
       "19847          [('ID=NKOAIHNC_01427', '14571', '15437')]   \n",
       "21404            [('ID=OMJCMEOF_01998', '6185', '7096')]   \n",
       "21440        [('ID=ANGLCDIO_00391', '100756', '101622')]   \n",
       "24219          [('ID=MLDCFJFN_01354', '62033', '62899')]   \n",
       "26561            [('ID=MPHFNKCK_01165', '8616', '9527')]   \n",
       "27219        [('ID=EEPDNCNB_00595', '599267', '600178')]   \n",
       "27993            [('ID=PHIIAFGJ_00211', '8449', '9315')]   \n",
       "28084          [('ID=DAJGLKOL_01121', '73262', '74173')]   \n",
       "28312          [('ID=GEFHLDIP_00072', '77824', '78735')]   \n",
       "28442        [('ID=IMAHBFIH_00346', '161386', '162252')]   \n",
       "30584            [('ID=KIPICGBL_01759', '4100', '4966')]   \n",
       "32475            [('ID=GMAILNMJ_00133', '3484', '4350')]   \n",
       "33316          [('ID=DEJDICEE_00396', '14571', '15437')]   \n",
       "33593            [('ID=PMBNEDJK_01978', '3261', '4127')]   \n",
       "\n",
       "                                             prokka_cas2 uSGB  \\\n",
       "274            [('ID=KMPIJEMK_00755', '19105', '19428')]   No   \n",
       "275            [('ID=KMPIJEMK_00755', '19105', '19428')]   No   \n",
       "2214         [('ID=OEDMBJBD_00130', '120660', '120983')]   No   \n",
       "2215         [('ID=OEDMBJBD_00130', '120660', '120983')]   No   \n",
       "3831           [('ID=IOOHGADO_00345', '63852', '64196')]   No   \n",
       "3854   [('ID=IKGOIMFN_00493', '11559', '11888'), ('ID...   No   \n",
       "3855   [('ID=IKGOIMFN_00493', '11559', '11888'), ('ID...   No   \n",
       "3949           [('ID=CMOIDPAO_00380', '36249', '36578')]   No   \n",
       "4002           [('ID=AIJKJLCA_00346', '35860', '36189')]   No   \n",
       "4006           [('ID=EJAINPIA_00529', '64129', '64473')]   No   \n",
       "4464           [('ID=KKNCMIMF_00311', '68777', '69100')]   No   \n",
       "4614                                                 NaN   No   \n",
       "6665           [('ID=LFBAJCND_01269', '38245', '38574')]   No   \n",
       "7622           [('ID=OMDKNLNP_01485', '15452', '15781')]   No   \n",
       "7637           [('ID=CDJLDPBF_00868', '15452', '15781')]   No   \n",
       "7638           [('ID=OCFCKFDJ_01246', '15444', '15773')]   No   \n",
       "7704           [('ID=CMOGHOMO_00557', '15449', '15778')]   No   \n",
       "10452        [('ID=BAEJBBCH_00523', '156647', '156970')]   No   \n",
       "10529              [('ID=PHEOENKP_01063', '470', '778')]   No   \n",
       "10585          [('ID=HLPGCKEJ_00910', '42156', '42485')]   No   \n",
       "10630          [('ID=GFDNGKMC_00297', '15445', '15774')]   No   \n",
       "19847          [('ID=NKOAIHNC_01428', '15449', '15778')]   No   \n",
       "21404            [('ID=OMJCMEOF_01999', '7098', '7421')]   No   \n",
       "21440        [('ID=ANGLCDIO_00390', '100415', '100744')]   No   \n",
       "24219          [('ID=MLDCFJFN_01353', '61692', '62021')]   No   \n",
       "26561            [('ID=MPHFNKCK_01166', '9529', '9852')]   No   \n",
       "27219        [('ID=EEPDNCNB_00594', '598942', '599265')]   No   \n",
       "27993            [('ID=PHIIAFGJ_00212', '9327', '9656')]   No   \n",
       "28084          [('ID=DAJGLKOL_01120', '72937', '73260')]   No   \n",
       "28312          [('ID=GEFHLDIP_00073', '78737', '79060')]   No   \n",
       "28442        [('ID=IMAHBFIH_00345', '161045', '161374')]   No   \n",
       "30584            [('ID=KIPICGBL_01760', '4963', '5307')]   No   \n",
       "32475            [('ID=GMAILNMJ_00132', '3143', '3487')]   No   \n",
       "33316          [('ID=DEJDICEE_00397', '15449', '15778')]   No   \n",
       "33593            [('ID=PMBNEDJK_01977', '2920', '3249')]   No   \n",
       "\n",
       "      Level of estimated taxonomy  \\\n",
       "274                       Species   \n",
       "275                       Species   \n",
       "2214                      Species   \n",
       "2215                      Species   \n",
       "3831                      Species   \n",
       "3854                      Species   \n",
       "3855                      Species   \n",
       "3949                      Species   \n",
       "4002                      Species   \n",
       "4006                      Species   \n",
       "4464                      Species   \n",
       "4614                      Species   \n",
       "6665                      Species   \n",
       "7622                      Species   \n",
       "7637                      Species   \n",
       "7638                      Species   \n",
       "7704                      Species   \n",
       "10452                     Species   \n",
       "10529                     Species   \n",
       "10585                     Species   \n",
       "10630                     Species   \n",
       "19847                     Species   \n",
       "21404                     Species   \n",
       "21440                     Species   \n",
       "24219                     Species   \n",
       "26561                     Species   \n",
       "27219                     Species   \n",
       "27993                     Species   \n",
       "28084                     Species   \n",
       "28312                     Species   \n",
       "28442                     Species   \n",
       "30584                     Species   \n",
       "32475                     Species   \n",
       "33316                     Species   \n",
       "33593                     Species   \n",
       "\n",
       "                                      Estimated taxonomy  \\\n",
       "274    k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "275    k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "2214   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "2215   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "3831   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "3854   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "3855   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "3949   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "4002   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "4006   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "4464   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "4614   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "6665   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "7622   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "7637   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "7638   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "7704   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "10452  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "10529  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "10585  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "10630  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "19847  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "21404  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "21440  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "24219  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "26561  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "27219  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "27993  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "28084  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "28312  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "28442  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "30584  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "32475  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "33316  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "33593  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
       "\n",
       "                                     prokka_cas9  \n",
       "274      ('ID=KMPIJEMK_00757', '20514', '20813')  \n",
       "275      ('ID=KMPIJEMK_00758', '20810', '21355')  \n",
       "2214   ('ID=OEDMBJBD_00132', '122069', '122356')  \n",
       "2215   ('ID=OEDMBJBD_00133', '122457', '125459')  \n",
       "3831     ('ID=IOOHGADO_00343', '58862', '62989')  \n",
       "3854      ('ID=IKGOIMFN_00491', '6554', '10681')  \n",
       "3855     ('ID=IKGOIMFN_00499', '15251', '18640')  \n",
       "3949     ('ID=CMOIDPAO_00382', '37456', '41583')  \n",
       "4002     ('ID=AIJKJLCA_00348', '37067', '41194')  \n",
       "4006     ('ID=EJAINPIA_00527', '59139', '63266')  \n",
       "4464     ('ID=KKNCMIMF_00309', '64299', '67691')  \n",
       "4614        ('ID=BAOICCCB_01645', '114', '4235')  \n",
       "6665     ('ID=LFBAJCND_01267', '33228', '37367')  \n",
       "7622     ('ID=OMDKNLNP_01483', '10441', '14574')  \n",
       "7637     ('ID=CDJLDPBF_00866', '10441', '14574')  \n",
       "7638     ('ID=OCFCKFDJ_01244', '10433', '14566')  \n",
       "7704     ('ID=CMOGHOMO_00555', '10450', '14571')  \n",
       "10452  ('ID=BAEJBBCH_00525', '158056', '161448')  \n",
       "10529      ('ID=PHEOENKP_01065', '1879', '2178')  \n",
       "10585    ('ID=HLPGCKEJ_00912', '43363', '47496')  \n",
       "10630    ('ID=GFDNGKMC_00295', '10446', '14567')  \n",
       "19847    ('ID=NKOAIHNC_01426', '10450', '14571')  \n",
       "21404      ('ID=OMJCMEOF_01997', '2620', '6012')  \n",
       "21440  ('ID=ANGLCDIO_00392', '101622', '105743')  \n",
       "24219    ('ID=MLDCFJFN_01355', '62899', '67032')  \n",
       "26561      ('ID=MPHFNKCK_01164', '5051', '8443')  \n",
       "27219  ('ID=EEPDNCNB_00596', '600351', '603743')  \n",
       "27993      ('ID=PHIIAFGJ_00210', '4328', '8449')  \n",
       "28084    ('ID=DAJGLKOL_01122', '74346', '76553')  \n",
       "28312    ('ID=GEFHLDIP_00071', '74259', '77651')  \n",
       "28442  ('ID=IMAHBFIH_00347', '162252', '166373')  \n",
       "30584        ('ID=KIPICGBL_01758', '36', '4100')  \n",
       "32475      ('ID=GMAILNMJ_00134', '4350', '6671')  \n",
       "33316    ('ID=DEJDICEE_00395', '10450', '14571')  \n",
       "33593      ('ID=PMBNEDJK_01979', '4127', '7297')  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gallo_data=cas_dataset[cas_dataset[\"SGB ID\"]==8018]\n",
    "gallo_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35, 17)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gallo_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'SGB_table' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-815919760076>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mSGB_abundance_in_dataset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgallo_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"SGB ID\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"Seq ID\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Seq ID\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m15\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mSGB_abundance_in_dataset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mSGB_abundance_in_dataset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrename\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"# Genomes with a cas9\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mSGB_rel_ab_in_dataset\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSGB_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mSGB_table\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"SGB ID\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0misin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSGB_abundance_in_dataset\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"SGB ID\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"# Reconstructed genomes\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mSGB_abundance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSGB_abundance_in_dataset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmerge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mSGB_rel_ab_in_dataset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mleft_on\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'SGB ID'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mright_on\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'SGB ID'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mSGB_abundance\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"Genomes relative abundance\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mSGB_abundance\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"# Genomes with act. wrk. cas9\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mSGB_abundance\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"# Reconstructed genomes\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'SGB_table' is not defined"
     ]
    }
   ],
   "source": [
    "SGB_abundance_in_dataset=gallo_data.groupby([\"SGB ID\"]).count().sort_values(by=\"Seq ID\", ascending=False)[\"Seq ID\"].head(15)\n",
    "SGB_abundance_in_dataset=SGB_abundance_in_dataset.rename(\"# Genomes with a cas9\")\n",
    "SGB_rel_ab_in_dataset=pd.DataFrame(SGB_table[SGB_table[\"SGB ID\"].isin(SGB_abundance_in_dataset.index)][[\"SGB ID\",\"# Reconstructed genomes\"]])\n",
    "SGB_abundance=pd.DataFrame(SGB_abundance_in_dataset).merge(SGB_rel_ab_in_dataset, left_on='SGB ID', right_on='SGB ID')\n",
    "SGB_abundance[\"Genomes relative abundance\"]=SGB_abundance[\"# Genomes with act. wrk. cas9\"]/SGB_abundance[\"# Reconstructed genomes\"]\n",
    "SGB_abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SEQ ID:  OEDMBJBD_00133\n",
      "NODE_71_length_152938_cov_11.0153\n",
      "BritoIL_2016\n",
      "BritoIL_2016__M1.1.ST__bin.108\n",
      "1000\n",
      "[('.', '118396', '119549')]\n",
      "----------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "seqid =\"OEDMBJBD_00133\"\n",
    "print(\"SEQ ID: \", seqid)\n",
    "print(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Contig\"].iloc[0])\n",
    "print(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Study\"].iloc[0])\n",
    "print(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Genome Name\"].iloc[0])\n",
    "print(len(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Seq\"].iloc[0]))\n",
    "print(cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"minced_CRISPR\"].iloc[0])\n",
    "print(\"----------------------------------------------------------------------\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              Seq ID                                    Seq Description  \\\n",
      "2215  OEDMBJBD_00133  OEDMBJBD_00133 CRISPR-associated endonuclease ...   \n",
      "\n",
      "                                                    Seq  \\\n",
      "2215  MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAE...   \n",
      "\n",
      "                                 Contig                     Genome Name  \\\n",
      "2215  NODE_71_length_152938_cov_11.0153  BritoIL_2016__M1.1.ST__bin.108   \n",
      "\n",
      "             Study Sample Name  SGB ID            Pos  \\\n",
      "2215  BritoIL_2016     M1.1.ST    8018  122457 125459   \n",
      "\n",
      "                       pilercr_CRISPR                minced_CRISPR  \\\n",
      "2215  [('ID=CRISPR1', '119549', '9')]  [('.', '118396', '119549')]   \n",
      "\n",
      "                                      prokka_cas1  \\\n",
      "2215  [('ID=OEDMBJBD_00131', '120985', '121896')]   \n",
      "\n",
      "                                      prokka_cas2 uSGB  \\\n",
      "2215  [('ID=OEDMBJBD_00130', '120660', '120983')]   No   \n",
      "\n",
      "     Level of estimated taxonomy  \\\n",
      "2215                     Species   \n",
      "\n",
      "                                     Estimated taxonomy  \\\n",
      "2215  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...   \n",
      "\n",
      "                                    prokka_cas9  \n",
      "2215  ('ID=OEDMBJBD_00133', '122457', '125459')  \n"
     ]
    }
   ],
   "source": [
    "this_cas=cas_dataset[cas_dataset[\"Seq ID\"]==seqid]\n",
    "print(this_cas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-----------------------------------------------------\n",
      "Cas9 id:\tOEDMBJBD_00133\n",
      "SGB:\t8018\n",
      "Genome Name:\t BritoIL_2016__M1.1.ST__bin.108\n",
      "Contig:\tNODE_71_length_152938_cov_11.0153 \n",
      "-----------------------------------------------------\n",
      "\n",
      "Sequence 'NODE_71_length_152938_cov_11.0153' (152938 bp)\n",
      "\n",
      "\n",
      "\n",
      "CRISPR 1   Range: 118396 - 119549\n",
      "\n",
      "POSITION\tREPEAT\t\t\t\tSPACER\n",
      "\n",
      "--------\t------------------------------------\t-----------------------------\n",
      "\n",
      "118396\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tTTTTCGTTCCACAATGTCAATCAAATCAAC\t[ 36, 30 ]\n",
      "\n",
      "118462\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tTAATTAGTGTCGTAGTAAAGACAATCATAT\t[ 36, 30 ]\n",
      "\n",
      "118528\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tGATTAGTATTATCTTCAAATATAGATCCTC\t[ 36, 30 ]\n",
      "\n",
      "118594\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tCACCTTTATTCAACTTAATTTCGTACATAA\t[ 36, 30 ]\n",
      "\n",
      "118660\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tTAGATGATATGTTAGAAAGTCTTGATATTG\t[ 36, 30 ]\n",
      "\n",
      "118726\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tGAACATAGTGAGAGGATTTAGTTTTTAAAA\t[ 36, 30 ]\n",
      "\n",
      "118792\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tCAACCTGTTGTTTGTAAAATTTCCTCAAC\t[ 36, 29 ]\n",
      "\n",
      "118857\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tGACGTTAAAGTCCAATTATACACGTGCGCC\t[ 36, 30 ]\n",
      "\n",
      "118923\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tTTTACTTTATCAGTTGTAACAATCACTAAA\t[ 36, 30 ]\n",
      "\n",
      "118989\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tTCGAATTAGGTTTCATAGTAATAATATAA\t[ 36, 29 ]\n",
      "\n",
      "119054\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tTCATTATCATTCACAAGGATTTTTATAAA\t[ 36, 29 ]\n",
      "\n",
      "119119\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tGTTTCAGACTATGTACATTTTTAATTTGAA\t[ 36, 30 ]\n",
      "\n",
      "119185\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tAGAGTTTAAATACCTCAAATATATTTCTGG\t[ 36, 30 ]\n",
      "\n",
      "119251\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tTCAAGTTGCCATTGACCATTCTTTTATGCT\t[ 36, 30 ]\n",
      "\n",
      "119317\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tGAAAAGTCTTGCTTGGTAAACCTGTATTCG\t[ 36, 30 ]\n",
      "\n",
      "119383\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tCGATAAATGTTATGGCATTTTTGGTGATA\t[ 36, 29 ]\n",
      "\n",
      "119448\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\tCAAATAATAAATTAGCAACCCAAGTAATAC\t[ 36, 30 ]\n",
      "\n",
      "119514\t\tGTTTTACGGTTACTTAAATCTTGAGAGTACAAAAAC\t\n",
      "\n",
      "--------\t------------------------------------\t-----------------------------\n",
      "\n",
      "Repeats: 18\tAverage Length: 36\t\tAverage Length: 29\n",
      "\n",
      "\n",
      "\n",
      "Time to find repeats: 336 ms\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Retrieving locus\n",
      "CRISPR si trova 118396 119549\n",
      "Cas9 si trova [122457, 125459]\n",
      "Cas2 si trova [120660, 120983]\n",
      "Cas1 si trova [120985, 121896]\n",
      "1000.0 1000\n",
      "Cas locus on minus strand\n",
      "Is Cas9 translated from the genome equal to the orignal annotation? True\n",
      "MTNGKILGLDIGIASVGVGIIEAKTGKVVHANSRLFSAANAENNAERRGFRGSRRLNRRKKHRVKRVRDLFEKYEIITDFRNLNLNPYELRVKGLTEQLTNEELFAALRTISKRRGISYLDDAEDDSTGSTDYAKSIDENRRLLKTKTPGQIQLERLEKYGQLRGNFTVYDENGDAHRLINVFSTSDYEKEARKILETQADYNKQITAEFIDDYVEILTQKRKYYHGPGNEKSRTDYGRFRTDGTTIENIFGILIGKCNFYLDEYRASKASYTAQEYNFLNDLNNLKVPTETGKLSTEQKESLVEFAKNTVTLGPAKLLKEIAKILDCKVDEIKGYREDSKGKPDLHTFEPYRKLKFNLTTVDIDNLSRDILDKLADILTLNTERESIEDAINRNLPEQFTKEQISEIIKVRKSQSTAFNKGWHSFSAKLMNELIPELYATSDEQMTILTRLEKFKVNKKSSKNTKTIDEKEVTDEIYNPVVAKSVRQTIKIINAAVKKYGDFDKIVIEMPRDKNADDEKKFIDKKNKENKKEKDDSLKRAAYLYNGTDKLPDDVFHGNKQLATKIRLWYQQGERCLYSGKPIPIQDLVHNSNNFEIDHILPLSLSFDDSLANKVLVYAWTNQEKGQKTPYQVIDSMDAAWSFREMKDYVLKQKGIGKKKRDYLLTTENIDKIEVKKKFIERNLVDTRYASRVVLNSLQSALRELGKDTRVSVVRGQFTSQLRRKWKIDKSRETYHHHAVDALIIAASSQLKLWQKHENLMFENYGENQVVNKETGEILSISDDEYKELVFQPPYQGFVNTISSKAFEDEILFSYQIDSKFNRKISDATIYSTRKAKLGKDKKEETYVLGKIKDIYSQDGFDTFIKRYKKDKTQFLMYQKDPLTWENVIEVILRDYPTTKKSENGKDDVKCNPFEEYRRENGLICKYSKQGKGTPIKSLKYYDTVLKTTSGLDITPDESRNKIILQSLNPWRTDVYFNAKTLKYELAGIRYCDLQYLKGG\n",
      "start:  118396\n",
      "end:  125459\n",
      "let's now blast the repeat against  the genome\n",
      "Running blastn of query file agianst all contigs\n",
      "blastn -out /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/8018__OEDMBJBD_00133__BritoIL_2016__M1.1.ST__bin.108__Cas9.trcr.blastout -outfmt \"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\" -query temp_repeat_seq -db /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb -evalue 0.001 -word_size 11 -penalty -2\n"
     ]
    }
   ],
   "source": [
    "spacers=[]\n",
    "repeats=[]\n",
    "repeat_start_pos=[]\n",
    "\n",
    "\n",
    "cas9_aa=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Seq\"].iloc[0]\n",
    "dataset=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Study\"].iloc[0]\n",
    "genomename=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Genome Name\"].iloc[0]\n",
    "SGB=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"SGB ID\"].iloc[0]\n",
    "contigname=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Contig\"].iloc[0]\n",
    "print(\"\\n-----------------------------------------------------\\n\"+feature+\" id:\\t\"+seqid+\"\\nSGB:\\t\"+str(SGB)+\"\\nGenome Name:\\t\",\\\n",
    "      genomename+\"\\nContig:\\t\"+cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Contig\"].iloc[0],\"\\n-----------------------------------------------------\\n\")\n",
    "\n",
    "######################################################################TODO THIS MUST BE INSIDE filename_discrepancies\n",
    "# take into account filename discrepancies\n",
    "if dataset.startswith(\"ZeeviD\"):  #TODO guarda se funziona quando finisci di girare minced_runner\n",
    "         s3_dataset=\"ZeeviD_2015\"\n",
    "         #genomename=filename_discrepancies.change_to_megahit(genomename)\n",
    "         s3_genomename=genomename\n",
    "         try:\n",
    "             genomename=genomename.replace(\"ZeeviD_2015\", \"ZeeviD_2015_B\")\n",
    "             prokkafile_ofsample=open(annodir+\"/justminced/ZeeviD_2015_B/\"+genomename+\".crisprcas.gff.minced\")\n",
    "             prokkafile_ofsample.close()\n",
    "             dataset=\"ZeeviD_2015_B\"\n",
    "         except:\n",
    "             genomename=genomename.replace(\"ZeeviD_2015_B\", \"ZeeviD_2015_A\")\n",
    "             dataset=\"ZeeviD_2015_A\"\n",
    "else:\n",
    "    s3_dataset=dataset\n",
    "    s3_genomename=genomename\n",
    "    dataset=filename_discrepancies.s3(s3_dataset,r=True) # get working dataset name from s3 dataset name\n",
    "    genomename=genomename.replace(s3_dataset, dataset)\n",
    "\n",
    "if filename_discrepancies.dataset_has_megahit(dataset,genomename):\n",
    "    genomename=filename_discrepancies.change_to_megahit(genomename)\n",
    "\n",
    "samplename = genomename.split(\"__\")[1]\n",
    "########################################################################\n",
    "\n",
    "# get CRISPR spacer and repeat sequence\n",
    "\n",
    "mincedCRISPRfilename=datadir+\"1crisprsearch/out/\"+s3_dataset+\"/\"+s3_genomename+\".fa.minced.out\"\n",
    "f=open(mincedCRISPRfilename, \"r\")\n",
    "for line in f.readlines():\n",
    "    print(line)\n",
    "    if line[0]== \"1\" or line[0]== \"2\" or line[0]== \"3\" or line[0]== \"4\" or line[0]== \"5\" or \\\n",
    "    line[0]== \"6\" or line[0]== \"7\" or line[0]== \"8\" or line[0]== \"9\" or line[0]== \"0\":\n",
    "\n",
    "        repeats.append(line.split(\"\\t\")[2])\n",
    "\n",
    "        repeat_start_pos.append(line.split(\"\\t\")[0])\n",
    "\n",
    "        if not line.split(\"\\t\")[3]==\"\\n\":\n",
    "            spacers.append(line.split(\"\\t\")[3])\n",
    "f.close()\n",
    "\n",
    "# get whole locus sequence todo this could become a function of its own, dentro utils, gli dai genomename della cas e/o seqid della cas (anzi famo solo seqid, che magari ci sn doppioni in genomi) e ti dà tt il locus\n",
    "\n",
    "# print start and stop of CRISPR\n",
    "unostranosplit=cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"minced_CRISPR\"].iloc[0].split(\"\\', \\'\")\n",
    "crispr_start, crispr_stop = int(unostranosplit[1]), int(unostranosplit[2].strip(\"\\')]\\\"\"))\n",
    "\n",
    "# > get start and stop position of all Cass. questo devi metterlo nella tabella di cas non ci sono santi.\n",
    "# scherzi? c'è !\n",
    "\n",
    "prokka_anno=tabellazza[tabellazza[\"Genome Name\"]==s3_genomename][\"prokka_cas\"].iloc[0]\n",
    "# so già che all casses inside here are on the same contig which is 'contig', xke le ho estratte per essere così\n",
    "\n",
    "prokkaslist=prokka_anno.split(\">\")\n",
    "prokkaslist.pop(0)\n",
    "# get contigname of current effectorcas   \n",
    "cas_position={}\n",
    "for casanno in prokkaslist:\n",
    "    unostranosplit=casanno.split(\"Name=\")\n",
    "    casnumber=unostranosplit[1][3] #because we know that we are loooking at 'good' loci, this is just going to be 9, 1 or 2\n",
    "    unaltrostranosplit=casanno.split(\"\\t\")\n",
    "    cas_start=unaltrostranosplit[3]\n",
    "    cas_stop=unaltrostranosplit[4]\n",
    "    cas_position[\"Cas\"+casnumber]=[int(cas_start),int(cas_stop)]\n",
    "\n",
    "genomefilename= \"/shares/CIBIO-Storage/CM/scratch/tmp_projects/epasolli_darkmatter/allcontigs/ALLreconstructedgenomes/\"+str(SGB)+\"/\"+genomename+\".fa\"\n",
    "print(\"Retrieving locus\")\n",
    "\n",
    "\n",
    "# access contig:\n",
    "\n",
    "for record in SeqIO.parse(genomefilename,\"fasta\"):  #c'è modo di non fare questo 'ciclo'?\n",
    "    if record.id.startswith(contigname):\n",
    "        #test se effettivamente la seq di cas è dove voglio io\n",
    "\n",
    "        SeqIO.write(record, \"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb\", \"fasta\")\n",
    "        print(\"CRISPR si trova\", crispr_start, crispr_stop)\n",
    "        print(\"Cas9 si trova\", cas_position[\"Cas9\"])\n",
    "        print(\"Cas2 si trova\", cas_position[\"Cas2\"])\n",
    "        print(\"Cas1 si trova\", cas_position[\"Cas1\"])\n",
    "\n",
    "        #estrai la posizione di cas9 dal contig, e traducila\n",
    "        cas9start=[int(x) for x in cas_position[\"Cas9\"]][0]\n",
    "        cas9stop=[int(x) for x in cas_position[\"Cas9\"]][1]\n",
    "        cas9_nnseq=record.seq[cas9start-1:cas9stop-3] #gff should have a 1-based positional annotation (ma sto andando a occhio finchè non sono uguali, è già un oggetto Bio.Seq\n",
    "\n",
    "        my_translated_cas9=cas9_nnseq.transcribe().translate()\n",
    "\n",
    "  \n",
    "        print(len(cas9_nnseq)/3,len(cas9_aa))\n",
    "\n",
    "\n",
    "        if my_translated_cas9==cas9_aa:\n",
    "            print(\"Cas locus on plus strand\")\n",
    "            plus=True\n",
    "        if not my_translated_cas9==cas9_aa:\n",
    "            #proviamo col reverse complement\n",
    "            cas9_nnseq=record.seq[cas9start+2:cas9stop]\n",
    "            my_translated_cas9=cas9_nnseq.reverse_complement().transcribe().translate()\n",
    "            print(\"Cas locus on minus strand\")\n",
    "        print(\"Is Cas9 translated from the genome equal to the orignal annotation?\", my_translated_cas9==cas9_aa)\n",
    "        print(my_translated_cas9)\n",
    "\n",
    "        #extact a locus, 200 bp a monte e a valle dei cosi che ho:\n",
    "        starts, stops=[startstop[0]for startstop in cas_position.values()], [startstop[1]for startstop in cas_position.values() ]\n",
    "        minimo=min([crispr_start]+starts)\n",
    "        print(\"start: \", minimo)\n",
    "        massimo=max([crispr_stop]+stops)\n",
    "        print(\"end: \", massimo)\n",
    "\n",
    "        #opzione 1 blast\n",
    "      #1. salva np.uniq(repeats) in un fasta file: build temporary query file for blastn search\n",
    "\n",
    "        print(\"let's now blast the repeat against  the genome\")\n",
    "        blast_folder=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/\"\n",
    "        os.chdir(blast_folder)\n",
    "        tempfile=open(\"temp_repeat_seq\", \"w\")\n",
    "        tempfile.close()\n",
    "        tempfile=open(\"temp_repeat_seq\", \"a\")\n",
    "        for n, repeat in enumerate([Seq(sequence) for sequence in np.unique(repeats)]):\n",
    "            tempfile.write(\">rpt\"+str(n+1)+\"|\"+contigname+\"|\"+genomename+\"|\"+seqid+\"\\n\"+str(repeat)+\"\\n\") \n",
    "        tempfile.close()\n",
    "\n",
    "        #2. Blast query file against db\n",
    "        print(\"Running blastn of query file agianst all contigs\")\n",
    "\n",
    "        #3. make db file\n",
    "        dbfile=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/tempblastdb\"\n",
    "        os.system(\"makeblastdb -in \"+dbfile+\" -parse_seqids  -dbtype nucl\")\n",
    "\n",
    "\n",
    "\n",
    "        blastoutfile=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracrRNA/\"\\\n",
    "        +str(SGB)+\"__\"+seqid+\"__\"+genomename+\"__\"+feature+\".trcr.blastout\"\n",
    "        blastn_command = \"blastn -out \"+blastoutfile+\" -outfmt \\\"6  qseqid sseqid pident qlen length mismatch gapopen qseq sseq sstart send evalue sstrand\\\" -query temp_repeat_seq -db \"+dbfile+\" -evalue 0.001 -word_size 11 -penalty -2\"\n",
    "        print(blastn_command)\n",
    "        os.system(blastn_command)\n",
    "       # os.system(\"rm /shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/7tracerRNA/temp*\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-----------------------------------------------------\n",
      "Parsing BLAST output, looking for hit which are NOT the same repeats\n",
      "i.e. not starting with: ['118396', '118462', '118528', '118594', '118660', '118726', '118792', '118857', '118923', '118989', '119054', '119119', '119185', '119251', '119317', '119383', '119448', '119514'] :\n",
      "-----------------------------------------------------\n",
      "\n",
      "rpt1|NODE_71_length_152938_cov_11.0153|BritoIL_2016__M1.1.ST__bin.108|OEDMBJBD_00133\tNODE_71_length_152938_cov_11.0153\t100.00\t36\t18\t0\t0\tACGGTTACTTAAATCTTG\tACGGTTACTTAAATCTTG\t122003\t121986\t2e-04\tminus\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#4. parse blast output: only print output if it is NOT one of the old repeats.\n",
    "print(\"\\n-----------------------------------------------------\\nParsing BLAST output, looking for hit which are NOT the same repeats\\ni.e. not starting with:\", repeat_start_pos,\":\\n-----------------------------------------------------\\n\")\n",
    "f=open(blastoutfile,\"r\")\n",
    "for line in f.readlines():\n",
    "    start_pos=line.split(\"\\t\")[9]\n",
    "    if start_pos not in repeat_start_pos:\n",
    "        if str(int(start_pos)+1) not in repeat_start_pos:\n",
    "            if str(int(start_pos)-1) not in repeat_start_pos:\n",
    "                print(line) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## extract sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# attenzioe: strano: poiché è sul minus strand, quindi start e stop sono al contrario. ma secondo i miei calcoli, per come è fatto blast, dovrebbe hittare sul plus strand..\n",
    "#qst vuol dire che sul - c'è la repeat,ee onn lantirepeat. quindi funge male....\n",
    "# che è testimoniato piu a valle dal fatto che trovo n  terminatore la  vicino, ma lo trovo dopo...\n",
    "#omg che sia un tracrRNA NON trans?\n",
    "# D:\n",
    "\n",
    "# questo potrebbe voler dire che il crispr arraay è in realtà sullo strand opposto?\n",
    "\n",
    "matched_rep_start=18665\n",
    "matched_rep_stop=18691"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GCGTTTTAATTGGAAAGGAAGCGATAA\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "27"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this is the sequence matching the repeat:\n",
    "matched_rep=Seq(\"TTATCGCTTCCTTTCCAATTAAAACGC\")\n",
    "\n",
    "#rpts which match:\n",
    "rpt2=Seq(\"GTTATCGCTTCCTTTCCAATCAAAACGCGGTATAAT\")\n",
    "\n",
    "# since repeat is on the + strand, and this is all on the + strand, tracrRNA should be its reverse complement:\n",
    "antirepeat=matched_rep.reverse_complement()\n",
    "#this should go frombo to bo  = x nt  (x+1)\n",
    "print(antirepeat)\n",
    "len(antirepeat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 18892      Both - CAATTCTTAAAGCCTTGAAGTCCGCCGCTTCAAGGCtTTTTCCGTATCAA -10.20\n",
      "\n",
      " 18895      Both + ATACGGAAAAAGCCTTGAAGCGGCGGACTTCAAGGCTTTAAGAATTGGT -10.20\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# print terminators\n",
    "\n",
    "f=open(blast_folder+str(SGB)+\"__\"+seqid+\"__\"+genomename+\".terminators\")\n",
    "for line in f.readlines():\n",
    "    if line.strip().startswith(\"18\"):\n",
    "        print(line)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminator length:  50\n"
     ]
    }
   ],
   "source": [
    "# this is the rho-independent trnascription terminator, going from 14719 or 14722 to bo (+1 perché conti il primo tipo), \n",
    "\n",
    "terminator_end=18895 -1 # qui è sbagliato, mavabbe proviamo\n",
    "terminator=Seq(\"CATACGGAAAAAGCCTTGAAGCGGCGGACTTCAAGGCTTTAAGAATTGGT\".upper())\n",
    "print(\"terminator length: \", len(terminator))\n",
    "terminator_start=terminator_end+len(terminator) #occhio che qui start e stop si riferisce al rev.compl, mentre nel antirepeat si riferisce alla direzione base\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terminator length:  49\n"
     ]
    }
   ],
   "source": [
    "#se terminator on + strand:\n",
    "terminator_start=18895-1\n",
    "terminator=Seq(\"ATACGGAAAAAGCCTTGAAGCGGCGGACTTCAAGGCTTTAAGAATTGGT\".upper())\n",
    "print(\"terminator length: \", len(terminator))\n",
    "terminator_end=terminator_start+len(terminator)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27\n",
      "test repeat:  GCGTTTTAATTGGAAAGGAAGCGATAA 27 True\n",
      "test terminator:  ATACGGAAAAAGCCTTGAAGCGGCGGACTTCAAGGCTTTAAGAATTGGT 49 True\n",
      "pre tracrRNA: CGTTTTAATTGGAAAGGAAGCGATAATAATCGATAAATTTCGAGTGGGGCTCCGGCGATCCGGTCTTGTATCGGGGTAGCCGTCCCCTAATTTTATACGACAAAAGTATTACTTTCACGCGCTTACTTTTTTCAGTAATCCTGCAGTTTTTATCGCTTTTGTTGAGTTTTTTTGAACAAAGAAGATATAGAACGGTTCCTTTTTTGGGAAAGGATTGTGGAACCGATTGATACGGAAAAAGCCTTGAAGCGGCGGACTTCAAGGCTTTAAGAATTGGT 278\n",
      "mario sequence: TAATCGATAAATTTCGAGTGGGGCTCCGGCGATCCGGTCTTGTATCGGGGTAGCCGTCCCCTAATTTTATACGACAAAAGTATTACTTTCACGCGCTTACTTTTTTCAGTAATCCTGCAGTTTTTATCGCTTTTGTTGAGTTTTTTTGAACAAAGAAGATATAGAACGGTTCCTTTTTTGGGAAAGGATTGTGGAACCGATTG 203\n"
     ]
    }
   ],
   "source": [
    "# Open genome and extract the whole sequence\n",
    "# find this inside the genome, and link it to the thing found by ARNold:\n",
    "#solo per lui:\n",
    "contigname=\"NODE_49_length_176785_cov_12.2725\"\n",
    "for record in SeqIO.parse(\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/selected_loci/\"+seqid+\"/\"+genomename+\".fa\", \"fasta\"):\n",
    "    if record.id==contigname:\n",
    "        contigseq=record.seq\n",
    "print(len(contigseq[matched_rep_start-1:matched_rep_stop]))\n",
    "#check antirepeat\n",
    "print(\"test repeat: \", antirepeat, len(antirepeat), contigseq[matched_rep_start-1:matched_rep_stop]==antirepeat) # -1 why? Indexing diverso evidentemente col blastout, vbb\n",
    "# check terminator\n",
    "print(\"test terminator: \" ,terminator, len(terminator), contigseq[terminator_start:terminator_end]==terminator)\n",
    "\n",
    "\n",
    "pre_tracrRNA=contigseq[matched_rep_start:terminator_end]\n",
    "print(\"pre tracrRNA:\", pre_tracrRNA, len(pre_tracrRNA))\n",
    "\n",
    "# add mario\n",
    "mario=contigseq[matched_rep_stop:terminator_start]\n",
    "print(\"mario sequence:\",  mario, len(mario))  # ho fatto i check, il -1 sopra è giusto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mario un po' lungo, ma a me mi sembra che al primo TTTTTATCG ci si possa fermare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTTTTTCCATAACAGCCTCTATCAAAAATAGTTTTAATATCAGCTTGACCCATAAAAACAGTTATGGCAATATCATTTTTATCGAAAGGGCGATTATTTGCGTTTTAATTGGAAAGGAAGCGATAATAATCGATAAATTTCGAGTGGGGCTCCGGCGATCCGGTCTTGTATCGGGGTAGCCGTCCCCTAATTTTATACGACAAAAGTATTACTTTCACGCGCTTACTTTTTTCAGTAATCCTGCAGTTTTTATCGCTTTTGTTGAGTTTTTTTGAACAAAGAAGATATAGAACGGTTCCTTTTTTGGGAAAGGATTGTGGAACCGATTGATACGGAAAAAGCCTTGAAGCGGCGGACTTCAAGGCTTTAAGAATTGGTGGGCGCGACAAGGATCGAACTTGTGACCCCTACGATGTCAACGTAGTGCTCTCCCGCTGAGCTACGCGCCCGTTGTTGTTGATTATCTCAACAAGATGAT\n"
     ]
    }
   ],
   "source": [
    "print(contigseq[matched_rep_start-100:terminator_end+100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GTTATCGCTTCCTTTCCAATC-AAAACGCGGTATAAT\n",
      " |||||||||||||||||||  |||||||\n",
      "-TTATCGCTTCCTTTCCAAT-TAAAACGC--------\n",
      "  Score=26\n",
      "\n",
      "rpt length: 36\n"
     ]
    }
   ],
   "source": [
    "# Now, align with the rest of the stuff to see what's up\n",
    "# first up, with the repeat, lets take the rpt3, which has only got 1 mismatch\n",
    "\n",
    "alignments = pairwise2.align.localxx(rpt2, antirepeat.reverse_complement())\n",
    "print(pairwise2.format_alignment(*alignments[0]))\n",
    "print(\"rpt length:\", len(rpt2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G-TTAT---CGCTTCCTTTCCAATC-AAAACGCGGTATAAT\n",
      "  ||||   |||||||||||||||  |||||||        \n",
      "-ATTATTATCGCTTCCTTTCCAAT-TAAAACGC--------\n",
      "  Score=26\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# align antirepat + mario with repeat\n",
    "alignments = pairwise2.align.globalxx( rpt2,(antirepeat+mario[:4]).reverse_complement())\n",
    "print(pairwise2.format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#no non fa nulla quel CTTT in più"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "----------------------------------------------------------------------\n",
      "         tracrRNA sequence for Cas9 protein id FLJLKKHA_00012\n",
      "----------------------------------------------------------------------\n",
      "\n",
      "- contig name: NODE_49_length_176785_cov_12.2725\n",
      "- Dataset:  BritoIL_2016\n",
      "- Bin:  BritoIL_2016__M1.29.ST__bin.30\n",
      "- SGB:  9311\n",
      "----------------------------------------------------------------------\n",
      "\n",
      "- Length:  278 \n",
      "- position:  18943 18691\n",
      "CGTTTTAATTGGAAAGGAAGCGATAATAATCGATAAATTTCGAGTGGGGCTCCGGCGATCCGGTCTTGTATCGGGGTAGCCGTCCCCTAATTTTATACGACAAAAGTATTACTTTCACGCGCTTACTTTTTTCAGTAATCCTGCAGTTTTTATCGCTTTTGTTGAGTTTTTTTGAACAAAGAAGATATAGAACGGTTCCTTTTTTGGGAAAGGATTGTGGAACCGATTGATACGGAAAAAGCCTTGAAGCGGCGGACTTCAAGGCTTTAAGAATTGGT\n"
     ]
    }
   ],
   "source": [
    "# I would probably say that the tracr is just, simply la parte di pre_tracrRNA la cui antirepeat si lega alla repeat, \n",
    "# che, in questo caso, è tutta.\n",
    "tracrRNA = pre_tracrRNA\n",
    "\n",
    "print(\"\\n----------------------------------------------------------------------\")\n",
    "print(\"         tracrRNA sequence for\", feature,\"protein id\", seqid)\n",
    "print(\"----------------------------------------------------------------------\\n\")\n",
    "print(\"- contig name:\", cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Contig\"].iloc[0])\n",
    "print(\"- Dataset: \",cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Study\"].iloc[0])\n",
    "print(\"- Bin: \",cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"Genome Name\"].iloc[0])\n",
    "print(\"- SGB: \",cas_dataset[cas_dataset[\"Seq ID\"]==seqid][\"SGB ID\"].iloc[0])\n",
    "print(\"----------------------------------------------------------------------\\n\")\n",
    "print(\"- Length: \", len(tracrRNA),\"\\n- position: \", terminator_end,matched_rep_stop)\n",
    "print(tracrRNA)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Print output\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write file to write output:\n",
    "strand=-1\n",
    "n_bases_after_TTTT=4  #to check\n",
    "#\n",
    "\n",
    "f=open(datadir+\"8outputv1/\"+seqid+\"/\"+seqid+\"inputforoutput\",\"w\")\n",
    "f.write(str(matched_rep)+\"\\n\")\n",
    "f.write(str(rpt3)+\"\\n\")    #check\n",
    "f.write(str(matched_rep_start)+\"\\n\")\n",
    "f.write(str(matched_rep_stop)+\"\\n\")\n",
    "f.write(str(strand)+\"\\n\")  #strand\n",
    "f.write(str(14653-1-(strand*n_bases_after_TTTT))+\"\\n\")  #terminator end from ARNold # -1, for indexing reasons. This position is the stop if strand = -1, the start, if strand = +1\n",
    "f.write(str(terminator[:-n_bases_after_TTTT])+\"\\n\")  # to change if n_bases_after_TTTT \n",
    "f.write(feature+\"\\n\")\n",
    "f.write(str(Cas9_strand))\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
