{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " <h1><center>Data mining for a working Cas9 variant.</center></h1>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<div style=\"text-align: right\">Lorenzo Federico Signorini, 09/08/2019</div>\n",
    "\n",
    "<div style=\"text-align: right\">Universit√† degli studi di Trento</div>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First created Thu Jul 11 09:51:15 CEST 2019\n",
    "# Made by L-F-S\n",
    "# At the University Of Trento, Italy\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from itertools import groupby\n",
    "from Bio import SeqIO\n",
    "from IPython.display import display, HTML\n",
    "sys.path.insert(0, '/home/lorenzo.signorini/cas_mining/utils/')\n",
    "import filename_discrepancies\n",
    "from Bio.Alphabet import generic_dna, generic_protein\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import AlignIO\n",
    "from Bio.Align.Applications import ClustalwCommandline\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio import pairwise2\n",
    "outpath=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/5caslocitable/out/\"\n",
    "datadir=\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# 1. Data:\n",
    "## 1.1 load information about SGBs (species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4930, 10)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SGB_table=pd.read_csv(\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/S4Segata.csv\", index_col=0)\n",
    "SGB_table.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 4930 different species (SGBs) in the dataset, Let us print the 20 most abundant ones:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SGB ID</th>\n",
       "      <th># Reconstructed genomes</th>\n",
       "      <th># Reference genomes</th>\n",
       "      <th>uSGB</th>\n",
       "      <th>Level of estimated taxonomy</th>\n",
       "      <th>Estimated taxonomy</th>\n",
       "      <th>Average distance from closest Reference Genome</th>\n",
       "      <th>Full taxonomic label of the closest genome</th>\n",
       "      <th>Taxonomic assignment based on 16S rRNA sequences (family level)</th>\n",
       "      <th>Taxonomic assignment based on 16S rRNA sequences (genus level)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1490</th>\n",
       "      <td>4933</td>\n",
       "      <td>3457</td>\n",
       "      <td>8</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.025725</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Lachnospiracea_incertae_sedis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1836</td>\n",
       "      <td>2983</td>\n",
       "      <td>20</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>0.017564</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>Bacteroidaceae</td>\n",
       "      <td>Bacteroides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672</th>\n",
       "      <td>2318</td>\n",
       "      <td>2560</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>0.014071</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>Rikenellaceae</td>\n",
       "      <td>Alistipes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>1934</td>\n",
       "      <td>2169</td>\n",
       "      <td>25</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>0.013706</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>Porphyromonadaceae</td>\n",
       "      <td>Parabacteroides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1024</th>\n",
       "      <td>4285</td>\n",
       "      <td>1925</td>\n",
       "      <td>5</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.024487</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Ruminococcaceae</td>\n",
       "      <td>Ruminococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>1814</td>\n",
       "      <td>1844</td>\n",
       "      <td>15</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>0.017179</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>Bacteroidaceae</td>\n",
       "      <td>Bacteroides</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4281</th>\n",
       "      <td>15286</td>\n",
       "      <td>1813</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Family</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.194245</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Ruminococcaceae</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1444</th>\n",
       "      <td>4874</td>\n",
       "      <td>1809</td>\n",
       "      <td>4</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.022680</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Fusicatenibacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1181</th>\n",
       "      <td>4540</td>\n",
       "      <td>1634</td>\n",
       "      <td>10</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.022782</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Anaerostipes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1559</th>\n",
       "      <td>5082</td>\n",
       "      <td>1629</td>\n",
       "      <td>4</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.022264</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Lachnospiracea_incertae_sedis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1486</th>\n",
       "      <td>4925</td>\n",
       "      <td>1587</td>\n",
       "      <td>3</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.020075</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Roseburia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4305</th>\n",
       "      <td>15316</td>\n",
       "      <td>1513</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.041104</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Ruminococcaceae</td>\n",
       "      <td>Faecalibacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1005</th>\n",
       "      <td>4262</td>\n",
       "      <td>1512</td>\n",
       "      <td>5</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.023576</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Ruminococcaceae</td>\n",
       "      <td>Ruminococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td>1965</td>\n",
       "      <td>1432</td>\n",
       "      <td>2</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>0.015707</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "      <td>Porphyromonadaceae</td>\n",
       "      <td>Barnesiella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3055</th>\n",
       "      <td>10068</td>\n",
       "      <td>1413</td>\n",
       "      <td>6193</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Proteobacteria|c__Gammaproteoba...</td>\n",
       "      <td>0.009558</td>\n",
       "      <td>k__Bacteria|p__Proteobacteria|c__Gammaproteoba...</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Escherichia/Shigella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4655</th>\n",
       "      <td>17244</td>\n",
       "      <td>1264</td>\n",
       "      <td>25</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteri...</td>\n",
       "      <td>0.019620</td>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteri...</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4659</th>\n",
       "      <td>17248</td>\n",
       "      <td>1245</td>\n",
       "      <td>59</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteri...</td>\n",
       "      <td>0.016429</td>\n",
       "      <td>k__Bacteria|p__Actinobacteria|c__Actinobacteri...</td>\n",
       "      <td>Bifidobacteriaceae</td>\n",
       "      <td>Bifidobacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4294</th>\n",
       "      <td>15300</td>\n",
       "      <td>1212</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.031550</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>Ruminococcaceae</td>\n",
       "      <td>Gemmiger</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2801</th>\n",
       "      <td>9226</td>\n",
       "      <td>1200</td>\n",
       "      <td>32</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Verrucomicrobia|c__Verrucomicro...</td>\n",
       "      <td>0.012297</td>\n",
       "      <td>k__Bacteria|p__Verrucomicrobia|c__Verrucomicro...</td>\n",
       "      <td>Verrucomicrobiaceae</td>\n",
       "      <td>Akkermansia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4258</th>\n",
       "      <td>15254</td>\n",
       "      <td>1188</td>\n",
       "      <td>0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Family</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "      <td>0.168840</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Firmicutes_unclas...</td>\n",
       "      <td>Ruminococcaceae</td>\n",
       "      <td>Oscillibacter</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      SGB ID  # Reconstructed genomes  # Reference genomes uSGB  \\\n",
       "1490    4933                     3457                    8   No   \n",
       "361     1836                     2983                   20   No   \n",
       "672     2318                     2560                    2   No   \n",
       "438     1934                     2169                   25   No   \n",
       "1024    4285                     1925                    5   No   \n",
       "339     1814                     1844                   15   No   \n",
       "4281   15286                     1813                    0  Yes   \n",
       "1444    4874                     1809                    4   No   \n",
       "1181    4540                     1634                   10   No   \n",
       "1559    5082                     1629                    4   No   \n",
       "1486    4925                     1587                    3   No   \n",
       "4305   15316                     1513                    1   No   \n",
       "1005    4262                     1512                    5   No   \n",
       "467     1965                     1432                    2   No   \n",
       "3055   10068                     1413                 6193   No   \n",
       "4655   17244                     1264                   25   No   \n",
       "4659   17248                     1245                   59   No   \n",
       "4294   15300                     1212                    1   No   \n",
       "2801    9226                     1200                   32   No   \n",
       "4258   15254                     1188                    0  Yes   \n",
       "\n",
       "     Level of estimated taxonomy  \\\n",
       "1490                     Species   \n",
       "361                      Species   \n",
       "672                      Species   \n",
       "438                      Species   \n",
       "1024                     Species   \n",
       "339                      Species   \n",
       "4281                      Family   \n",
       "1444                     Species   \n",
       "1181                     Species   \n",
       "1559                     Species   \n",
       "1486                     Species   \n",
       "4305                     Species   \n",
       "1005                     Species   \n",
       "467                      Species   \n",
       "3055                     Species   \n",
       "4655                     Species   \n",
       "4659                     Species   \n",
       "4294                     Species   \n",
       "2801                     Species   \n",
       "4258                      Family   \n",
       "\n",
       "                                     Estimated taxonomy  \\\n",
       "1490  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "361   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "672   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "438   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "1024  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "339   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "4281  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1444  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1181  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1559  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1486  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "4305  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1005  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "467   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "3055  k__Bacteria|p__Proteobacteria|c__Gammaproteoba...   \n",
       "4655  k__Bacteria|p__Actinobacteria|c__Actinobacteri...   \n",
       "4659  k__Bacteria|p__Actinobacteria|c__Actinobacteri...   \n",
       "4294  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "2801  k__Bacteria|p__Verrucomicrobia|c__Verrucomicro...   \n",
       "4258  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "\n",
       "      Average distance from closest Reference Genome  \\\n",
       "1490                                        0.025725   \n",
       "361                                         0.017564   \n",
       "672                                         0.014071   \n",
       "438                                         0.013706   \n",
       "1024                                        0.024487   \n",
       "339                                         0.017179   \n",
       "4281                                        0.194245   \n",
       "1444                                        0.022680   \n",
       "1181                                        0.022782   \n",
       "1559                                        0.022264   \n",
       "1486                                        0.020075   \n",
       "4305                                        0.041104   \n",
       "1005                                        0.023576   \n",
       "467                                         0.015707   \n",
       "3055                                        0.009558   \n",
       "4655                                        0.019620   \n",
       "4659                                        0.016429   \n",
       "4294                                        0.031550   \n",
       "2801                                        0.012297   \n",
       "4258                                        0.168840   \n",
       "\n",
       "             Full taxonomic label of the closest genome  \\\n",
       "1490  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "361   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "672   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "438   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "1024  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "339   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "4281  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1444  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1181  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1559  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1486  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "4305  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "1005  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "467   k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...   \n",
       "3055  k__Bacteria|p__Proteobacteria|c__Gammaproteoba...   \n",
       "4655  k__Bacteria|p__Actinobacteria|c__Actinobacteri...   \n",
       "4659  k__Bacteria|p__Actinobacteria|c__Actinobacteri...   \n",
       "4294  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...   \n",
       "2801  k__Bacteria|p__Verrucomicrobia|c__Verrucomicro...   \n",
       "4258  k__Bacteria|p__Firmicutes|c__Firmicutes_unclas...   \n",
       "\n",
       "     Taxonomic assignment based on 16S rRNA sequences (family level)  \\\n",
       "1490                                    Lachnospiraceae                \n",
       "361                                      Bacteroidaceae                \n",
       "672                                       Rikenellaceae                \n",
       "438                                  Porphyromonadaceae                \n",
       "1024                                    Ruminococcaceae                \n",
       "339                                      Bacteroidaceae                \n",
       "4281                                    Ruminococcaceae                \n",
       "1444                                    Lachnospiraceae                \n",
       "1181                                    Lachnospiraceae                \n",
       "1559                                    Lachnospiraceae                \n",
       "1486                                    Lachnospiraceae                \n",
       "4305                                    Ruminococcaceae                \n",
       "1005                                    Ruminococcaceae                \n",
       "467                                  Porphyromonadaceae                \n",
       "3055                                 Enterobacteriaceae                \n",
       "4655                                 Bifidobacteriaceae                \n",
       "4659                                 Bifidobacteriaceae                \n",
       "4294                                    Ruminococcaceae                \n",
       "2801                                Verrucomicrobiaceae                \n",
       "4258                                    Ruminococcaceae                \n",
       "\n",
       "     Taxonomic assignment based on 16S rRNA sequences (genus level)  \n",
       "1490                      Lachnospiracea_incertae_sedis              \n",
       "361                                         Bacteroides              \n",
       "672                                           Alistipes              \n",
       "438                                     Parabacteroides              \n",
       "1024                                       Ruminococcus              \n",
       "339                                         Bacteroides              \n",
       "4281                                   Faecalibacterium              \n",
       "1444                                   Fusicatenibacter              \n",
       "1181                                       Anaerostipes              \n",
       "1559                      Lachnospiracea_incertae_sedis              \n",
       "1486                                          Roseburia              \n",
       "4305                                   Faecalibacterium              \n",
       "1005                                       Ruminococcus              \n",
       "467                                         Barnesiella              \n",
       "3055                               Escherichia/Shigella              \n",
       "4655                                    Bifidobacterium              \n",
       "4659                                    Bifidobacterium              \n",
       "4294                                           Gemmiger              \n",
       "2801                                        Akkermansia              \n",
       "4258                                      Oscillibacter              "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SGB_table.sort_values(by=\"# Reconstructed genomes\", ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.2 load information about Cas9s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33978, 16)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cas9dataset=pd.read_csv(\"/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/5caslocitable/known_Cas9_variants_table.csv\", index_col=0)\n",
    "cas9dataset.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 33978 Cas9s in the whole dataset of 154723 genomes.\n",
    "This dataset was created following step 5, from the original Crisprcas_hits_table.csv. Every line is a different Cas9. Here is information about the sequence, the contig (piece of contiguous DNA in one genome),  other cas features on the same contig, and phylogenetical information, printed out for the first 10 Cas9s:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>Seq ID</th>\n",
       "      <th>Seq Description</th>\n",
       "      <th>Seq</th>\n",
       "      <th>Contig</th>\n",
       "      <th>Genome Name</th>\n",
       "      <th>Study</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SGB ID</th>\n",
       "      <th>pilercr_CRISPR</th>\n",
       "      <th>minced_CRISPR</th>\n",
       "      <th>prokka_cas1</th>\n",
       "      <th>prokka_cas2</th>\n",
       "      <th>uSGB</th>\n",
       "      <th>Level of estimated taxonomy</th>\n",
       "      <th>Estimated taxonomy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>EIBEGOMO_01279</td>\n",
       "      <td>EIBEGOMO_01279 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MAEEIITDITIFGESKKMLRETLRKKFASCLDDEAIKKLSKLRYRD...</td>\n",
       "      <td>NODE_1073_length_1994_cov_2.17225</td>\n",
       "      <td>AsnicarF_2017__MV_FEI4_t1Q14__bin.1</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI4_t1Q14</td>\n",
       "      <td>6939</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Negativicutes|o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>KBEGFPIJ_01435</td>\n",
       "      <td>KBEGFPIJ_01435 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MKNIVGLDLGTNSIGWAVVNGSVNDDGSEQLVKIQASGSRIIPMDA...</td>\n",
       "      <td>NODE_45_length_75974_cov_38.2495</td>\n",
       "      <td>AsnicarF_2017__MV_FEI4_t1Q14__bin.4</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI4_t1Q14</td>\n",
       "      <td>1814</td>\n",
       "      <td>[('ID=CRISPR1', '40172', '12')]</td>\n",
       "      <td>[('.', '39283', '40172')]</td>\n",
       "      <td>[('ID=KBEGFPIJ_01437', '37910', '38842')]</td>\n",
       "      <td>[('ID=KBEGFPIJ_01438', '38842', '39180')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>OBHKLEMF_00864</td>\n",
       "      <td>OBHKLEMF_00864 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MTKLNQPYGIGLDIGSNSIGFAVVDANSHLLRLKGETAIGARLFRE...</td>\n",
       "      <td>NODE_51_length_94666_cov_13.1517</td>\n",
       "      <td>AsnicarF_2017__MV_FEI4_t2Q15__bin.2</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI4_t2Q15</td>\n",
       "      <td>7144</td>\n",
       "      <td>[('ID=CRISPR1', '23272', '22')]</td>\n",
       "      <td>[('.', '21587', '23272')]</td>\n",
       "      <td>[('ID=OBHKLEMF_00863', '24253', '25158')]</td>\n",
       "      <td>[('ID=OBHKLEMF_00862', '23970', '24275')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>EPJEOPIP_01092</td>\n",
       "      <td>EPJEOPIP_01092 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MGDRKYNLGLDIGTSSIGFAAVDENNQPIRVKGKTAIGVRLFEEGK...</td>\n",
       "      <td>NODE_70_length_82951_cov_19.4828</td>\n",
       "      <td>AsnicarF_2017__MV_FEI4_t2Q15__bin.5</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI4_t2Q15</td>\n",
       "      <td>7168</td>\n",
       "      <td>[('ID=CRISPR1', '24497', '3')]</td>\n",
       "      <td>[('.', '24197', '25552')]</td>\n",
       "      <td>[('ID=EPJEOPIP_01091', '26532', '27437')]</td>\n",
       "      <td>[('ID=EPJEOPIP_01090', '26249', '26554')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>AHBHJMAE_01497</td>\n",
       "      <td>AHBHJMAE_01497 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MGYILGLDIGVASVGYAIIDENYNVLISGVRLFREGTAEENVARRG...</td>\n",
       "      <td>NODE_123_length_29049_cov_29.8501</td>\n",
       "      <td>AsnicarF_2017__MV_FEI5_t1Q14__bin.12</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI5_t1Q14</td>\n",
       "      <td>6939</td>\n",
       "      <td>[('ID=CRISPR1', '13059', '23')]</td>\n",
       "      <td>[('.', '11506', '13060')]</td>\n",
       "      <td>[('ID=AHBHJMAE_01496', '7315', '8232')]</td>\n",
       "      <td>[('ID=AHBHJMAE_01495', '6974', '7318')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Negativicutes|o__...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>PPJNGKMK_00620</td>\n",
       "      <td>PPJNGKMK_00620 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MGKTHIIGVGLDLGGTYTGTFITSHPSDEAEHRDHSSAFTVVNSEK...</td>\n",
       "      <td>NODE_30_length_116148_cov_68.0265</td>\n",
       "      <td>AsnicarF_2017__MV_FEI5_t2Q14__bin.2</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI5_t2Q14</td>\n",
       "      <td>9262</td>\n",
       "      <td>[('ID=CRISPR1', '29853', '6')]</td>\n",
       "      <td>[('.', '29462', '29925'), ('.', '36897', '3764...</td>\n",
       "      <td>[('ID=PPJNGKMK_00621', '34746', '35738')]</td>\n",
       "      <td>[('ID=PPJNGKMK_00622', '35742', '36029')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Proteobacteria|c__Betaproteobac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>AMIFCEHF_02229</td>\n",
       "      <td>AMIFCEHF_02229 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MKTILGLDLGTTSIGWALVKEAEPFSDEKSEIVRLGVRVIPLTVDE...</td>\n",
       "      <td>NODE_175_length_31817_cov_9.16545</td>\n",
       "      <td>AsnicarF_2017__MV_FEM1_t1Q14__bin.6</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEM1_t1Q14</td>\n",
       "      <td>2230</td>\n",
       "      <td>[('ID=CRISPR2', '25256', '18')]</td>\n",
       "      <td>[('.', '23902', '25256')]</td>\n",
       "      <td>[('ID=AMIFCEHF_02230', '22548', '23462')]</td>\n",
       "      <td>[('ID=AMIFCEHF_02231', '23495', '23800')]</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>HBHGIHPK_01033</td>\n",
       "      <td>HBHGIHPK_01033 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MKSIYEDLKKEKMVISMAKELHFTSKDMPISDFDRMELKGRIETNP...</td>\n",
       "      <td>NODE_4986_length_3564_cov_6.73668</td>\n",
       "      <td>AsnicarF_2017__MV_FEM2_t1Q14__bin.2</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEM2_t1Q14</td>\n",
       "      <td>1024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>No</td>\n",
       "      <td>Species</td>\n",
       "      <td>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>FBOIKEJM_00490</td>\n",
       "      <td>FBOIKEJM_00490 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MILSTQKYIIDTKGCNMKYKLGLDLGSTSLGWAIVELNEKDTPIRL...</td>\n",
       "      <td>NODE_9_length_125774_cov_10.3523</td>\n",
       "      <td>AsnicarF_2017__MV_FEM2_t1Q14__bin.30</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEM2_t1Q14</td>\n",
       "      <td>17322</td>\n",
       "      <td>[('ID=CRISPR1', '113006', '6')]</td>\n",
       "      <td>[('.', '112378', '113006')]</td>\n",
       "      <td>[('ID=FBOIKEJM_00492', '111321', '112217')]</td>\n",
       "      <td>[('ID=FBOIKEJM_00491', '110999', '111337')]</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Other</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>BNAIOCPL_01158</td>\n",
       "      <td>BNAIOCPL_01158 CRISPR-associated endonuclease ...</td>\n",
       "      <td>MDKKYYIGADIGTESVGWAVTDEDYNILKARGRELWGSYLFDKAES...</td>\n",
       "      <td>NODE_2411_length_6235_cov_4.42071</td>\n",
       "      <td>AsnicarF_2017__MV_FEM2_t1Q14__bin.32</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEM2_t1Q14</td>\n",
       "      <td>14099</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[('ID=BNAIOCPL_01159', '4512', '5375')]</td>\n",
       "      <td>[('ID=BNAIOCPL_01160', '5392', '5685')]</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Other</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0.1          Seq ID  \\\n",
       "0             0  EIBEGOMO_01279   \n",
       "1             1  KBEGFPIJ_01435   \n",
       "2             2  OBHKLEMF_00864   \n",
       "3             3  EPJEOPIP_01092   \n",
       "4             4  AHBHJMAE_01497   \n",
       "5             5  PPJNGKMK_00620   \n",
       "6             6  AMIFCEHF_02229   \n",
       "7             7  HBHGIHPK_01033   \n",
       "8             8  FBOIKEJM_00490   \n",
       "9             9  BNAIOCPL_01158   \n",
       "\n",
       "                                     Seq Description  \\\n",
       "0  EIBEGOMO_01279 CRISPR-associated endonuclease ...   \n",
       "1  KBEGFPIJ_01435 CRISPR-associated endonuclease ...   \n",
       "2  OBHKLEMF_00864 CRISPR-associated endonuclease ...   \n",
       "3  EPJEOPIP_01092 CRISPR-associated endonuclease ...   \n",
       "4  AHBHJMAE_01497 CRISPR-associated endonuclease ...   \n",
       "5  PPJNGKMK_00620 CRISPR-associated endonuclease ...   \n",
       "6  AMIFCEHF_02229 CRISPR-associated endonuclease ...   \n",
       "7  HBHGIHPK_01033 CRISPR-associated endonuclease ...   \n",
       "8  FBOIKEJM_00490 CRISPR-associated endonuclease ...   \n",
       "9  BNAIOCPL_01158 CRISPR-associated endonuclease ...   \n",
       "\n",
       "                                                 Seq  \\\n",
       "0  MAEEIITDITIFGESKKMLRETLRKKFASCLDDEAIKKLSKLRYRD...   \n",
       "1  MKNIVGLDLGTNSIGWAVVNGSVNDDGSEQLVKIQASGSRIIPMDA...   \n",
       "2  MTKLNQPYGIGLDIGSNSIGFAVVDANSHLLRLKGETAIGARLFRE...   \n",
       "3  MGDRKYNLGLDIGTSSIGFAAVDENNQPIRVKGKTAIGVRLFEEGK...   \n",
       "4  MGYILGLDIGVASVGYAIIDENYNVLISGVRLFREGTAEENVARRG...   \n",
       "5  MGKTHIIGVGLDLGGTYTGTFITSHPSDEAEHRDHSSAFTVVNSEK...   \n",
       "6  MKTILGLDLGTTSIGWALVKEAEPFSDEKSEIVRLGVRVIPLTVDE...   \n",
       "7  MKSIYEDLKKEKMVISMAKELHFTSKDMPISDFDRMELKGRIETNP...   \n",
       "8  MILSTQKYIIDTKGCNMKYKLGLDLGSTSLGWAIVELNEKDTPIRL...   \n",
       "9  MDKKYYIGADIGTESVGWAVTDEDYNILKARGRELWGSYLFDKAES...   \n",
       "\n",
       "                              Contig                           Genome Name  \\\n",
       "0  NODE_1073_length_1994_cov_2.17225   AsnicarF_2017__MV_FEI4_t1Q14__bin.1   \n",
       "1   NODE_45_length_75974_cov_38.2495   AsnicarF_2017__MV_FEI4_t1Q14__bin.4   \n",
       "2   NODE_51_length_94666_cov_13.1517   AsnicarF_2017__MV_FEI4_t2Q15__bin.2   \n",
       "3   NODE_70_length_82951_cov_19.4828   AsnicarF_2017__MV_FEI4_t2Q15__bin.5   \n",
       "4  NODE_123_length_29049_cov_29.8501  AsnicarF_2017__MV_FEI5_t1Q14__bin.12   \n",
       "5  NODE_30_length_116148_cov_68.0265   AsnicarF_2017__MV_FEI5_t2Q14__bin.2   \n",
       "6  NODE_175_length_31817_cov_9.16545   AsnicarF_2017__MV_FEM1_t1Q14__bin.6   \n",
       "7  NODE_4986_length_3564_cov_6.73668   AsnicarF_2017__MV_FEM2_t1Q14__bin.2   \n",
       "8   NODE_9_length_125774_cov_10.3523  AsnicarF_2017__MV_FEM2_t1Q14__bin.30   \n",
       "9  NODE_2411_length_6235_cov_4.42071  AsnicarF_2017__MV_FEM2_t1Q14__bin.32   \n",
       "\n",
       "           Study    Sample Name  SGB ID                   pilercr_CRISPR  \\\n",
       "0  AsnicarF_2017  MV_FEI4_t1Q14    6939                              NaN   \n",
       "1  AsnicarF_2017  MV_FEI4_t1Q14    1814  [('ID=CRISPR1', '40172', '12')]   \n",
       "2  AsnicarF_2017  MV_FEI4_t2Q15    7144  [('ID=CRISPR1', '23272', '22')]   \n",
       "3  AsnicarF_2017  MV_FEI4_t2Q15    7168   [('ID=CRISPR1', '24497', '3')]   \n",
       "4  AsnicarF_2017  MV_FEI5_t1Q14    6939  [('ID=CRISPR1', '13059', '23')]   \n",
       "5  AsnicarF_2017  MV_FEI5_t2Q14    9262   [('ID=CRISPR1', '29853', '6')]   \n",
       "6  AsnicarF_2017  MV_FEM1_t1Q14    2230  [('ID=CRISPR2', '25256', '18')]   \n",
       "7  AsnicarF_2017  MV_FEM2_t1Q14    1024                              NaN   \n",
       "8  AsnicarF_2017  MV_FEM2_t1Q14   17322  [('ID=CRISPR1', '113006', '6')]   \n",
       "9  AsnicarF_2017  MV_FEM2_t1Q14   14099                              NaN   \n",
       "\n",
       "                                       minced_CRISPR  \\\n",
       "0                                                NaN   \n",
       "1                          [('.', '39283', '40172')]   \n",
       "2                          [('.', '21587', '23272')]   \n",
       "3                          [('.', '24197', '25552')]   \n",
       "4                          [('.', '11506', '13060')]   \n",
       "5  [('.', '29462', '29925'), ('.', '36897', '3764...   \n",
       "6                          [('.', '23902', '25256')]   \n",
       "7                                                NaN   \n",
       "8                        [('.', '112378', '113006')]   \n",
       "9                                                NaN   \n",
       "\n",
       "                                   prokka_cas1  \\\n",
       "0                                          NaN   \n",
       "1    [('ID=KBEGFPIJ_01437', '37910', '38842')]   \n",
       "2    [('ID=OBHKLEMF_00863', '24253', '25158')]   \n",
       "3    [('ID=EPJEOPIP_01091', '26532', '27437')]   \n",
       "4      [('ID=AHBHJMAE_01496', '7315', '8232')]   \n",
       "5    [('ID=PPJNGKMK_00621', '34746', '35738')]   \n",
       "6    [('ID=AMIFCEHF_02230', '22548', '23462')]   \n",
       "7                                          NaN   \n",
       "8  [('ID=FBOIKEJM_00492', '111321', '112217')]   \n",
       "9      [('ID=BNAIOCPL_01159', '4512', '5375')]   \n",
       "\n",
       "                                   prokka_cas2 uSGB  \\\n",
       "0                                          NaN   No   \n",
       "1    [('ID=KBEGFPIJ_01438', '38842', '39180')]   No   \n",
       "2    [('ID=OBHKLEMF_00862', '23970', '24275')]   No   \n",
       "3    [('ID=EPJEOPIP_01090', '26249', '26554')]   No   \n",
       "4      [('ID=AHBHJMAE_01495', '6974', '7318')]   No   \n",
       "5    [('ID=PPJNGKMK_00622', '35742', '36029')]   No   \n",
       "6    [('ID=AMIFCEHF_02231', '23495', '23800')]   No   \n",
       "7                                          NaN   No   \n",
       "8  [('ID=FBOIKEJM_00491', '110999', '111337')]  Yes   \n",
       "9      [('ID=BNAIOCPL_01160', '5392', '5685')]  Yes   \n",
       "\n",
       "  Level of estimated taxonomy  \\\n",
       "0                     Species   \n",
       "1                     Species   \n",
       "2                     Species   \n",
       "3                     Species   \n",
       "4                     Species   \n",
       "5                     Species   \n",
       "6                     Species   \n",
       "7                     Species   \n",
       "8                       Other   \n",
       "9                       Other   \n",
       "\n",
       "                                  Estimated taxonomy  \n",
       "0  k__Bacteria|p__Firmicutes|c__Negativicutes|o__...  \n",
       "1  k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...  \n",
       "2  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...  \n",
       "3  k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactob...  \n",
       "4  k__Bacteria|p__Firmicutes|c__Negativicutes|o__...  \n",
       "5  k__Bacteria|p__Proteobacteria|c__Betaproteobac...  \n",
       "6  k__Bacteria|p__Bacteroidetes|c__Bacteroidia|o_...  \n",
       "7  k__Bacteria|p__Firmicutes|c__Clostridia|o__Clo...  \n",
       "8                                                  -  \n",
       "9                                                  -  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cas9dataset.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1.3. Initial Cas9 lengths distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7f488140cbe0>"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA64AAAOICAYAAADW86xZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzs3X+MpVd93/HPN2xCKjayHZGsXGNpkOJKIVghYWWQ8s9saMGwf5hIRTJCxE6InD9MlUqWmiFSBA1xu1VDkJISVEe24jQ0WyshsuV1ilyXLUIqAZxSFuNGbGFL/EO2UjubbKBIjk7/mMfJYGZ27u7OnfnuM6+XtJqZc8/c51yfnR9vP/c+W2OMAAAAQFfftdcLAAAAgPMRrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaO7DXCzifV77ylWNlZWVXjvU3f/M3ecUrXrErx2K57OW82M95sZ/zYj/nw17Oi/2cl7nv56OPPvoXY4wf2G5e63BdWVnJ5z//+V051smTJ7O6urorx2K57OW82M95sZ/zYj/nw17Oi/2cl7nvZ1X9n0XmeaowAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrB/Z6AQAAcKFW1k4sNO/MsaNLXgmwG5xxBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFrbNlyr6nur6rNV9T+r6rGq+pfT+Kur6k+q6itV9Z+q6num8ZdPH5+ebl/ZcF/vm8b/rKresqwHBQAAwHwscsb1W0l+cozxo0lel+TGqnpjkn+T5MNjjOuSPJ/kPdP89yR5fozxQ0k+PM1LVb0myc1JfiTJjUl+q6petpMPBgAAgPnZNlzHunPTh989/RlJfjLJH0zj9yZ5+/T+TdPHmW5/U1XVNH58jPGtMcbXkpxOcsOOPAoAAABmq8YY209aPzP6aJIfSvKRJP82yWems6qpqmuT/PEY47VV9aUkN44xnphu+99J3pDkA9Pn/N40fvf0OX/wkmPdluS2JDl06NDrjx8/vhOPc1vnzp3LwYMHd+VYLJe9nBf7OS/2c17s53xcjnt56smzC827/porlrySfi7H/WRrc9/PI0eOPDrGOLzdvAOL3NkY42+TvK6qrkzyR0l+eLNp09va4ratxl96rLuS3JUkhw8fHqurq4ss8ZKdPHkyu3Uslstezov9nBf7OS/2cz4ux728de3EQvPOvGt1uQtp6HLcT7ZmP9dd0FWFxxh/meRkkjcmubKqXgzfVyV5anr/iSTXJsl0+xVJnts4vsnnAAAAwKYWuarwD0xnWlNV/yDJP07yeJJPJvmn07Rbktw/vf/A9HGm2//rWH8+8gNJbp6uOvzqJNcl+exOPRAAAADmaZGnCl+d5N7pda7fleS+McaDVfXlJMer6leT/I8kd0/z707yH6rqdNbPtN6cJGOMx6rqviRfTvJCktunpyADAADAlrYN1zHGF5P82CbjX80mVwUeY/y/JO/Y4r7uTHLnhS8TAACA/eqCXuMKAAAAu024AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWDuz1AgAA4EUrayf2eglAQ864AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANDatuFaVddW1Ser6vGqeqyqfmEa/0BVPVlVX5j+vG3D57yvqk5X1Z9V1Vs2jN84jZ2uqrXlPCQAAADm5MACc15IcscY40+r6vuSPFpVD0+3fXiM8WsbJ1fVa5LcnORHkvzDJP+lqv7RdPNHkvyTJE8k+VxVPTDG+PJOPBAAAADmadtwHWM8neTp6f2/rqrHk1xznk+5KcnxMca3knytqk4nuWG67fQY46tJUlXHp7nCFQAAgC1d0Gtcq2olyY8l+ZNp6L1V9cWquqeqrprGrkny5xs+7YlpbKtxAAAA2FKNMRabWHUwyX9LcucY4+NVdSjJXyQZST6Y5Ooxxs9W1UeS/Pcxxu9Nn3d3koeyHslvGWP83DT+7iQ3jDH+2UuOc1uS25Lk0KFDrz9+/PgOPMztnTt3LgcPHtyVY7Fc9nJe7Oe82M95sZ/z0WkvTz15dkfv7/prrtjR+7scdNpPLt3c9/PIkSOPjjEObzdvkde4pqq+O8kfJvnYGOPjSTLGeGbD7b+d5MHpwyeSXLvh01+V5Knp/a3G/84Y464kdyXJ4cOHx+rq6iJLvGQnT57Mbh2L5bKX82I/58V+zov9nI9Oe3nr2okdvb8z71rd0fu7HHTaTy6d/Vy3yFWFK8ndSR4fY/z6hvGrN0z7qSRfmt5/IMnNVfXyqnp1kuuSfDbJ55JcV1WvrqrvyfoFnB7YmYcBAADAXC1yxvUnkrw7yamq+sI09ktJ3llVr8v6U4XPJPn5JBljPFZV92X9oksvJLl9jPG3SVJV703yiSQvS3LPGOOxHXwsAAAAzNAiVxX+dJLa5KaHzvM5dya5c5Pxh873eQAAAPBSF3RVYQAAANhtwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKC1bcO1qq6tqk9W1eNV9VhV/cI0/v1V9XBVfWV6e9U0XlX1G1V1uqq+WFU/vuG+bpnmf6WqblnewwIAAGAuFjnj+kKSO8YYP5zkjUlur6rXJFlL8sgY47okj0wfJ8lbk1w3/bktyUeT9dBN8v4kb0hyQ5L3vxi7AAAAsJVtw3WM8fQY40+n9/86yeNJrklyU5J7p2n3Jnn79P5NSX53rPtMkiur6uokb0ny8BjjuTHG80keTnLjjj4aAAAAZqfGGItPrlpJ8qkkr03y9THGlRtue36McVVVPZjk2Bjj09P4I0l+Mclqku8dY/zqNP7LSb45xvi1lxzjtqyfqc2hQ4def/z48Yt+cBfi3LlzOXjw4K4ci+Wyl/NiP+fFfs6L/ZyPTnt56smzO3p/119zxY7e3+Wg035y6ea+n0eOHHl0jHF4u3kHFr3DqjqY5A+T/PMxxl9V1ZZTNxkb5xn/9oEx7kpyV5IcPnx4rK6uLrrES3Ly5Mns1rFYLns5L/ZzXuznvNjP+ei0l7eundjR+zvzrtUdvb/LQaf95NLZz3ULXVW4qr4769H6sTHGx6fhZ6anAGd6++w0/kSSazd8+quSPHWecQAAANjSIlcVriR3J3l8jPHrG256IMmLVwa+Jcn9G8Z/erq68BuTnB1jPJ3kE0neXFVXTRdlevM0BgAAAFta5KnCP5Hk3UlOVdUXprFfSnIsyX1V9Z4kX0/yjum2h5K8LcnpJN9I8jNJMsZ4rqo+mORz07xfGWM8tyOPAgAAgNnaNlyniyxt9YLWN20yfyS5fYv7uifJPReyQAAAAPa3hV7jCgAAAHtFuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABo7cBeLwAAYC+srJ1YaN6ZY0eXvBIAtuOMKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABo7cBeLwAAAC4XK2snFpp35tjRJa8E9hdnXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQ2oG9XgAAALBzTj15Nreundh23pljR3dhNbAznHEFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALS2bbhW1T1V9WxVfWnD2Aeq6smq+sL0520bbntfVZ2uqj+rqrdsGL9xGjtdVWs7/1AAAACYo0XOuP5Okhs3Gf/wGON105+HkqSqXpPk5iQ/Mn3Ob1XVy6rqZUk+kuStSV6T5J3TXAAAADivA9tNGGN8qqpWFry/m5IcH2N8K8nXqup0khum206PMb6aJFV1fJr75QteMQAAAPtKjTG2n7Qerg+OMV47ffyBJLcm+askn09yxxjj+ar6d0k+M8b4vWne3Un+eLqbG8cYPzeNvzvJG8YY793kWLcluS1JDh069Prjx49fwsNb3Llz53Lw4MFdORbLZS/nxX7Oi/2cl8t9P089eXaheddfc8WSV7L3Ou3lovuyqJ3ev8vh782zz53NM9/cft5++Ls9B52+PpfhyJEjj44xDm83b9szrlv4aJIPJhnT2w8l+dkktcnckc2fkrxpMY8x7kpyV5IcPnx4rK6uXuQSL8zJkyezW8diuezlvNjPebGf83K57+etaycWmnfmXavLXUgDnfZy0X1Z1E7v3+Xw9+Y3P3Z/PnRq+1/z98Pf7Tno9PW5ly4qXMcYz7z4flX9dpIHpw+fSHLthqmvSvLU9P5W4wAAALCli/rncKrq6g0f/lSSF684/ECSm6vq5VX16iTXJflsks8lua6qXl1V35P1Czg9cPHLBgAAYL/Y9oxrVf1+ktUkr6yqJ5K8P8lqVb0u60/3PZPk55NkjPFYVd2X9YsuvZDk9jHG3073894kn0jysiT3jDEe2/FHAwAAwOwsclXhd24yfPd55t+Z5M5Nxh9K8tAFrQ4AAIB976KeKgwAAAC7RbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGjtwF4vAABgJ62sndjrJQCww5xxBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWjuw1wsAAIBlWVk7sdC8M8eOLnklwKVwxhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWDuz1AgAA5mBl7cTCc88cO7rElQDMjzOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANDagb1eAADAIlbWTuz1EgDYI864AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0dmCvFwBzsrJ2YqF5Z44dXfJKAABgPpxxBQAAoDXhCgAAQGueKgwAcJnzUhVg7pxxBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtLZtuFbVPVX1bFV9acPY91fVw1X1lentVdN4VdVvVNXpqvpiVf34hs+5ZZr/laq6ZTkPBwAAgLlZ5Izr7yS58SVja0keGWNcl+SR6eMkeWuS66Y/tyX5aLIeuknen+QNSW5I8v4XYxcAAADOZ9twHWN8KslzLxm+Kcm90/v3Jnn7hvHfHes+k+TKqro6yVuSPDzGeG6M8XySh/OdMQwAAADf4WJf43pojPF0kkxvf3AavybJn2+Y98Q0ttU4AAAAnFeNMbafVLWS5MExxmunj/9yjHHlhtufH2NcVVUnkvzrMcanp/FHkvyLJD+Z5OVjjF+dxn85yTfGGB/a5Fi3Zf1pxjl06NDrjx8/fmmPcEHnzp3LwYMHd+VYLNde7uWpJ88uNO/6a65Y8krmw9fmvNjPednt/Vz0e+xOW/R79oWsb6d/Dlzqz59OX5tz2ee9/Fn/7HNn88w3t5/n95HLQ6evz2U4cuTIo2OMw9vNO3CR9/9MVV09xnh6eirws9P4E0mu3TDvVUmemsZXXzJ+crM7HmPcleSuJDl8+PBYXV3dbNqOO3nyZHbrWCzXXu7lrWsnFpp35l2ry13IjPjanBf7OS+7vZ+Lfo/daYt+z76Q9e30z4FL/fnT6WtzLvu8lz/rf/Nj9+dDp7b/Nd/vI5eHTl+fe+liw/WBJLckOTa9vX/D+Hur6njWL8R0dorbTyT5VxsuyPTmJO+7+GUDAOyOlT0KKQD+3rbhWlW/n/Wzpa+sqieyfnXgY0nuq6r3JPl6kndM0x9K8rYkp5N8I8nPJMkY47mq+mCSz03zfmWM8dILPgEAAMB32DZcxxjv3OKmN20ydyS5fYv7uSfJPRe0OgAAAPa9i72qMAAAAOwK4QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaO7DXCwAAgL22snZir5cAnIczrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaO7DXCwAAYHMrayf2egkALTjjCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoLUDe70AAACYm5W1EwvNO3Ps6JJXAvPgjCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGsH9noBwKVbWTux8Nwzx44ucSUAALDznHEFAACgNeEKAABAa8IVAACA1rzGlT236OszvTYTAAD2J+EKAAD7kJMHXE48VRgAAIDWhCsAAACtCVcAAABaE64AAAC0JlwBAABoTbgCAADQmnAFAACgNeEKAABAa8IVAACA1oQrAAAArQlXAAAAWrukcK2qM1V1qqq+UFWfn8a+v6oerqqvTG+vmsarqn6jqk5X1Rer6sd34gEAAAAwbztxxvXIGON1Y4zD08drSR4ZY1yX5JHp4yR5a5Lrpj+3JfnoDhwbAACAmVvGU4VvSnLv9P69Sd6+Yfx3x7rPJLmyqq5ewvEBAACYkRpjXPwnV30tyfNJRpJ/P8a4q6r+coxx5YY5z48xrqqqB5McG2N8ehp/JMkvjjE+/5L7vC3rZ2Rz6NCh1x8/fvyi13chzp07l4MHD+7Ksfh2p548u9C866+5YqF5e7mXO/1Ydvq4yzj2svnanBf7OS+7vZ8X8r2uu0W/F+/0Y97quJ2+Nue0z4tYxs/lZ587m2e+uXP3d7n97jA3nb4+l+Gra1vDAAAMkUlEQVTIkSOPbnj27pYOXOJxfmKM8VRV/WCSh6vqf51nbm0y9h3VPMa4K8ldSXL48OGxurp6iUtczMmTJ7Nbx+Lb3bp2YqF5Z961utC8vdzLnX4sO33cZRx72Xxtzov9nJfd3s8L+V7X3aLfi3f6MW913E5fm3Pa50Us4+fyb37s/nzo1KX+mv/3LrffHeam09fnXrqkpwqPMZ6a3j6b5I+S3JDkmRefAjy9fXaa/kSSazd8+quSPHUpxwcAAGD+Ljpcq+oVVfV9L76f5M1JvpTkgSS3TNNuSXL/9P4DSX56urrwG5OcHWM8fdErBwAAYF+4lOcQHEryR1X14v38xzHGf66qzyW5r6rek+TrSd4xzX8oyduSnE7yjSQ/cwnHBgAAYJ+46HAdY3w1yY9uMv5/k7xpk/GR5PaLPR4AAAD70869apsdsbLoxX2OHV3ySgAAAHpYxr/jCgAAADtGuAIAANCacAUAAKA1r3Hlgnkd7u5Z9L81AADMmTOuAAAAtCZcAQAAaM1ThfcBT+0FAAAuZ864AgAA0JozrpcpF+0BAAD2C2dcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWDuz1AgCA/W1l7cReLwGA5pxxBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWDuz1AgAAmLeVtRN7vYRZWPS/4x3XL3khsAeEK0vjhxQAALAThCt/R2gCAAAdCVcAYGEb/yfnHde/kFvP8z89zxw7uhtLAmAfcHEmAAAAWhOuAAAAtCZcAQAAaE24AgAA0JpwBQAAoDXhCgAAQGvCFQAAgNaEKwAAAK0JVwAAAFoTrgAAALQmXAEAAGhNuAIAANCacAUAAKA14QoAAEBrwhUAAIDWhCsAAACtHdjrBcBeWlk7sdC8M8eO7slxAQAAZ1wBAABoTrgCAADQmnAFAACgNeEKAABAay7OBACwy1ykD+DCOOMKAABAa8IVAACA1oQrAAAArQlXAAAAWnNxJgAAYEuLXkzszLGjS14J+5lwBTblhxQAAF14qjAAAACtCVcAAABaE64AAAC0JlwBAABozcWZYAGLXqgIAADYec64AgAA0JozrszOqSfP5lZnSIGZ809WAbCfOOMKAABAa864AuwxZ84AAM7PGVcAAABac8YV9hlXSL58OTMLAOxXzrgCAADQmjOuwK5wthAAgIslXAEA9omt/ifiHde/8G3/lJz/iQh0I1yBVl76S9VLf5l6kV+qAAD2D+HKZWPRp5recf2SF8JseTrz7vHfGgC4EMIVYJ+6kCtMC0gAYC8JVwBgKfzzWwDsFP8cDgAAAK054wpwgbw+EwBgdwlXYNb241MV9+NjBgDmzVOFAQAAaM0ZVwDa8rRsACARrsBlytNhYWf5mgKgM+EKsCRCYH9wVhgAlk+4ArBjtou4O65/IbfuYdDvx8j0P1AAmAPhClwSvxQzR/5eA0AvwhUAGhHNAPCdhCsA7AJBCmzG9wZYjH/HFQAAgNaEKwAAAK0JVwAAAFrzGlcAtuU1WADAXhKuAFz2hDUAzJtw3SV+qQIA5sbvN8Bu8RpXAAAAWhOuAAAAtOapwgAAfBtPAf7/7d1/qN11Hcfx54stpR+KE01kmpthwv7SFWaIIlS6WbgKlEmo9AMzNJIImgUl9o8aBQVRWI4U/BklDdJ0f0T9k+aPLLfUnGvlck3KSMFItt79cT7Xna1773bMfc/3nPt8wJfzPZ997z2fw2vv7/1+7vdzPldS3zhw/T95YpckSZKkg8upwpIkSZKkXnPgKkmSJEnqNQeukiRJkqRec+AqSZIkSeo1B66SJEmSpF5z4CpJkiRJ6jUHrpIkSZKkXnPgKkmSJEnqNQeukiRJkqRec+AqSZIkSeo1B66SJEmSpF5bPO4OSJIkSZp8y9b99ICO23bdBw5yTzSNvOMqSZIkSeq1zu+4JlkFfBNYBHy/qq7rug+SJEmSxuNA78yCd2e1R6d3XJMsAr4NrAZWABclWdFlHyRJkiRJk6XrqcKnAVuqamtVvQLcAazpuA+SJEmSpAnS9VThpcCzQ8+3A+/uuA+SJEmSJoALPmlGqqq7F0suAM6tqk+25xcDp1XVZ4aOuQy4rD09GXiqo+4dBfyto9fSwWWW08U8p4t5ThfznB5mOV3Mc7pMe54nVNXR+zuo6zuu24Hjh54fBzw3fEBV3Qjc2GWnAJI8XFXv6vp19fozy+lintPFPKeLeU4Ps5wu5jldzHOg68+4PgSclGR5kkOAtcCGjvsgSZIkSZognd5xrapdSa4E7mPw53DWV9XmLvsgSZIkSZosnf8d16q6B7in69c9AJ1PT9ZBY5bTxTyni3lOF/OcHmY5XcxzupgnHS/OJEmSJEnSqLr+jKskSZIkSSNZ8APXJKuSPJVkS5J14+6P9i/J8Ul+nuSJJJuTfLa1X5PkL0kea9t5Q19zdcv4qSTnjq/3mk2SbUkeb7k93NqOTLIxydPtcUlrT5JvtTx/l2TleHuvGUlOHqq/x5K8mOQqa3NyJFmf5Pkkm4baRq7FJJe2459Ocuk43ovmzPNrSZ5smd2d5IjWvizJv4bq9LtDX/POdo7e0jLPON7PQjdHniOfX732Hb85srxzKMdtSR5r7dbmjKpasBuDBaKeAU4EDgF+C6wYd7/c9pvbscDKtn8Y8AdgBXAN8PlZjl/Rsj0UWN4yXzTu9+G2V0bbgKP2absBWNf21wHXt/3zgHuBAKcDD467/26zZroI+CtwgrU5ORtwFrAS2DTUNlItAkcCW9vjkra/ZNzvbSFuc+R5DrC47V8/lOey4eP2+T6/Bt7Tsr4XWD3u97YQtznyHOn86rVvP7bZstzn378OfLntW5ttW+h3XE8DtlTV1qp6BbgDWDPmPmk/qmpHVT3a9l8CngCWzvMla4A7qurfVfVHYAuD7NVva4Cb2/7NwIeG2m+pgQeAI5IcO44Oal7vBZ6pqj/Nc4y12TNV9UvghX2aR63Fc4GNVfVCVf0D2AisOvi9175my7Oq7q+qXe3pA8Bx832PlunhVfWrGlwp38Ke/wPq0Bz1OZe5zq9e+/bAfFm2u6YXArfP9z0WYm0u9IHrUuDZoefbmX8ApJ5Jsgw4FXiwNV3Zpj+tn5nOhjlPggLuT/JIksta2zFVtQMGv6wA3trazXMyrGXvH7rW5uQatRbNdXJ8nMFdmhnLk/wmyS+SnNnaljLIcIZ59s8o51frs//OBHZW1dNDbdYmDlxnmwfuMssTIslbgB8BV1XVi8B3gLcDpwA7GEyzAHOeBGdU1UpgNXBFkrPmOdY8ey7JIcD5wA9bk7U5nebKz1wnQJIvAbuAW1vTDuBtVXUq8DngtiSHY559N+r51Tz77yL2/sWvtdks9IHrduD4oefHAc+NqS8aQZI3MBi03lpVPwaoqp1Vtbuq/gN8jz1TDs2556rqufb4PHA3g+x2zkwBbo/Pt8PNs/9WA49W1U6wNqfAqLVorj3XFsz6IPDRNsWQNqX0723/EQafg3wHgzyHpxObZ4+8hvOr9dljSRYDHwHunGmzNvdY6APXh4CTkixvdwjWAhvG3CftR5v7fxPwRFV9Y6h9+HOOHwZmVmrbAKxNcmiS5cBJDD7Mrh5I8uYkh83sM1g4ZBOD3GZWI70U+Enb3wBc0lY0PR3458w0RvXGXr8ttjYn3qi1eB9wTpIlbdriOa1NPZBkFfAF4Pyqenmo/egki9r+iQzqcWvL9KUkp7efv5ew5/+Axuw1nF+99u239wFPVtWrU4CtzT0Wj7sD41RVu5JcyeAH6iJgfVVtHnO3tH9nABcDj88sFQ58EbgoySkMpklsAz4FUFWbk9wF/J7BtKgrqmp3573WXI4B7m4ruC8GbquqnyV5CLgrySeAPwMXtOPvYbCa6RbgZeBj3XdZc0nyJuD9tPprbrA2J0OS24GzgaOSbAe+AlzHCLVYVS8k+SqDC2SAa6vqQBeU0etojjyvZrDS7MZ23n2gqi5nsMrptUl2AbuBy4dy+zTwA+CNDD4TO/y5WHVkjjzPHvX86rXv+M2WZVXdxP+uDwHW5qvSZohIkiRJktRLC32qsCRJkiSp5xy4SpIkSZJ6zYGrJEmSJKnXHLhKkiRJknrNgaskSZIkqdccuEqSJEmSes2BqyRJkiSp1xy4SpIkSZJ67b9dgufzcheH6QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1152x1152 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "counts=cas9dataset[\"Seq\"].str.count(\"\")\n",
    "counts.hist(bins=80, figsize=(20*0.8,20*0.8))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1784"
      ]
     },
     "execution_count": 190,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.max()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Data filtering: Extract a list of the shortest working cas9 from most abundant and most unknown genomes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.1 Filter by: active locus\n",
    "An active locus is defined by having at least 1 CRISPR array, 1 Cas1, 1 Cas2, and 1 effector Cas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Checking that the next step actually does work properly:\n",
    "# If there are no NaNs values in columns [\"Estimated taxonomy\",\"Level of estimated taxonomy\", \"SGB ID\",\n",
    "#                                            \"Sample Name\", \"Study\", \"Genome Name\", \"Contig\", \"Seq\", \n",
    "#                                            \"Seq Description\", \"Seq ID\"],\n",
    "# then, it returns True\n",
    "temp=cas9dataset[cas9dataset.isna().any(axis=1)]\n",
    "temp[temp[\"Estimated taxonomy\"]==None].shape==temp[temp[\"Level of estimated taxonomy\"]==None].shape==\\\n",
    "temp[temp[\"uSGB\"]==None].shape==temp[temp[\"SGB ID\"]==None].shape==temp[temp[\"Sample Name\"]==None].shape==\\\n",
    "temp[temp[\"Study\"]==None].shape==temp[temp[\"Genome Name\"]==None].shape==temp[temp[\"Contig\"]==None].shape==\\\n",
    "temp[temp[\"Seq\"]==None].shape==temp[temp[\"Seq Description\"]==None].shape==temp[temp[\"Seq ID\"]==None].shape==\\\n",
    "(0,16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(18875, 16)\n"
     ]
    }
   ],
   "source": [
    "#Drop Nan rows\n",
    "activecas9s=cas9dataset.dropna(how=\"any\")\n",
    "print(activecas9s.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 18875 active loci. Let us print the lengths distribution of cas9 in these loci:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Filter by: working active locus\n",
    "A working locus is defined by being an active locus with a Cas9 of a length falling within the peak of the distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA64AAAOICAYAAADW86xZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvnQurowAAIABJREFUeJzs3W9srmdh3/Hf1biwrn9EUALKkmhmUzQ1mzTKopSp0kSHBIHzIlQaErxYs4opUwcb3VZN7l4M1qrSaaetElJHla0RQVpLUVfUaCcbZKgbmjRaDl0WSCnLKT3AaTKSLi1txdSK9toL36eYxPbxn8f279ifj2TZvnz7ua7Lfh7bX9+PH485ZwAAAKDVN5z0AgAAAGA3whUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqaye9gN3cdNNNc319/aSXAQAAwBH45Cc/+dtzzpuvdVx1uK6vr+fixYsnvQwAAACOwBjj83s5zl2FAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqLZ20gsAAACubX3jwp6Ou3z+3BGvBI6fM64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUu2a4jjFuH2P80hjjM2OMJ8YY71zGXzrGeHSM8eTy/MZlfIwx3jPGuDTGeHyM8aotl3XfcvyTY4z7jm5bAAAAnBZ7OeP61ST/ZM757UleneTtY4w7k2wk+eic844kH11eT5I3JLljebo/yXuTzdBN8q4k35nk7iTvuhq7AAAAsJNrhuuc8+k5568uL/9+ks8kuTXJvUkeWg57KMmblpfvTfL+uenjSV4yxrglyeuTPDrnfG7O+TtJHk1yz0p3AwAAwKmzr79xHWOsJ/mOJL+c5OVzzqeTzbhN8rLlsFuTfHHLu11ZxnYaf/4c948xLo4xLj777LP7WR4AAACn0J7DdYzxLUn+Q5IfmHP+3m6HbjM2dxn/+oE5H5hz3jXnvOvmm2/e6/IAAAA4pfYUrmOMb8xmtP77OecvLMNfWu4CnOX5M8v4lSS3b3n325I8tcs4AAAA7Ggvjyo8kvx0ks/MOf/1ljc9nOTqIwPfl+QXt4x/7/Lowq9O8uXlrsQfTvK6McaNy4MyvW4ZAwAAgB2t7eGY70ryt5N8aozx2DL2z5KcT/LBMcbbknwhyZuXtz2S5I1JLiX5SpLvS5I553NjjB9J8onluB+ecz63kl0AAABwal0zXOec/z3b/31qkrx2m+NnkrfvcFkPJnlwPwsEAADgbNvXowoDAADAcROuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVBOuAAAAVFs76QUAAADXv/WNC3s67vL5c0e8Ek4jZ1wBAACoJlwBAACoJlwBAACoJlwBAACoJlwBAACoJlwBAACoJlwBAACoJlwBAACoJlwBAACoJlwBAACotnbSCwAAgLNqfePCSS8BrgvOuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFDtmuE6xnhwjPHMGOPTW8bePcb4rTHGY8vTG7e87YfGGJfGGJ8dY7x+y/g9y9ilMcbG6rcCAADAabSXM67vS3LPNuM/Med85fL0SJKMMe5M8pYkf3l5n38zxrhhjHFDkp9M8oYkdyZ563IsAAAA7GrtWgfMOT82xljf4+Xdm+QDc84/TPKbY4xLSe5e3nZpzvm5JBljfGA59tf2vWIAAADOlMP8jes7xhiPL3clvnEZuzXJF7ccc2UZ22kcAAAAdnXQcH1vkr+Y5JVJnk7yr5bxsc2xc5fxFxhj3D/GuDjGuPjss88ecHkAAACcFgcK1znnl+acfzzn/JMk/zZfuzvwlSS3bzn0tiRP7TK+3WU/MOe8a855180333yQ5QEAAHCKHChcxxi3bHn1e5JcfcThh5O8ZYzx4jHGK5LckeRXknwiyR1jjFeMMV6UzQdwevjgywYAAOCsuOaDM40xfjbJa5LcNMa4kuRdSV4zxnhlNu/ueznJ30uSOecTY4wPZvNBl76a5O1zzj9eLucdST6c5IYkD845n1j5bgAAADh19vKowm/dZvindzn+R5P86DbjjyR5ZF+rAwAA4Mw7zKMKAwAAwJETrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFRbO+kFAABwOOsbF/Z03OXz5454JQBHwxlXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqglXAAAAqq2d9AIAAACOw/rGhT0dd/n8uSNeCfvljCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADVhCsAAADV1k56AQAAwOqsb1zY03GXz5874pXA6jjjCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQLVrhusY48ExxjNjjE9vGXvpGOPRMcaTy/Mbl/ExxnjPGOPSGOPxMcartrzPfcvxT44x7jua7QAAAHDa7OWM6/uS3PO8sY0kH51z3pHko8vrSfKGJHcsT/cneW+yGbpJ3pXkO5PcneRdV2MXAAAAdnPNcJ1zfizJc88bvjfJQ8vLDyV505bx989NH0/ykjHGLUlen+TROedzc87fSfJoXhjDAAAA8AIH/RvXl885n06S5fnLlvFbk3xxy3FXlrGdxl9gjHH/GOPiGOPis88+e8DlAQAAcFqs+sGZxjZjc5fxFw7O+cCc8645510333zzShcHAADA9eeg4fql5S7AWZ4/s4xfSXL7luNuS/LULuMAAACwq4OG68NJrj4y8H1JfnHL+Pcujy786iRfXu5K/OEkrxtj3Lg8KNPrljEAAADY1dq1Dhhj/GyS1yS5aYxxJZuPDnw+yQfHGG9L8oUkb14OfyTJG5NcSvKVJN+XJHPO58YYP5LkE8txPzznfP4DPgEAAMALXDNc55xv3eFNr93m2Jnk7TtczoNJHtzX6gAAADjzVv3gTAAAALBSwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqwhUAAIBqaye9AAAA4Pitb1zY03GXz5874pXAtTnjCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDXhCgAAQDX/DgcA4Jj5NyQA++OMKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANXWTnoBAABsb33jwkkvAaCCM64AAABUE64AAABUE64AAABUE64AAABUE64AAABU86jCAAB8nb0+mvHl8+eOeCUAm5xxBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoJpwBQAAoNraSS8AAACuWt+4sKfjLp8/d8QrAZo44woAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEA14QoAAEC1Q4XrGOPyGONTY4zHxhgXl7GXjjEeHWM8uTy/cRkfY4z3jDEujTEeH2O8ahUbAAAA4HRbxRnX755zvnLOedfy+kaSj84570jy0eX1JHlDkjuWp/uTvHcFcwMAAHDKHcVdhe9N8tDy8kNJ3rRl/P1z08eTvGSMccsRzA8AAMApcthwnUk+Msb45Bjj/mXs5XPOp5Nkef6yZfzWJF/c8r5XlrGvM8a4f4xxcYxx8dlnnz3k8gAAALjerR3y/b9rzvnUGONlSR4dY/z6LseObcbmCwbmfCDJA0ly1113veDtAAAAnC2HCtc551PL82fGGB9KcneSL40xbplzPr3cFfiZ5fArSW7f8u63JXnqMPMDALB36xsXTnoJAAdy4LsKjzG+eYzxrVdfTvK6JJ9O8nCS+5bD7kvyi8vLDyf53uXRhV+d5MtX71IMAAAAOznMGdeXJ/nQGOPq5fzMnPM/jzE+keSDY4y3JflCkjcvxz+S5I1JLiX5SpLvO8TcAAAAnBEHDtc55+eS/NVtxv9vktduMz6TvP2g8wEAAHA2HcW/wwEAAICVEa4AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUE64AAABUWzvpBQAAAFyP1jcu7Om4y+fPHfFKTj9nXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKgmXAEAAKi2dtILAAAAeq1vXDjpJYAzrgAAAHQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFQTrgAAAFRbO+kFAAAAZ8f6xoU9H3v5/LkjXAnXE2dcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqCZcAQAAqOZRhQEAuO7s9ZFpPSotnA7OuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBNuAIAAFBt7aQXAABwGqxvXDjpJQCcWs64AgAAUE24AgAAUE24AgAAUE24AgAAUE24AgAAUE24AgAAUE24AgAAUE24AgAAUE24AgAAUE24AgAAUG3tpBcAAMD1aX3jwp6Ou3z+3BGvBDjtnHEFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACgmnAFAACg2tpJLwAAAKDJ+saFk14Cz+OMKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANWEKwAAANXWTnoBAAAnYX3jwp6Ou3z+3BGvBIBrccYVAACAasIVAACAasIVAACAasIVAACAasIVAACAasIVAACAasIVAACAasIVAACAasIVAACAasIVAACAasIVAACAasIVAACAasIVAACAamsnvQAAAE639Y0LJ70E4DonXAEAYMXEOqyWcAUAAK5rflFw+vkbVwAAAKoJVwAAAKoJVwAAAKoJVwAAAKoJVwAAAKoJVwAAAKr5dzgAwKni32IAnD7OuAIAAFDNGVcA4Ejs9czn5fPnjnglnGWuh3A6OOMKAABANWdc4RTYz99z+Y0yAADXG+EKAM/jl0Fs5cGeAE6euwoDAABQzRlXKOa3/AAA4IwrAAAA5Zxx5dTxt2kAAHC6OOMKAABANWdc4QSc5N+urnpuZ63Phr1eb1wfOAjXLwCuRbgCnFFn8cG/zuKeAeA0EK6wB84GcJRO0/XrNO3lrPG5A6CZcAWOhR+K2cr1AQDYD+EKsBBTANDFn3hwlXAFDqX9G0r7+oC9c3sGOLuEK3Bduh5+gL0e1nha+FgDwOkmXAHgEE7qLubu2g5w/fA1+/CE63VqP2cXTuoG4AbKaeXsHkdp1dev9ssDgL0QrvypsxiafgDjenIWr69ncc9AN1+X4GQIV1gh38wA4Prkezh0E67s20l9YfcN5Wzweea0ct0GgIMTrmWO4gcbPywBAADXM+HKiRPWAADAbr7hpBcAAAAAuxGuAAAAVHNX4WPi7rAAAAAHI1wPSZACAADHba8dcvn8uSNeyfE49rsKjzHuGWN8doxxaYyxcdzzAwAAcH051nAdY9yQ5CeTvCHJnUneOsa48zjXAAAAwPXluO8qfHeSS3POzyXJGOMDSe5N8mvHvA4AAIAq/gxxZ8d9V+Fbk3xxy+tXljEAAADY1nGfcR3bjM2vO2CM+5Pcv7z6B2OMzx7xmm5K8ttHPMdxzXOa9nJc89jL2Z7nNO3luOaxl7M9z2nay3HNYy9ne57TtJfjmsdeVjzP+LHjmecQ/vxeDjrucL2S5PYtr9+W5KmtB8w5H0jywHEtaIxxcc5512mY5zTt5bjmsZezPc9p2stxzWMvZ3ue07SX45rHXs72PKdpL8c1j72YZyfHfVfhTyS5Y4zxijHGi5K8JcnDx7wGAAAAriPHesZ1zvnVMcY7knw4yQ1JHpxzPnGcawAAAOD6ctx3Fc6c85Ekjxz3vLs4rrslH8c8p2kvxzWPvZzteU7TXo5rHns52/Ocpr0c1zz2crbnOU17Oa557MU82xpzzmsfBQAAACfkuP/GFQAAAPZnznlmn5Lck+SzSS4l2Vjh5V5O8qkkjyW5uIy9NMmjSZ5cnt94gMt9MMkzST69ZWzby83mvx56z7K3x5O86pDzvDvJby17eizJG7e87YeWeT6b5PV7nOP2JL+U5DNJnkjyzlXvZ5c5Vr2XP5PkV5L8r2Wef7GMvyLJLy97+bkkL1rGX7y8fml5+/oh53lfkt/csp9XruA6cEOS/5nkPx7FXnaZ5yj2cjl7vD0edJ4d5ljp9Wx5v5ck+fkkv75cr//6EexluzlWupckf2nLZT2W5PeS/MAR7GWneVa9n3+Uzdvkp5P8bDZvqyu/zewwz/uywttMkncul/9Ekh84itvLLvMc+vOSFX2fTHLfcvyTSe6jvWZnAAALoklEQVQ75DyvSfLlLfv651veZ8efQ3aY483Lx+xPktz1vOO3/RjtNsd+50mynuT/bdnLT21521/L5tfBS8vHdexhnn+Zza81jyf5UJKXHGY/+5njCPbyI8scjyX5SJI/d5jr2T7nONB1bKd5trztB7P5bytvOorbzC7zHGg/O3zM3p19fl05yMdsp3kOej3b6eOV5B8sa3siyY8fxV52muege1nl08ov8Hp5yuYPzL+R5C8keVE2Y+DOFV325as3vi1jP371CpNkI8mPHeBy/0aSVz3vhrLt5SZ5Y5L/lM0vMq9O8suHnOfdSX5wm2PvXD52L87mD2q/keSGPcxxS5YveEm+Ncn/Xi5rZfvZZY5V72Uk+Zbl5W/M5g+jr07ywSRvWcZ/Ksn3Ly///as39mw+svbP7fHzstM870vyt7Y5/jDXgX+c5GfytaBc6V52meco9nI5e7w9HnSeHeZY6fVsed+Hkvzd5eUXZTMyV72X7eZY+V62XMYNSf5PNv+H28q/nu0wz8r2k+TWbIbjN225rfydVd9mdpnnfVnRbSbJX8lmTP7ZbD4Gxn9JcscRXMd2mufQn5es4PtkNgP0c8vzG5eXbzzEPK/J8jVum+vkjj+H7DDHt2fzFzL/NV8flNt+jK41xwHmWc824bG87Vey+YuusXxc37CHeV6XZG15+ce2fMwOtJ99zrHqvXzblpf/Yb52Oz/Q9WyfcxzoOrbTPMv47dl8MNXP52tBudLbzC7zHGg/O3zM3p19fF056Mdsl3kOdD3bYY7vzubXyxcvr7/siPay0zwH2ssqn87yXYXvTnJpzvm5OecfJflAknuPcL57s/kDYZbnb9rvBcw5P5bkuT1e7r1J3j83fTzJS8YYtxxinp3cm+QDc84/nHP+ZjZ/03L3HuZ4es75q8vLv5/NMzu3rnI/u8yx6r3MOecfLK9+4/I0k/zNbJ652m4vV/f480leO8YYh5hnt/3s+zowxrgtybkk/255fax6L9vNcw0Hvj7vcnkrvd3sc+59X8/GGN+WzW8wP50kc84/mnP+bla4l13mWOlenue1SX5jzvn5Ve7lGvPs5KD7WUvyTWOMtWzG2NM5gtvMNvM8tcuxB/mYfXuSj885vzLn/GqS/5bke7L6z8tO8+y2lz19Xlb0ffL1SR6dcz435/ydbJ49vecQ8+xk159DtptjzvmZOednt7msnT5G1/xZZ5/zbGv5uH3bnPN/zDlnkvfnefvfYZ6PLNeBJPl4ktsOs599zrHqvfzelle/OV/7/nyg69k+59jJgT7/i59I8k+fN8eqbzM7zXOg/ewyx3ZWepvZr2tdz3aY4/uTnJ9z/uFyzDNHtJed5jnQXlbpLIfrrUm+uOX1K9k9avZjJvnIGOOTY4z7l7GXzzmfTjaDKsnLVjTXTpd7FPt7xxjj8THGg2OMG1c1zxhjPcl3ZPMM4pHs53lzJCveyxjjhjHGY9m8u8Wj2fwN1+9u+Wa59bL+dJ7l7V/O/2/v7GLsmqI4/ltBSxEfRXy0aEVJCCU09dFGSkQTRoqHShsNHnmSiMgkEg/efCWIB8QDKlFtigdf0RBPSoepifFRH9Epij6okEixPKx9zZnp2adz9l1HRrp+yWTuPffe/b//c9baZ++z99kXZpfoqGrPz33Jz0MiMrNPPw9jJ5G/0/PZXXip0enh6QXa5WOpTp0G+MbZfOAn4GkR+VBEnhSRQ5295DS8vVRZiU17xdlLkw44+VHVHcD9wLdYh/UXYAvOOVOno6pvpJe9cmYEWCois0VkFja6Mhf/45LTgW7irO33L9VrOs9fJCLDIvKqiJy1D/0SvL00MS/VD++IyJKK/lifOrdgIzW98rrwU9UAZy8icp+IbAdWAfdUynPzktEAxxgTkQFgh6oOT3rJ20tOB3xzpk290k+M1emAX5wtAJaIyHuprAs78pLT8fRSxP7cca27wj2Vqz1T4RJVPR9YDtwmIkudym2Dt7/HgdOAhVij6QEPHRE5DFiP3eO0u+mtpTo1Gu5eVPUvVV2IXcldhI0o5Mpy0xGRs7H7Gs4ELsSm6dxVqiMiVwM/quqW6uaGcoq8ZHTA0UuFNvlYqlOn4R1nB2LTeR5X1fOA37DpiDlKdHIaXeX/DGAAWLevtzrruPlJjZNrselZJ2KjIMsbyinNmb10RGQ1jjmjqqPYVMo3gdewaWV/NnykyEuDTidx1kCuXG+9IeAUVT0XeATYuA/9Ev4rL98DJ6f64Q5grdhMjX5zdBCLged6mzLl9dMWmKzh7kVVB1V1btK4vSedKa80f+o03GIsXUwaZGKn+N+XM+V563jmTNt6pfT453Q84+xAbCr2YuBO4AURkYay+mlv1Ol0kv9t2J87rmOMX+EF6wg0TbuaMqr6Xfr/I7YQwCJgZxpK7w2pNw67tyBXrqs/Vd2ZOk1/A08wPk2rWEdEDsI6lM+p6oa02dVPnUYXXnqoTat8G0v2I8Wm9E0u61+d9PoRtJxyUtG5Sm1KtKYpHU/36ecSYEBEvsGmlCzDRka9veylIyLPOnsBWudjkU6dRgdxNgaMVUbZX8Q6mZ5eajU6zJnlwJCq7kzPu6rPJug4+7kC+FpVf1LVPcAG4GL8c6ZWxztnVPUpVT1fVZem7/UFHRyXOp0O46zt9y/Vq9VR1d2abvNQ+y37g0TkGAdfVby91KI2FXFXerwFm120IOlUp+C2aQusAa4GVqlqr7Hr6qdOowsvFdYC13fhpU7DOcZOwy6QDadz9BxgSESOd/aS1fH0U1CvlNadtTrOcTYGbEh1/mZsxlrTfumnvbGXTsc5MyX2547r+8DpIjIvXY1fCbzcb6EicqiIHN57jC0KMJLKXpPetgZ4qV+tRK7cl4GbxFiMTSv7vlREJt63tALz1NNZKSIzRWQetsDG5imUJ9h9dKOq+mAXfnIaHXg5VkSOTI8PwRqYo9iKxjdkvPQ83gBsqpys2+p8WmkoCXZPQdVPq32mqner6hxVPRXLiU2qusrbS0ZntaeXVE7bfCyJs1oN7zhT1R+A7SJyRtp0OfCJp5echreXCjcycfpuV/XZBB1nP98Ci0VkVorb3nFxzZmMzmgHOXNc+n8ycB2239yPS51Oh3HW9vu/DlwpIkeJjXRfmbYV6YjI8en4ICKLsLbXLnzbIbl95NrWSeehA9Lj+Unnq7TffhWRxcnrTUyhnSMiV2GzBAZU9fcu/OQ0OvByeuXpALaScc+LS5zlNDxjTFU/VtXjVPXUdI4ewy5g/uDppUnH009BvVKUMzkd5zjbiA0oICILsAWXfvb2ktPxzpkitIMVn/4vf9h9NZ9jVwwGncqcj0176v1kyWDaPht4C7t6/RZwdEHZz2PD9HuwBL81Vy42bP9Y8vYxk5bNL9B5JpWzFQv6EyrvH0w6nzHFVcSAS7FpBL1l3T9Kx8PNT4OGt5dzsJ902YpVVPdUYmEzdpP8OsZXZzs4Pd+WXp/fp86m5GcEeJbxlYeLYyB9/jLGV/t19dKg4+qFlvlYGGc5Ddc4S59bCHyQytyITeVxrQMyGl14mYU1RI6obOuiPqvT8a4D7sUakCOp7Jl0kDMZHe+ceRfreA8Dl3d4XOp0+j4uOJ0nsXsht6W/m/vUuR2rG4axxYEurpSTbYdkNFakx38AO4HX97WPmjTa6mAjfD0vQ8A1lXIuSHH4JfAo7PUTMnU627B78Op+XqO1nzYaHXhZnz6zFXgFOKmfOGupURRjOZ1Jr3/DxJ/DccuZBh3PnGldr5Tss5wOhXGW0ZiB1fMjqaxlHXmp1Sn14vknSSwIgiAIgiAIgiAIpiX781ThIAiCIAiCIAiC4H9AdFyDIAiCIAiCIAiCaU10XIMgCIIgCIIgCIJpTXRcgyAIgiAIgiAIgmlNdFyDIAiCIAiCIAiCaU10XIMgCIIgCIIgCIJpTXRcgyAIgiAIgiAIgmlNdFyDIAiCIAiCIAiCac0/VCXzhETtn9IAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1152x1152 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "activecas9counts=activecas9s[\"Seq\"].str.count(\"\")\n",
    "plt.figure(figsize=(20*0.8,20*0.8))\n",
    "plt.hist(activecas9counts ,bins=80)\n",
    "plt.xticks(np.arange(0, 1700, step=50))\n",
    "print()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The length intreval where there seems to be a working protein is around (1050,1170) , (1330,1450) and maybe (1500,1580). We shall define these intrevals as working lengths."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3-2.4 Filter by: sequence length of 949-1099 amino acid, unknown SGB\n",
    "An unknown SGB is an SGB with no reference genome in literature."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(628, 16)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_activecas9counts=activecas9s[\"Seq\"].str.count(\"\").sort_values()\n",
    "working_intreval=sorted_activecas9counts[sorted_activecas9counts>=949]\n",
    "working_intreval=working_intreval[working_intreval<=1099]\n",
    "active_working_cas9s=activecas9s.loc[working_intreval.index]\n",
    "active_working_unknwon_cas9s=active_working_cas9s[active_working_cas9s.uSGB==\"Yes\"]\n",
    "active_working_unknwon_cas9s.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 628 active and working Cas9s from unknown species."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test=active_working_unknwon_cas9s.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NKGLJMBJ_00036 type II-A, B or C\n",
      "KJDMGLGH_01090 type II-C variant\n",
      "JJNJAONG_00079 type II-A, B or C\n",
      "PNKCOBEG_00068 type II-C variant\n",
      "GCEOGADH_01212 type II-C variant\n",
      "MOJKMKJH_00496 type II-C variant\n",
      "HFHGEGGB_01281 type II-C variant\n",
      "OJFAPCEG_01446 type II-A, B or C\n",
      "POLOOKHD_01143 type II-A, B or C\n",
      "FKPPKKCJ_01237 type II-C variant\n"
     ]
    }
   ],
   "source": [
    "for i, cas9 in test.iterrows():\n",
    "    cas1start=cas9[\"prokka_cas1\"].split(\",\")[1].strip(\" '\")\n",
    "    cas2start=cas9[\"prokka_cas2\"].split(\",\")[1].strip(\" '\")\n",
    "    if int(cas1start)<int(cas2start):\n",
    "        print(cas9[\"Seq ID\"], \"type II-A, B or C\")\n",
    "    else:\n",
    "        print(cas9[\"Seq ID\"], \"type II-C variant\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "/test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.5 Filter by: most abundant SGB "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let us check which SGBs are the most abundant among the Cas loci we have filtered so far."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SGB ID</th>\n",
       "      <th># Genomes with act. wrk. cas9</th>\n",
       "      <th># Reconstructed genomes</th>\n",
       "      <th>Genomes relative abundance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9340</td>\n",
       "      <td>105</td>\n",
       "      <td>242</td>\n",
       "      <td>0.433884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15299</td>\n",
       "      <td>51</td>\n",
       "      <td>99</td>\n",
       "      <td>0.515152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8767</td>\n",
       "      <td>34</td>\n",
       "      <td>71</td>\n",
       "      <td>0.478873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15095</td>\n",
       "      <td>22</td>\n",
       "      <td>70</td>\n",
       "      <td>0.314286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8769</td>\n",
       "      <td>21</td>\n",
       "      <td>18</td>\n",
       "      <td>1.166667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9710</td>\n",
       "      <td>17</td>\n",
       "      <td>28</td>\n",
       "      <td>0.607143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9281</td>\n",
       "      <td>17</td>\n",
       "      <td>146</td>\n",
       "      <td>0.116438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9311</td>\n",
       "      <td>16</td>\n",
       "      <td>22</td>\n",
       "      <td>0.727273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>14454</td>\n",
       "      <td>11</td>\n",
       "      <td>83</td>\n",
       "      <td>0.132530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9280</td>\n",
       "      <td>10</td>\n",
       "      <td>145</td>\n",
       "      <td>0.068966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>15286</td>\n",
       "      <td>10</td>\n",
       "      <td>1813</td>\n",
       "      <td>0.005516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>8774</td>\n",
       "      <td>10</td>\n",
       "      <td>25</td>\n",
       "      <td>0.400000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>14370</td>\n",
       "      <td>9</td>\n",
       "      <td>44</td>\n",
       "      <td>0.204545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>9307</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>0.321429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>4329</td>\n",
       "      <td>9</td>\n",
       "      <td>34</td>\n",
       "      <td>0.264706</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    SGB ID  # Genomes with act. wrk. cas9  # Reconstructed genomes  \\\n",
       "0     9340                            105                      242   \n",
       "1    15299                             51                       99   \n",
       "2     8767                             34                       71   \n",
       "3    15095                             22                       70   \n",
       "4     8769                             21                       18   \n",
       "5     9710                             17                       28   \n",
       "6     9281                             17                      146   \n",
       "7     9311                             16                       22   \n",
       "8    14454                             11                       83   \n",
       "9     9280                             10                      145   \n",
       "10   15286                             10                     1813   \n",
       "11    8774                             10                       25   \n",
       "12   14370                              9                       44   \n",
       "13    9307                              9                       28   \n",
       "14    4329                              9                       34   \n",
       "\n",
       "    Genomes relative abundance  \n",
       "0                     0.433884  \n",
       "1                     0.515152  \n",
       "2                     0.478873  \n",
       "3                     0.314286  \n",
       "4                     1.166667  \n",
       "5                     0.607143  \n",
       "6                     0.116438  \n",
       "7                     0.727273  \n",
       "8                     0.132530  \n",
       "9                     0.068966  \n",
       "10                    0.005516  \n",
       "11                    0.400000  \n",
       "12                    0.204545  \n",
       "13                    0.321429  \n",
       "14                    0.264706  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SGB_abundance_in_dataset=active_working_unknwon_cas9s.groupby([\"SGB ID\"]).count().sort_values(by=\"Seq ID\", ascending=False)[\"Seq ID\"].head(15)\n",
    "SGB_abundance_in_dataset=SGB_abundance_in_dataset.rename(\"# Genomes with act. wrk. cas9\")\n",
    "SGB_rel_ab_in_dataset=pd.DataFrame(SGB_table[SGB_table[\"SGB ID\"].isin(SGB_abundance_in_dataset.index)][[\"SGB ID\",\"# Reconstructed genomes\"]])\n",
    "SGB_abundance=pd.DataFrame(SGB_abundance_in_dataset).merge(SGB_rel_ab_in_dataset, left_on='SGB ID', right_on='SGB ID')\n",
    "SGB_abundance[\"Genomes relative abundance\"]=SGB_abundance[\"# Genomes with act. wrk. cas9\"]/SGB_abundance[\"# Reconstructed genomes\"]\n",
    "SGB_abundance"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All SGBs have a pretty high relative abundance, except for 15286 a, 14454 and 9280. And 8769 does not make any sense.Check that later.\n",
    "\n",
    "Let us keep all the loci from the SGBs with 10 or more genomes (each one of these genomes contains a working, active cas9 and belongs to the same species), that is, the first 12 SGBs.\n",
    "\n",
    "\n",
    "But, first, let's check whether all samples are coming from different datasets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SGB 9710 genomes all come from dataset HMP_2012\n"
     ]
    }
   ],
   "source": [
    "#returns SGBs with samples that only come from one dataset, if any.\n",
    "for SGB in SGB_abundance_in_dataset.index:\n",
    "    first_dataset=True\n",
    "    diff=False\n",
    "    for ind, Cas9 in active_working_unknwon_cas9s.iterrows():\n",
    "        if Cas9[\"SGB ID\"]==SGB:\n",
    "            dataset=Cas9[\"Genome Name\"].split(\"__\")[0]\n",
    "            if not first_dataset:\n",
    "                if dataset != temp_unique_dataset:\n",
    "                    diff=True\n",
    "            else:\n",
    "                temp_unique_dataset=dataset\n",
    "                first_dataset=False\n",
    "    if not diff:\n",
    "        print(\"SGB\",SGB,\"genomes all come from dataset\",dataset)\n",
    "\n",
    "                "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This might indicate a certain degree of replication in the data for that dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3 Data Clustering\n",
    "## 3.1 Cluster together identical sequences in the same SGB\n",
    "for the 12 most abundant unknown SGBs with a working active Cas9s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#trovata cosa ceh non va: dov'√® 14454 in questo file?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 204 unique sequences in 12 SGBs\n"
     ]
    }
   ],
   "source": [
    "def seq_getter(s): return str(s.seq)\n",
    "cons_table_949_1048=\"\"\n",
    "for SGB in [9340, 15299, 8767, 15095, 8769, 9710, 9281, 9311, 14454, 4329, 8774, 15286]:\n",
    "    current_SGB=active_working_unknwon_cas9s[active_working_unknwon_cas9s[\"SGB ID\"]==SGB]\n",
    "    alignments=[]\n",
    "    for index, row in current_SGB.iterrows():\n",
    "        tempseq=SeqRecord(Seq(row.Seq), id=str(row[\"SGB ID\"])+\"__\"+row[\"Seq ID\"]+\"__\"+row[\"Genome Name\"], description=row[\"Genome Name\"]+\"__\"+row[\"Seq ID\"]+\"__\"+str(row[\"SGB ID\"]))\n",
    "        alignments.append(tempseq)\n",
    "\n",
    "    filename=\"SGB\"+str(SGB)\n",
    "    SeqIO.write(alignments, outpath+filename+\".faa\", \"fasta\")\n",
    "    \n",
    "    fastafile=outpath+filename+\".faa\"\n",
    "    records = list(SeqIO.parse(fastafile,'fasta'))    \n",
    "    records.sort(key=seq_getter)\n",
    "    n=0\n",
    "    for seq,equal in groupby(records, seq_getter):\n",
    "        ids = ';'.join(s.id for s in equal)\n",
    "        N=ids.count(';')+1\n",
    "        \n",
    "        line=\">\"+ids+\";length\"+str(len(seq))+\";#sequences\"+str(N)+\"\\n\"+seq+\"\\n\"\n",
    "        cons_table_949_1048+=line\n",
    "filename2= outpath+\"949_1099_aa_Cas9_sequences_from_active_loci_from_12_most_abundant_most_unknown_species_identical\"       \n",
    "f=open(filename2+\".faa\",\"w\")\n",
    "f.write(cons_table_949_1048)\n",
    "f.close()\n",
    "sequences=list(SeqIO.parse(outpath+\"949_1099_aa_Cas9_sequences_from_active_loci_from_12_most_abundant_most_unknown_species_identical.faa\",'fasta'))\n",
    "print(\"There are\", len(sequences),\"unique sequences in 12 SGBs\")# clusters\",\"(\"+str(len(without_9710))+\" without 9710 SGB) at \"+identity_score+\" identity.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.2 ClustalW multiple sequence alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "clustalw -infile=/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/5caslocitable/out/949_1099_aa_Cas9_sequences_from_active_loci_from_12_most_abundant_most_unknown_species_identical.faa -outfile=/shares/CIBIO-Storage/CM/scratch/tmp_projects/signorini_cas/5caslocitable/out/949_1099_aa_Cas9_sequences_from_active_loci_from_12_most_abundant_most_unknown_species_identical.faa.aln\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alignments=[]\n",
    "cline= ClustalwCommandline(\"clustalw\", infile=filename2+\".faa\", outfile=filename2+\".faa.aln\")\n",
    "print(cline)\n",
    "os.system(str(cline))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3.3 Sequences clustering\n",
    "Cluster together sequences of up to 90% similarity, and extract one representative sequence for each cluster.\n",
    "Different clustering algorithms are available,the one that works the easiest was fast uclust https://drive5.com/usearch/manual/uclust_algo.html. Uclust performs centroid-based clustering and returns the centroid of each cluster as representative sequence.\n",
    "\n",
    "```\n",
    "usearch -cluster_fast 949_1099_aa_Cas9_sequences_from_active_loci_from_12_most_abundant_most_unknown_species_identical.faa -id 0.90 -centroids 9aCsfalf1musi_centroid90.faa -uc clusters.uc\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "uclust output parsing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 204 sequence clusters (21 without SGBs 9710,8769, 15286,14454, 9280) at 90 identity.\n",
      "SGB(s)\t# Sequence clusters\n",
      "---------------------------\n",
      "9311     4\n",
      "15299    3\n",
      "9710     2\n",
      "15286    2\n",
      "8769     2\n",
      "9340     2\n",
      "8774     1\n",
      "15095    1\n",
      "8767     1\n",
      "14454    1\n",
      "9281     1\n",
      "4329     1\n",
      "dtype: int64 \n",
      "\n",
      "There are 204 sequence clusters (24 without SGBs 9710,8769, 15286,14454, 9280) at 93 identity.\n",
      "SGB(s)\t# Sequence clusters\n",
      "---------------------------\n",
      "9311     5\n",
      "15299    4\n",
      "15286    3\n",
      "9710     2\n",
      "8769     2\n",
      "9340     2\n",
      "8774     1\n",
      "15095    1\n",
      "8767     1\n",
      "14454    1\n",
      "9281     1\n",
      "4329     1\n",
      "dtype: int64 \n",
      "\n",
      "There are 204 sequence clusters (28 without SGBs 9710,8769, 15286,14454, 9280) at 95 identity.\n",
      "SGB(s)\t# Sequence clusters\n",
      "---------------------------\n",
      "9311      5\n",
      "15299     4\n",
      "9710      3\n",
      "15286     3\n",
      "8774      2\n",
      "4329      2\n",
      "14454     2\n",
      "8769      2\n",
      "9340      2\n",
      "15095     1\n",
      "8767_C    1\n",
      "9281      1\n",
      "dtype: int64 \n",
      "\n",
      "There are 204 sequence clusters (38 without SGBs 9710,8769, 15286,14454, 9280) at 97 identity.\n",
      "SGB(s)\t# Sequence clusters\n",
      "---------------------------\n",
      "9311     7\n",
      "15299    6\n",
      "9710     3\n",
      "8774     3\n",
      "14454    3\n",
      "4329     3\n",
      "15286    3\n",
      "8769     3\n",
      "9340     3\n",
      "15095    2\n",
      "9281     1\n",
      "8767     1\n",
      "dtype: int64 \n",
      "\n",
      "There are 204 sequence clusters (83 without SGBs 9710,8769, 15286,14454, 9280) at 99 identity.\n",
      "SGB(s)\t# Sequence clusters\n",
      "---------------------------\n",
      "15299    13\n",
      "9311     13\n",
      "9340     10\n",
      "8774      9\n",
      "15095     7\n",
      "4329      7\n",
      "8767      6\n",
      "15286     6\n",
      "14454     4\n",
      "8769      4\n",
      "9710      3\n",
      "9281      1\n",
      "dtype: int64 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "for identity_score in [\"90\", \"93\",\"95\",\"97\", \"99\"]:\n",
    "    without_9710, without_8769, without_15286, without_14454, without_9280, without_sketcy_sgbs=[], [],[],[],[],[]\n",
    "    n_cluster_per_SGB=[]\n",
    "    for record in SeqIO.parse(outpath+\"9aCsfalf1musi_centroid\"+identity_score+\".faa\",'fasta'):\n",
    "        if not record.id.startswith(\"9710\") or not record.id.startswith(\"8769\") or not record.id.startswith(\"15286\") or not record.id.startswith(\"14454\") or not record.id.startswith(\"9280\"):\n",
    "            without_sketcy_sgbs.append(SeqRecord(record.seq, id=record.id, description=record.description))\n",
    "        SGBsnames=[oneid[:6].rstrip(\"__\") for oneid in record.id.split(\";\")]\n",
    "        SGBsnames=np.unique(SGBsnames) \n",
    "        SGBsnames=\",\".join(SGBsnames)\n",
    "        n_cluster_per_SGB.append(SGBsnames)\n",
    "    print(\"There are\", len(sequences),\"sequence clusters\",\"(\"+str(len(without_sketcy_sgbs))+\" without SGBs 9710,8769, 15286,14454, 9280) at \"+identity_score+\" identity.\")\n",
    "    print(\"SGB(s)\\t# Sequence clusters\\n---------------------------\")\n",
    "    print(pd.Series(n_cluster_per_SGB).value_counts(),\"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4 Pairwise alignment with known sequences\n",
    "\n",
    "perform pairwise alignment of cluster centroids with known working and not working cas9s. And visualize them nicely."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build a fasta file of the things to be aligned against\n",
    "#and of the bad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "not all arguments converted during string formatting",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-7-9e7427dcb19f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0;31m#print(pairwise2.format_alignment(*alignments[0]))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m       \u001b[0;31m#  print(\"##########################################################################\")\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m         \u001b[0mAlignIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malignments\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutpath\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"test_aln\"\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\"fasta\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/Bio/AlignIO/__init__.py\u001b[0m in \u001b[0;36mwrite\u001b[0;34m(alignments, handle, format)\u001b[0m\n\u001b[1;32m    227\u001b[0m                 \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malignment\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mMultipleSeqAlignment\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    228\u001b[0m                     raise TypeError(\"Expect a list or iterator of MultipleSeqAlignment \"\n\u001b[0;32m--> 229\u001b[0;31m                                     \"objects, got: %r\" % alignment)\n\u001b[0m\u001b[1;32m    230\u001b[0m                 \u001b[0mSeqIO\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0malignment\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m                 \u001b[0mcount\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: not all arguments converted during string formatting"
     ]
    }
   ],
   "source": [
    "from Bio import AlignIO\n",
    "identity_score='95'\n",
    "for record in SeqIO.parse(outpath+\"9aCsfalf1musi_centroid\"+identity_score+\".faa\", \"fasta\"):\n",
    "    for referencecas9record in SeqIO.parse(outpath+\"uniprot_working_Cas9s.fasta\", \"fasta\"):\n",
    "        alignments = pairwise2.align.globalxx(record.seq, referencecas9record.seq)\n",
    "        #print(pairwise2.format_alignment(*alignments[0]))\n",
    "      #  print(\"##########################################################################\")\n",
    "        AlignIO.write(alignments, outpath+\"test_aln\",\"fasta\"),\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Raw Cell Format",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
